<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Dina-Leigh Simons">

<title>Formatting and quality control of eDNA metabarcoding data (intertidal dataset)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="intertidal-eDNA-formatting-and-quality-control_files/libs/clipboard/clipboard.min.js"></script>
<script src="intertidal-eDNA-formatting-and-quality-control_files/libs/quarto-html/quarto.js"></script>
<script src="intertidal-eDNA-formatting-and-quality-control_files/libs/quarto-html/popper.min.js"></script>
<script src="intertidal-eDNA-formatting-and-quality-control_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="intertidal-eDNA-formatting-and-quality-control_files/libs/quarto-html/anchor.min.js"></script>
<link href="intertidal-eDNA-formatting-and-quality-control_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="intertidal-eDNA-formatting-and-quality-control_files/libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="intertidal-eDNA-formatting-and-quality-control_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="intertidal-eDNA-formatting-and-quality-control_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="intertidal-eDNA-formatting-and-quality-control_files/libs/bootstrap/bootstrap-1bc8a17f135ab3d594c857e9f48e611b.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#load-packages" id="toc-load-packages" class="nav-link active" data-scroll-target="#load-packages">Load packages</a></li>
  <li><a href="#importing-data" id="toc-importing-data" class="nav-link" data-scroll-target="#importing-data">Importing data</a>
  <ul class="collapse">
  <li><a href="#taxonomic-assignments" id="toc-taxonomic-assignments" class="nav-link" data-scroll-target="#taxonomic-assignments">Taxonomic assignments</a></li>
  <li><a href="#asv-data" id="toc-asv-data" class="nav-link" data-scroll-target="#asv-data">ASV data</a></li>
  <li><a href="#metadata" id="toc-metadata" class="nav-link" data-scroll-target="#metadata">Metadata</a></li>
  </ul></li>
  <li><a href="#data-wrangling-to-create-a-single-data-set" id="toc-data-wrangling-to-create-a-single-data-set" class="nav-link" data-scroll-target="#data-wrangling-to-create-a-single-data-set">Data wrangling to create a single data set</a>
  <ul class="collapse">
  <li><a href="#merge-taxonomic-assignments-and-asv-information" id="toc-merge-taxonomic-assignments-and-asv-information" class="nav-link" data-scroll-target="#merge-taxonomic-assignments-and-asv-information">Merge taxonomic assignments and ASV information</a></li>
  <li><a href="#remove-undetermined-sequences-in-18s-data" id="toc-remove-undetermined-sequences-in-18s-data" class="nav-link" data-scroll-target="#remove-undetermined-sequences-in-18s-data">Remove undetermined sequences in 18S data</a></li>
  <li><a href="#convert-to-long-format" id="toc-convert-to-long-format" class="nav-link" data-scroll-target="#convert-to-long-format">Convert to long format</a></li>
  <li><a href="#join-data-frames" id="toc-join-data-frames" class="nav-link" data-scroll-target="#join-data-frames">Join data frames</a></li>
  </ul></li>
  <li><a href="#cleaning-taxonomic-names" id="toc-cleaning-taxonomic-names" class="nav-link" data-scroll-target="#cleaning-taxonomic-names">Cleaning taxonomic names</a>
  <ul class="collapse">
  <li><a href="#remove-ambiguous-names" id="toc-remove-ambiguous-names" class="nav-link" data-scroll-target="#remove-ambiguous-names">Remove ambiguous names</a></li>
  <li><a href="#using-clean_names-in-janitor-package" id="toc-using-clean_names-in-janitor-package" class="nav-link" data-scroll-target="#using-clean_names-in-janitor-package">Using clean_names() in janitor package</a></li>
  </ul></li>
  <li><a href="#extracting-information-from-worms" id="toc-extracting-information-from-worms" class="nav-link" data-scroll-target="#extracting-information-from-worms">Extracting information from WoRMS</a>
  <ul class="collapse">
  <li><a href="#defining-functions" id="toc-defining-functions" class="nav-link" data-scroll-target="#defining-functions">Defining functions</a></li>
  <li><a href="#get-aphia-ids" id="toc-get-aphia-ids" class="nav-link" data-scroll-target="#get-aphia-ids">Get Aphia IDs</a></li>
  <li><a href="#get-taxonomic-metadata" id="toc-get-taxonomic-metadata" class="nav-link" data-scroll-target="#get-taxonomic-metadata">Get taxonomic metadata</a></li>
  <li><a href="#get-broad-functional-groups" id="toc-get-broad-functional-groups" class="nav-link" data-scroll-target="#get-broad-functional-groups">Get broad functional groups</a></li>
  </ul></li>
  <li><a href="#create-phyloseq-object" id="toc-create-phyloseq-object" class="nav-link" data-scroll-target="#create-phyloseq-object">Create phyloseq object</a>
  <ul class="collapse">
  <li><a href="#create-tax_table" id="toc-create-tax_table" class="nav-link" data-scroll-target="#create-tax_table">Create tax_table()</a></li>
  <li><a href="#create-sample_data" id="toc-create-sample_data" class="nav-link" data-scroll-target="#create-sample_data">Create sample_data()</a></li>
  <li><a href="#create-otu_table" id="toc-create-otu_table" class="nav-link" data-scroll-target="#create-otu_table">Create otu_table()</a></li>
  </ul></li>
  <li><a href="#decontamination" id="toc-decontamination" class="nav-link" data-scroll-target="#decontamination">Decontamination</a>
  <ul class="collapse">
  <li><a href="#visualizing-contaminants" id="toc-visualizing-contaminants" class="nav-link" data-scroll-target="#visualizing-contaminants">Visualizing contaminants</a></li>
  <li><a href="#removing-contaminants" id="toc-removing-contaminants" class="nav-link" data-scroll-target="#removing-contaminants">Removing contaminants</a></li>
  </ul></li>
  <li><a href="#filtering-steps" id="toc-filtering-steps" class="nav-link" data-scroll-target="#filtering-steps">Filtering steps</a>
  <ul class="collapse">
  <li><a href="#removal-of-low-read-samples" id="toc-removal-of-low-read-samples" class="nav-link" data-scroll-target="#removal-of-low-read-samples">Removal of low read samples</a></li>
  <li><a href="#filtering-for-target-functional-groups-broad" id="toc-filtering-for-target-functional-groups-broad" class="nav-link" data-scroll-target="#filtering-for-target-functional-groups-broad">Filtering for target functional groups (broad)</a></li>
  <li><a href="#filtering-for-target-taxa-more-specific" id="toc-filtering-for-target-taxa-more-specific" class="nav-link" data-scroll-target="#filtering-for-target-taxa-more-specific">Filtering for target taxa (more specific)</a></li>
  </ul></li>
  <li><a href="#further-quality-control" id="toc-further-quality-control" class="nav-link" data-scroll-target="#further-quality-control">Further quality control</a>
  <ul class="collapse">
  <li><a href="#set-up-functions-and-variables" id="toc-set-up-functions-and-variables" class="nav-link" data-scroll-target="#set-up-functions-and-variables">Set-up (functions and variables)</a></li>
  <li><a href="#explore-read-depths" id="toc-explore-read-depths" class="nav-link" data-scroll-target="#explore-read-depths">Explore read depths</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Formatting and quality control of eDNA metabarcoding data (intertidal dataset)</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Dina-Leigh Simons </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="load-packages" class="level2">
<h2 class="anchored" data-anchor-id="load-packages">Load packages</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>list.of.packages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"dplyr"</span>, </span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"tidyverse"</span>, </span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"phyloseq"</span>, </span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"seqinr"</span>, </span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"dada2"</span>, </span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"sjmisc"</span>, </span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"worrms"</span>,</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"taxize"</span>,</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"tibble"</span>, </span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"taxadb"</span>,</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"reshape2"</span>,</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"decontam"</span>,</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"microViz"</span>,</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"cowplot"</span>,</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"devtools"</span>,</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"qiime2R"</span>, </span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"microbiome"</span>,</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"vegan"</span>,</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"ggforce"</span>,</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"ANCOMBC"</span>)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>new.packages <span class="ot">&lt;-</span> list.of.packages[<span class="sc">!</span>(list.of.packages <span class="sc">%in%</span> <span class="fu">installed.packages</span>()[,<span class="st">"Package"</span>])]</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(new.packages)) <span class="fu">install.packages</span>(new.packages)</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a><span class="fu">invisible</span>(<span class="fu">lapply</span>(list.of.packages, library, <span class="at">character.only =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="importing-data" class="level2">
<h2 class="anchored" data-anchor-id="importing-data">Importing data</h2>
<p>We’ll first load all the required data, including outputs from the taxonomic assignment (.txt files), ASV data (.tsv and .fasta) and metadata (.csv).</p>
<section id="taxonomic-assignments" class="level3">
<h3 class="anchored" data-anchor-id="taxonomic-assignments">Taxonomic assignments</h3>
<p>We BLASTed ASVs against curated databases (MIDORI2 for CO1, Silva for 18S) separately for five regions: Scotland, North and South Wales, Northeast England, and Southwest England.</p>
<p>The BLAST results were then processed using MEtaGenome ANalyzer (MEGAN), which parses the alignments and assigns taxonomy using the Lowest Common Ancestor (LCA) algorithm. This approach ensures accurate taxonomic classification by considering multiple hits while conservatively resolving ambiguous assignments.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Scotland----</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>SCH_18S_taxa<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/MEGAN_data/SCH_18S_ASVs_blast_out_SILVA_98-ex.txt"</span>, <span class="at">header =</span> F) <span class="co">#Scotland 18S</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(SCH_18S_taxa)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ASV"</span>, <span class="st">"taxonomy"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>SCH_CO1_taxa<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/MEGAN_data/SCH_CO1_ASVs_blast_UNIQ_MIDORI_98-ex.txt"</span>, <span class="at">header =</span> F) <span class="co">#Scotland CO1</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(SCH_CO1_taxa)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ASV"</span>, <span class="st">"taxonomy"</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co">#North wales----</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>NW_18S_taxa<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/MEGAN_data/NW_18S_ASVs_blast_out_SILVA_98-ex.txt"</span>, <span class="at">header =</span> F) <span class="co">#North Wales 18S</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(NW_18S_taxa)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ASV"</span>, <span class="st">"taxonomy"</span>)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>NW_CO1_taxa<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/MEGAN_data/NW_CO1_ASVs_blast_UNIQ_MIDORI_98-ex_1per.txt"</span>, <span class="at">header =</span> F) <span class="co">#North Wales CO1</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(NW_CO1_taxa)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ASV"</span>, <span class="st">"taxonomy"</span>)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="co">#South Wales----</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>SW_18S_taxa<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/MEGAN_data/SW_18S_ASVs_blast_out_SILVA_98-ex.txt"</span>, <span class="at">header =</span> F) <span class="co">#South Wales 18S</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(SW_18S_taxa)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ASV"</span>, <span class="st">"taxonomy"</span>)</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>SW_CO1_taxa<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/MEGAN_data/SW_CO1_ASVs_blast_UNIQ_MIDORI_98-ex.txt"</span>, <span class="at">header =</span> F) <span class="co">#South Wales CO1</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(SW_CO1_taxa)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ASV"</span>, <span class="st">"taxonomy"</span>)</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="co">#Northeast England----</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>NE_18S_taxa<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/MEGAN_data/NE_18S_ASVs_blast_out_SILVA_98-ex.txt"</span>, <span class="at">header =</span> F) <span class="co">#Northeast 18S</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(NE_18S_taxa)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ASV"</span>, <span class="st">"taxonomy"</span>)</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>NE_CO1_taxa<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/MEGAN_data/NE_CO1_ASVs_blast_UNIQ_MIDORI_98-ex.txt"</span>, <span class="at">header =</span> F) <span class="co">#Northwast CO1</span></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(NE_CO1_taxa)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ASV"</span>, <span class="st">"taxonomy"</span>)</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a><span class="co">#Southwest England (Cornwall)----</span></span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>CN_18S_taxa<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/MEGAN_data/CN_18S_ASVs_blast_out_SILVA-1per_98.txt"</span>, <span class="at">header =</span> F) <span class="co">#Cornwall 18S</span></span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(CN_18S_taxa)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ASV"</span>, <span class="st">"taxonomy"</span>)</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>CN_CO1_taxa<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/MEGAN_data/CN_CO1_ASVs_blast_UNIQ_MIDORI-1per_98.txt"</span>, <span class="at">header =</span> F) <span class="co">#Cornwall CO1</span></span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(CN_CO1_taxa)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ASV"</span>, <span class="st">"taxonomy"</span>)</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a><span class="co">#Controls----</span></span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>Controls_18S_taxa<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/MEGAN_data/Controls_reps_18S_ASVs_blast_out_SILVA_98-ex.txt"</span>, <span class="at">header =</span> F) <span class="co">#Controls 18S</span></span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Controls_18S_taxa)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ASV"</span>, <span class="st">"taxonomy"</span>)</span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a>Controls_CO1_taxa<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/MEGAN_data/Repeats_CO1_ASVs_blast_UNIQ_MIDORI_98-ex.txt"</span>, <span class="at">header =</span> F) <span class="co">#Controls CO1</span></span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Controls_CO1_taxa)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ASV"</span>, <span class="st">"taxonomy"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s take a look at one of the MEGAN outputs. This file shows the ASVs detected in Scotland derived from the CO1 gene.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(SCH_CO1_taxa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        ASV                taxonomy
1   ASV_820 Paramoeba pemaquidensis
2 ASV_18576      Vexillifera kereti
3 ASV_22679      Vexillifera kereti
4   ASV_280       Parvamoeba rugata
5 ASV_10580       Parvamoeba rugata
6 ASV_20723       Parvamoeba rugata</code></pre>
</div>
</div>
</section>
<section id="asv-data" class="level3">
<h3 class="anchored" data-anchor-id="asv-data">ASV data</h3>
<p>Now we import the ASV count and sequence information for each region, obtained using DADA2. We have to do a bit of wrangling to get the fasta file into the correct format.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Scotland----</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>SCH_asv_count_18S <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/HPC_processed_data/SCH_18S/06_ASV_counts_SCH_18S.tsv"</span>) <span class="co">#Scotland 18S counts</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>SCH_asv_fasta_18S <span class="ot">&lt;-</span> <span class="fu">read.fasta</span>(<span class="st">"Input_Data/HPC_processed_data/SCH_18S/06_ASV_seqs_SCH_18S.fasta"</span>)<span class="co">#Scotland 18S fasta</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>SCH_asv_fasta_18S_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ASV=</span><span class="fu">names</span>(SCH_asv_fasta_18S), </span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">Seqs=</span><span class="fu">unlist</span>(<span class="fu">getSequence</span>(SCH_asv_fasta_18S, <span class="at">as.string=</span>T))) <span class="co">#turn into data frame</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>SCH_asv_fasta_18S_df<span class="sc">$</span>Seqs<span class="ot">&lt;-</span> <span class="fu">toupper</span>(SCH_asv_fasta_18S_df<span class="sc">$</span>Seqs) <span class="co">#upper case Seqs</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>SCH_asv_count_CO1 <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/HPC_processed_data/SCH_COI/06_ASV_counts_SCH_CO1_names.tsv"</span>) <span class="co">#Scotland CO1S counts</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>SCH_asv_fasta_CO1 <span class="ot">&lt;-</span> <span class="fu">read.fasta</span>(<span class="st">"Input_Data/HPC_processed_data/SCH_COI/06_ASV_seqs_SCH_CO1_names.fasta"</span>)<span class="co">#Scotland CO1 fasta</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>SCH_asv_fasta_CO1_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ASV=</span><span class="fu">names</span>(SCH_asv_fasta_CO1), </span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">Seqs=</span><span class="fu">unlist</span>(<span class="fu">getSequence</span>(SCH_asv_fasta_CO1, <span class="at">as.string=</span>T))) <span class="co">#turn into data frame</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>SCH_asv_fasta_CO1_df<span class="sc">$</span>Seqs<span class="ot">&lt;-</span> <span class="fu">toupper</span>(SCH_asv_fasta_CO1_df<span class="sc">$</span>Seqs) <span class="co">#upper case Seqs</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="co">#North Wales----</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>NW_asv_count_18S <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/HPC_processed_data/NW_18S/06_ASV_counts.tsv"</span>) <span class="co">#North Wales 18S counts</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>NW_asv_fasta_18S <span class="ot">&lt;-</span> <span class="fu">read.fasta</span>(<span class="st">"Input_Data/HPC_processed_data/NW_18S/06_ASV_seqs.fasta"</span>)<span class="co">#North Wales 18S fasta</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>NW_asv_fasta_18S_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ASV=</span><span class="fu">names</span>(NW_asv_fasta_18S), </span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">Seqs=</span><span class="fu">unlist</span>(<span class="fu">getSequence</span>(NW_asv_fasta_18S, <span class="at">as.string=</span>T))) <span class="co">#turn into data frame</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>NW_asv_fasta_18S_df<span class="sc">$</span>Seqs<span class="ot">&lt;-</span> <span class="fu">toupper</span>(NW_asv_fasta_18S_df<span class="sc">$</span>Seqs) <span class="co">#upper case Seqs</span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>NW_asv_count_CO1 <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/HPC_processed_data/NW_CO1/06_ASV_counts.tsv"</span>) <span class="co">#North Wales CO1 counts</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>NW_asv_fasta_CO1 <span class="ot">&lt;-</span> <span class="fu">read.fasta</span>(<span class="st">"Input_Data/HPC_processed_data/NW_CO1/06_ASV_seqs.fasta"</span>)<span class="co">#North Wales CO1 fasta</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>NW_asv_fasta_CO1_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ASV=</span><span class="fu">names</span>(NW_asv_fasta_CO1), </span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">Seqs=</span><span class="fu">unlist</span>(<span class="fu">getSequence</span>(NW_asv_fasta_CO1, <span class="at">as.string=</span>T))) <span class="co">#turn into data frame</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>NW_asv_fasta_CO1_df<span class="sc">$</span>Seqs<span class="ot">&lt;-</span> <span class="fu">toupper</span>(NW_asv_fasta_CO1_df<span class="sc">$</span>Seqs) <span class="co">#upper case Seqs</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a><span class="co">#South Wales----</span></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>SW_asv_count_18S <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/HPC_processed_data/SW_18S/06_ASV_counts_SW_18S.tsv"</span>) <span class="co">#South Wales 18S counts</span></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>SW_asv_fasta_18S <span class="ot">&lt;-</span> <span class="fu">read.fasta</span>(<span class="st">"Input_Data/HPC_processed_data/SW_18S/06_ASV_seqs_SW_18S.fasta"</span>) <span class="co">#South Wales 18S fasta</span></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>SW_asv_fasta_18S_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ASV=</span><span class="fu">names</span>(SW_asv_fasta_18S), </span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">Seqs=</span><span class="fu">unlist</span>(<span class="fu">getSequence</span>(SW_asv_fasta_18S, <span class="at">as.string=</span>T))) <span class="co">#turn into data frame</span></span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>SW_asv_fasta_18S_df<span class="sc">$</span>Seqs<span class="ot">&lt;-</span> <span class="fu">toupper</span>(SW_asv_fasta_18S_df<span class="sc">$</span>Seqs) <span class="co">#upper case Seqs</span></span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>SW_asv_count_CO1 <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/HPC_processed_data/SW_CO1/06_ASV_counts_SW_CO1_names.tsv"</span>) <span class="co">#South Wales CO1 counts</span></span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>SW_asv_fasta_CO1 <span class="ot">&lt;-</span> <span class="fu">read.fasta</span>(<span class="st">"Input_Data/HPC_processed_data/SW_CO1/06_ASV_seqs_SW_CO1_names.fasta"</span>) <span class="co">#South Wales CO1 fasta</span></span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>SW_asv_fasta_CO1_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ASV=</span><span class="fu">names</span>(SW_asv_fasta_CO1), </span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">Seqs=</span><span class="fu">unlist</span>(<span class="fu">getSequence</span>(SW_asv_fasta_CO1, <span class="at">as.string=</span>T))) <span class="co">#turn into data frame</span></span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a>SW_asv_fasta_CO1_df<span class="sc">$</span>Seqs<span class="ot">&lt;-</span> <span class="fu">toupper</span>(SW_asv_fasta_CO1_df<span class="sc">$</span>Seqs) <span class="co">#upper case Seqs</span></span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a><span class="co">#Northeast----</span></span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a>NE_asv_count_18S <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/HPC_processed_data/NE_18S/06_ASV_counts_NE.tsv"</span>) <span class="co">#Northeast 18S counts</span></span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a>NE_asv_fasta_18S <span class="ot">&lt;-</span> <span class="fu">read.fasta</span>(<span class="st">"Input_Data/HPC_processed_data/NE_18S/06_ASV_seqs_NE.fasta"</span>)<span class="co">#Northeast 18S fasta</span></span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a>NE_asv_fasta_18S_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ASV=</span><span class="fu">names</span>(NE_asv_fasta_18S), </span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">Seqs=</span><span class="fu">unlist</span>(<span class="fu">getSequence</span>(NE_asv_fasta_18S, <span class="at">as.string=</span>T))) <span class="co">#turn into data frame</span></span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a>NE_asv_fasta_18S_df<span class="sc">$</span>Seqs<span class="ot">&lt;-</span> <span class="fu">toupper</span>(NE_asv_fasta_18S_df<span class="sc">$</span>Seqs) <span class="co">#upper case Seqs</span></span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a>NE_asv_count_CO1 <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/HPC_processed_data/NE_CO1/06_ASV_counts.tsv"</span>) <span class="co">#Northeast CO1 counts</span></span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a>NE_asv_fasta_CO1 <span class="ot">&lt;-</span> <span class="fu">read.fasta</span>(<span class="st">"Input_Data/HPC_processed_data/NE_CO1/06_ASV_seqs.fasta"</span>)<span class="co">#Northeast CO1 fasta</span></span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a>NE_asv_fasta_CO1_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ASV=</span><span class="fu">names</span>(NE_asv_fasta_CO1), </span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">Seqs=</span><span class="fu">unlist</span>(<span class="fu">getSequence</span>(NE_asv_fasta_CO1, <span class="at">as.string=</span>T))) <span class="co">#turn into data frame</span></span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a>NE_asv_fasta_CO1_df<span class="sc">$</span>Seqs<span class="ot">&lt;-</span> <span class="fu">toupper</span>(NE_asv_fasta_CO1_df<span class="sc">$</span>Seqs) <span class="co">#upper case Seqs</span></span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a><span class="co">#Cornwall----</span></span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a>CN_asv_count_18S <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/HPC_processed_data/CN_18S/06_ASV_counts.tsv"</span>) <span class="co">#Cornwall 18S counts</span></span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a>CN_asv_fasta_18S <span class="ot">&lt;-</span> <span class="fu">read.fasta</span>(<span class="st">"Input_Data/HPC_processed_data/CN_18S/06_ASV_seqs.fasta"</span>)<span class="co">#Cornwall 18S fasta</span></span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a>CN_asv_fasta_18S_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ASV=</span><span class="fu">names</span>(CN_asv_fasta_18S), </span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">Seqs=</span><span class="fu">unlist</span>(<span class="fu">getSequence</span>(CN_asv_fasta_18S, <span class="at">as.string=</span>T))) <span class="co">#turn into data frame</span></span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a>CN_asv_fasta_18S_df<span class="sc">$</span>Seqs<span class="ot">&lt;-</span> <span class="fu">toupper</span>(CN_asv_fasta_18S_df<span class="sc">$</span>Seqs) <span class="co">#upper case Seqs</span></span>
<span id="cb5-59"><a href="#cb5-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-60"><a href="#cb5-60" aria-hidden="true" tabindex="-1"></a>CN_asv_count_CO1 <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/HPC_processed_data/CN_CO1/06_ASV_counts.tsv"</span>) <span class="co">#Cornwall CO1 counts</span></span>
<span id="cb5-61"><a href="#cb5-61" aria-hidden="true" tabindex="-1"></a>CN_asv_fasta_CO1 <span class="ot">&lt;-</span> <span class="fu">read.fasta</span>(<span class="st">"Input_Data/HPC_processed_data/CN_CO1/06_ASV_seqs.fasta"</span>)<span class="co">#Cornwall CO1 fasta</span></span>
<span id="cb5-62"><a href="#cb5-62" aria-hidden="true" tabindex="-1"></a>CN_asv_fasta_CO1_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ASV=</span><span class="fu">names</span>(CN_asv_fasta_CO1), </span>
<span id="cb5-63"><a href="#cb5-63" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">Seqs=</span><span class="fu">unlist</span>(<span class="fu">getSequence</span>(CN_asv_fasta_CO1, <span class="at">as.string=</span>T))) <span class="co">#turn into data frame</span></span>
<span id="cb5-64"><a href="#cb5-64" aria-hidden="true" tabindex="-1"></a>CN_asv_fasta_CO1_df<span class="sc">$</span>Seqs<span class="ot">&lt;-</span> <span class="fu">toupper</span>(CN_asv_fasta_CO1_df<span class="sc">$</span>Seqs) <span class="co">#upper case Seqs</span></span>
<span id="cb5-65"><a href="#cb5-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-66"><a href="#cb5-66" aria-hidden="true" tabindex="-1"></a><span class="co">#Controls</span></span>
<span id="cb5-67"><a href="#cb5-67" aria-hidden="true" tabindex="-1"></a>controls_PhD2_asv_count_18S <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/HPC_processed_data/Controls_18S/06_ASV_counts_Controls.tsv"</span>) <span class="co">#controls 18S counts</span></span>
<span id="cb5-68"><a href="#cb5-68" aria-hidden="true" tabindex="-1"></a>controls_PhD2_asv_fasta_18S <span class="ot">&lt;-</span> <span class="fu">read.fasta</span>(<span class="st">"Input_Data/HPC_processed_data/Controls_18S/06_ASV_seqs_Controls.fasta"</span>)<span class="co">#controls 18S fasta</span></span>
<span id="cb5-69"><a href="#cb5-69" aria-hidden="true" tabindex="-1"></a>controls_PhD2_asv_fasta_18S_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ASV=</span><span class="fu">names</span>(controls_PhD2_asv_fasta_18S), </span>
<span id="cb5-70"><a href="#cb5-70" aria-hidden="true" tabindex="-1"></a>                                             <span class="at">Seqs=</span><span class="fu">unlist</span>(<span class="fu">getSequence</span>(controls_PhD2_asv_fasta_18S, <span class="at">as.string=</span>T))) <span class="co">#turn into data frame</span></span>
<span id="cb5-71"><a href="#cb5-71" aria-hidden="true" tabindex="-1"></a>controls_PhD2_asv_fasta_18S_df<span class="sc">$</span>Seqs<span class="ot">&lt;-</span> <span class="fu">toupper</span>(controls_PhD2_asv_fasta_18S_df<span class="sc">$</span>Seqs) <span class="co">#upper case Seqs</span></span>
<span id="cb5-72"><a href="#cb5-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-73"><a href="#cb5-73" aria-hidden="true" tabindex="-1"></a>controls_PhD2_asv_count_CO1 <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Input_Data/HPC_processed_data/Repeats_CO1/06_ASV_counts.tsv"</span>) <span class="co">#controls CO1 counts</span></span>
<span id="cb5-74"><a href="#cb5-74" aria-hidden="true" tabindex="-1"></a>controls_PhD2_asv_fasta_CO1 <span class="ot">&lt;-</span> <span class="fu">read.fasta</span>(<span class="st">"Input_Data/HPC_processed_data/Repeats_CO1/06_ASV_seqs.fasta"</span>)<span class="co">#controls  CO1 fasta</span></span>
<span id="cb5-75"><a href="#cb5-75" aria-hidden="true" tabindex="-1"></a>controls_PhD2_asv_fasta_CO1_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ASV=</span><span class="fu">names</span>(controls_PhD2_asv_fasta_CO1), </span>
<span id="cb5-76"><a href="#cb5-76" aria-hidden="true" tabindex="-1"></a>                                             <span class="at">Seqs=</span><span class="fu">unlist</span>(<span class="fu">getSequence</span>(controls_PhD2_asv_fasta_CO1, <span class="at">as.string=</span>T))) <span class="co">#turn into data frame</span></span>
<span id="cb5-77"><a href="#cb5-77" aria-hidden="true" tabindex="-1"></a>controls_PhD2_asv_fasta_CO1_df<span class="sc">$</span>Seqs<span class="ot">&lt;-</span> <span class="fu">toupper</span>(controls_PhD2_asv_fasta_CO1_df<span class="sc">$</span>Seqs) <span class="co">#upper case Seqs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s take a look at an example of the two imported ASV files. Again, these shows the ASVs detected in Scotland derived from the CO1 gene.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(SCH_asv_count_CO1) <span class="co">#counts</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   25524 obs. of  69 variables:
 $ Adapter3616.SCH01.04rep: int  4093 1277 1446 285 247 8426 200 1132 74 76 ...
 $ Adapter3618.SCH04.09rep: int  6540 0 4951 13980 0 25 124 1022 141 31 ...
 $ Adapter3619.SCH07.02rep: int  4514 5476 44 0 7098 37 0 1062 0 2265 ...
 $ Adapter385.SCH01.01    : int  14226 34 0 3645 55 14877 72 3995 0 0 ...
 $ Adapter386.SCH01.02    : int  17167 97 57 128 29 11294 34 3222 0 0 ...
 $ Adapter387.SCH01.03    : int  11759 13 0 65 19 12901 0 2801 0 0 ...
 $ Adapter388.SCH01.04    : int  7043 1235 2119 409 116 10103 265 1971 212 99 ...
 $ Adapter389.SCH01.05    : int  6969 149 220 0 12746 8130 14 1907 0 0 ...
 $ Adapter390.SCH01.06    : int  1595 54 41322 2382 480 3344 25 578 0 0 ...
 $ Adapter391.SCH01.07    : int  5347 31 444 553 9 14428 4346 1909 0 82 ...
 $ Adapter392.SCH01.08    : int  3505 202 113 1132 9 17493 10475 1000 0 74 ...
 $ Adapter393.SCH01.09    : int  4292 57 152 1186 0 8626 3700 957 0 83 ...
 $ Adapter394.SCH01.C     : int  0 8 0 0 0 0 0 0 0 0 ...
 $ Adapter395.SCH02.01    : int  11959 62 120 715 0 392 1396 1826 0 98 ...
 $ Adapter396.SCH02.02    : int  2087 16 0 131 0 48 537 386 8 52 ...
 $ Adapter398.SCH02.04    : int  38 0 0 1782 0 0 0 0 56662 0 ...
 $ Adapter400.SCH02.06    : int  602 0 0 8866 0 40 0 162 24149 0 ...
 $ Adapter401.SCH02.C     : int  14 0 0 12 0 22 0 5 10 0 ...
 $ Adapter402.SCHNC1ex    : int  0 0 0 0 0 0 0 0 4 0 ...
 $ Adapter403.SCH03.01    : int  18895 304 5239 1427 75 417 470 3982 40 10 ...
 $ Adapter404.SCH03.02    : int  7478 22 3145 2473 34 185 13917 1091 0 0 ...
 $ Adapter405.SCH03.03    : int  17104 892 1097 250 19 725 569 3290 0 0 ...
 $ Adapter406.SCH03.04    : int  15268 24 2484 947 18 418 7 2957 259 0 ...
 $ Adapter407.SCH03.05    : int  7185 0 4943 2569 40 298 896 1131 673 0 ...
 $ Adapter408.SCH03.06    : int  13432 48 2614 572 59 633 241 2778 59 7 ...
 $ Adapter409.SCH03.07    : int  13391 63 91 70 52 1427 252 2817 0 197 ...
 $ Adapter410.SCH03.08    : int  15437 21 42 0 11 339 17 3480 0 56 ...
 $ Adapter411.SCH03.09    : int  11703 64 246 340 76 1711 242 2267 0 295 ...
 $ Adapter412.SCH03.C     : int  9 0 0 0 0 0 0 0 0 0 ...
 $ Adapter413.SCH04.01    : int  21451 6 247 99 43 221 2253 3042 0 299 ...
 $ Adapter414.SCH04.02    : int  37259 15 448 87 28561 340 7067 4422 0 226 ...
 $ Adapter416.SCH04.04    : int  46854 1890 2082 245 0 14 11 8926 0 0 ...
 $ Adapter417.SCH04.05    : int  26878 84 4754 5350 58 416 30879 4635 0 62 ...
 $ Adapter418.SCH04.06    : int  20521 6672 1237 0 1141 654 284 3087 0 161 ...
 $ Adapter419.SCH04.07    : int  5368 0 417 6578 0 0 0 778 0 5 ...
 $ Adapter420.SCH04.08    : int  10112 8 7201 11499 0 6 993 1698 81 68 ...
 $ Adapter421.SCH04.09    : int  10247 0 6344 18824 0 184 309 1865 382 132 ...
 $ Adapter422.SCH04.C     : int  45 0 9 0 21 0 9 13 10 0 ...
 $ Adapter423.SCHNC2ex    : int  48 0 10 0 23 9 0 5 0 0 ...
 $ Adapter424.SCH05.01    : int  11751 2887 69 91 131 455 117 2468 0 2658 ...
 $ Adapter425.SCH05.02    : int  8230 1134 57 0 77 272 33 1500 0 1177 ...
 $ Adapter426.SCH05.03    : int  10757 97 0 8 0 15 0 1994 0 116 ...
 $ Adapter427.SCH05.C     : int  0 0 0 0 0 0 0 0 0 0 ...
 $ Adapter428.SCH06.01    : int  6952 526 9164 238 3594 28 1040 942 0 1649 ...
 $ Adapter429.SCH06.02    : int  10309 26677 40 0 1339 30 969 1732 0 6878 ...
 $ Adapter430.SCH06.03    : int  6651 2433 92 0 25632 25 1550 972 0 6207 ...
 $ Adapter431.SCH06.04    : int  8922 14828 25 0 3810 42 77 1228 0 4183 ...
 $ Adapter432.SCH06.05    : int  5348 9687 17 0 2432 35 36 960 0 3236 ...
 $ Adapter433.SCH06.06    : int  9116 11166 8 0 3847 68 64 1360 0 7572 ...
 $ Adapter434.SCH06.C     : int  12 5 0 0 4 0 0 0 0 0 ...
 $ Adapter435.SCH07.01    : int  8558 23342 609 0 24425 61 117 1565 0 10786 ...
 $ Adapter436.SCH07.02    : int  11804 10739 138 0 14036 61 0 3911 0 4291 ...
 $ Adapter437.SCH07.03    : int  2289 2718 142 0 5587 24 106 528 0 6173 ...
 $ Adapter438.SCH07.C     : int  6 0 0 0 0 0 0 0 0 0 ...
 $ Adapter439.SCH08.01    : int  10664 77000 117 0 3881 222 92 2333 0 0 ...
 $ Adapter440.SCH08.02    : int  4825 60236 0 0 3213 49 0 775 0 0 ...
 $ Adapter441.SCH08.03    : int  2271 40980 38 0 2888 54 0 513 0 0 ...
 $ Adapter442.SCH08.04    : int  159 36 15855 12775 42 3 7671 57 0 0 ...
 $ Adapter443.SCH08.05    : int  15650 17873 38455 9310 460 61 2396 2530 0 0 ...
 $ Adapter444.SCH08.06    : int  11373 195 40361 818 17 58 213 2263 17 0 ...
 $ Adapter445.SCH08.07    : int  21058 66 6206 15798 11 178 19735 5069 1045 0 ...
 $ Adapter446.SCH08.08    : int  149 20 1213 37507 6 5 1803 59 924 0 ...
 $ Adapter447.SCH08.09    : int  21 0 0 453 0 0 0 0 25 0 ...
 $ Adapter448.SCH08.C     : int  0 0 0 0 0 0 0 0 4 0 ...
 $ Adapter449.SCHNC3ex    : int  31 296 0 0 45 0 0 9 0 0 ...
 $ Adapter450.SCHNCPCR    : int  0 0 105 0 0 0 51 0 0 0 ...
 $ Adapter451.SCHPC       : int  0 0 0 0 0 0 0 0 0 0 ...
 $ Adapter452.SCHPC2803   : int  0 0 0 0 0 0 0 0 0 0 ...
 $ Adapter453.SCHNC2803   : int  95 0 26 53 0 0 91 24 0 0 ...</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(SCH_asv_fasta_CO1_df) <span class="co">#sequences</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    ASV
1 ASV_1
2 ASV_2
3 ASV_3
4 ASV_4
5 ASV_5
6 ASV_6
                                                                                                                                                                                                                                                                                                                                                              Seqs
1                                        ACTAAGTCATATTACTAGTCACTCAGGAGGTGCTGTAGATTTAGCTATTTTTAGTTTACACGTTTCAGGTGCAAGTAGTATTTTAGGTGCAATCAACTTTATTACTACCATCTTTAATATGCGTGGACCTGGATTATTAATGCATCGCTTACCTTTATTTGTATGGAGTGTTTTAATTACAGCGTTTTTACTTCTTCTCTCATTACCTGTACTTGCAGGTGCAATTACAATGTTACTAACTGATCGAAACTTTTCAACCAGTTTCTTTGACCCAAGTGGTGGAGGAGACCCTATTTTATATCAACATTTATTC
2                                        CCTTAGTGGTATTCAGGCTCACTCGGGGCCTTCTGTTGATTTAGCTATATTCAGTCTTCATCTCTCAGGTGCTGCTTCTATTTTAGGTGCTATAAATTTTATCACAACAATTTTTAATATGAGAGCTCCTGGTATGACAATGGATAGAGTACCTCTTTTTGTATGGTCTGTCTTAATCACAGCGTTTTTGTTACTGTTATCGCTTCCTGTTTTGGCAGGTGGTATTACAATGTTATTGACAGATCGTAACTTTAATACTACTTTCTTTGATCCGGCAGGTGGTGGTGATCCAGTATTGTATCAGCATTTATTT
3                                     ATTAAGCACTTCTTTCCTCAGTTTATCACCTTCAAGTACAGCTTTCATAGTCTTTGGATTATTAATGTCAGGTATATCCTCATCTCTCACATCTGTAAACTTTTGGACAACAATTATAAACATGAGATCTTATTATCTGACAATGAAGACTATGCCATTATTCCCTTGGAGCCTTTTGATAACTTCTGGAATGCTTTTATTAACATTACCAATCTTATCAGGAGCTCTTCTAATGGTCTTGGGTGATCTTCATTCTAATACACTTTTCTTTGATCCAATCTTTGGAGGAGATCCTATATTCTATCAACACTTATTT
4                                     ATTAAGCACTTCTTTCCTCAGTTTATCACCTTCAAGTACAGCTTTCATAGTCTTTGGATTATTAATGTCAGGTATATCCTCATCTCTCACATCTGTAAACTTTTGGACAACAATTATAAACATGAGATCTTATTATCTGACAATGAAGACTATGCCATTATTCCCTTGGAGCCTTTTGATAACTTCTGGAATGCTTTTATTAACATTACCAATCTTATCAGGAGCTCTTCTAATGGTCTTCGGTGATCTTCATTCTAATACACTTTTCTTTGATCCAATCTTTGGAGGAGATCCTATATTCTATCAACACTTATTT
5                                        ATTAAGTGGTATTCAAGCACACTCAGGGCCTTCAGTTGATTTAGCTATTTTTAGTTTACATTTATCTGGAGCGGCTTCAATTTTAGGAGCTATTAACTTTATTACTACAATCTTTAATATGCGTGCTCCTGGTATGACTATGGATAGATTACCTCTTTTTGTTTGGTCTGTTTTAATAACAGCATTCTTACTTTTATTATCCCTTCCTGTTTTAGCTGGTGGTATTACAATGCTTTTAACAGATAGAAATTTTAATACTACTTTTTTTGACCCTGCTGGTGGTGGTGATCCTGTATTATACCAACACTTATTT
6 ATTAAGTATTTTGGAACAAGCTATACCTGGATCGGGACTTGGAATGACTTTGTGGTTATTAAGTATGACACTGTTTATAGCTTCCTCATTAATTGGGGCGCTTAATTACATTGTTACAATTATTAACTTAAGAACCATTGGAATGAAAATGACAAGATTGCCACTTACTATTTGGGCATTCTTTGTAACTGCAATTTTAGGTGTACTTTCTTTTCCAGTATTAGTATCAGCAGTATTGTTGTTATTAATGGATAAGACTTTTGGAACAAGTTTCTACTTGTCTGATATCTTTATTGGAGGAGAAGCATTAGATGCATCTGGAGGTTCACCAATATTATACCAACATTTATTT</code></pre>
</div>
</div>
</section>
<section id="metadata" class="level3">
<h3 class="anchored" data-anchor-id="metadata">Metadata</h3>
<p>Finally, let’s import any metadata associated to our samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>sites <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"Input_Data/Metadata/eDNA_Site_Rockpool_Data_Oct2023.csv"</span>, <span class="at">na.strings=</span><span class="fu">c</span>(<span class="st">""</span>,<span class="st">"NA"</span>))</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>IDs_stageone <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"Input_Data/Metadata/sample_ID_matching_stageone.csv"</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>IDs_stagetwo <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"Input_Data/Metadata/sample_ID_matching_stagetwo.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And let’s take a quick look at the metadata files too. You can see there are 63 variables which are associated to our samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(sites)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   374 obs. of  63 variables:
 $ eventID              : chr  "PHD1-2023.03.23-Neg-bead" "PHD1-2023.03.29-Neg-PCRrep" "PHD1-2023.03.29-Pos-PCRrep" "PHD1-2022.06.12-SCH01-01" ...
 $ fieldID              : chr  "PHD1-NCbead" "PHD1-NCrep" "PHD1-PCrep" "PHD1-SCH01-01" ...
 $ projectID            : chr  "PHD1" "PHD1" "PHD1" "PHD1" ...
 $ localityID           : chr  NA NA NA "SCH01" ...
 $ sampleID             : chr  NA NA NA "1" ...
 $ sampleType           : chr  "lab-negative-control" "lab-negative-control" "lab-positive-control" "sample" ...
 $ controlCheck         : chr  "control" "control" "positive" "sample" ...
 $ negCheckLogical      : logi  TRUE TRUE FALSE FALSE FALSE FALSE ...
 $ pairedRepLogical     : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
 $ batch_extraction     : int  NA NA NA 1 1 1 1 1 1 1 ...
 $ batch_PCR_CO1        : int  1 1 1 2 2 2 2 1 2 2 ...
 $ plate_col_CO1        : int  7 8 9 1 1 1 1 8 1 1 ...
 $ plate_row_CO1        : chr  "H" "G" "D" "A" ...
 $ batch_PCR_18S        : int  NA 3 3 4 4 4 4 3 4 4 ...
 $ plate_col_18S        : int  NA 8 9 1 1 1 1 8 1 1 ...
 $ plate_row_18S        : chr  NA "D" "A" "A" ...
 $ date_extraction      : chr  NA NA NA "24.02.2023" ...
 $ date_amplified_CO1   : chr  "22.03.2023" "29.03.2023" "29.03.2023" "28.03.2023" ...
 $ date_amplified_18S   : chr  NA "29.03.2023" "29.03.2023" "27.03.2023" ...
 $ sample_18S           : chr  NA "S60" "S65" "S66" ...
 $ sample_CO1           : chr  "Adapter3608.NCbead" "Adapter3615.NCPCRrep" "Adapter3620.PCrep" "Adapter385.SCH01.01" ...
 $ COUNT                : int  1 2 2 2 2 2 2 2 2 2 ...
 $ dna_concentration_CO1: num  4.63 28.76 73.22 82.54 70.27 ...
 $ dna_concentration_18S: num  NA 46.3 81.6 55.4 37 ...
 $ continent            : chr  NA NA NA "Europe" ...
 $ country              : chr  NA NA NA "Scotland" ...
 $ county               : chr  NA NA NA "Sutherland" ...
 $ waterBody            : chr  NA NA NA "rocky" ...
 $ verbatimLocality     : chr  NA NA NA "Scourie" ...
 $ exposure             : chr  NA NA NA "Exposed" ...
 $ weather              : chr  NA NA NA "moderate rain, strong winds, very cloudy" ...
 $ tide                 : chr  NA NA NA "outgoing" ...
 $ lowWater             : chr  NA NA NA "12:23" ...
 $ decimalLongitude     : num  NA NA NA -5.17 -5.17 ...
 $ decimalLatitude      : num  NA NA NA 58.4 58.4 ...
 $ recordedBy           : chr  NA NA NA "Dina-Leigh Simons; Nova Mieszkowska" ...
 $ year                 : int  NA NA NA 2022 2022 2022 2022 2022 2022 2022 ...
 $ month                : int  NA NA NA 6 6 6 6 6 6 6 ...
 $ day                  : int  NA NA NA 12 12 12 12 12 12 12 ...
 $ eventTime            : chr  NA NA NA "09:31" ...
 $ samplingProtocol     : chr  NA NA NA "Sterivex-0.22" ...
 $ pumpType             : chr  NA NA NA "Syringe 60mL, sealant gun" ...
 $ repVol..mL.          : int  NA NA NA 1000 1000 800 1000 1000 1000 1000 ...
 $ algaeCover.          : int  NA NA NA 0 5 60 90 90 90 95 ...
 $ remarks              : chr  NA NA NA "Filtered on shore" ...
 $ daysBeforefreeze     : int  NA NA NA 9 9 9 9 9 9 9 ...
 $ preservation         : chr  NA NA NA "EtOH" ...
 $ type                 : chr  NA NA NA "Rockpool" ...
 $ shorePosition        : chr  NA NA NA "High" ...
 $ rockpoolarea.cm.2.   : num  NA NA NA 7443 23218 ...
 $ depth1               : num  NA NA NA NA NA NA NA NA NA NA ...
 $ depth2               : num  NA NA NA NA NA NA NA NA NA NA ...
 $ averageDepth..cm.    : num  NA NA NA 25 19.1 13.9 1100 1100 24.7 25.6 ...
 $ rockpoolVol.m.3.     : num  NA NA NA 0.186 0.444 ...
 $ pH1                  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ pH2                  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ averagePH            : num  NA NA NA 8.59 8.78 8.47 8.63 8.63 9.54 8.84 ...
 $ temp1                : chr  NA NA NA NA ...
 $ temp2                : chr  NA NA NA NA ...
 $ averageTemp          : num  NA NA NA 13 13.6 ...
 $ siteImage            : chr  NA NA NA "IMG_SCH01-01.jpeg" ...
 $ rockType             : chr  NA NA NA "metamorphic" ...
 $ rockFold             : chr  NA NA NA NA ...</code></pre>
</div>
</div>
<p>We also have some data to help tidy up sample IDs later in the pipeline.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(IDs_stageone) <span class="co">#labels</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            sample_ID sample_18S           sample_CO1
1 PHD1-SCHNC-PCR_1703       &lt;NA&gt;  Adapter450.SCHNCPCR
2     PHD1-SCHPC_1703       &lt;NA&gt;     Adapter451.SCHPC
3      PHD1-SCHNC2803       &lt;NA&gt; Adapter453.SCHNC2803
4      PHD1-SCHPC2803       &lt;NA&gt; Adapter452.SCHPC2803
5         PHD1-NCbead       &lt;NA&gt;   Adapter3608.NCbead
6  PHD1-SWNC_PCR_2203       &lt;NA&gt;  Adapter3607.SWNCPCR</code></pre>
</div>
</div>
</section>
</section>
<section id="data-wrangling-to-create-a-single-data-set" class="level2">
<h2 class="anchored" data-anchor-id="data-wrangling-to-create-a-single-data-set">Data wrangling to create a single data set</h2>
<p>We have lots of different files here which need to be manipulated into one single data frame for each primer. A few steps are required to achieve this.</p>
<section id="merge-taxonomic-assignments-and-asv-information" class="level3">
<h3 class="anchored" data-anchor-id="merge-taxonomic-assignments-and-asv-information">Merge taxonomic assignments and ASV information</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Scotland----</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>SCH_asv_18S_merged <span class="ot">&lt;-</span> <span class="fu">merge</span>(SCH_asv_fasta_18S_df, SCH_asv_count_18S, <span class="at">by.x =</span> <span class="st">"Seqs"</span>, <span class="at">by.y =</span> <span class="st">"row.names"</span>) <span class="co">#add ASVs to count data</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>SCH_taxa_18S <span class="ot">&lt;-</span> <span class="fu">full_join</span>(SCH_18S_taxa, SCH_asv_18S_merged, <span class="at">by =</span> <span class="st">"ASV"</span>) <span class="co">#full_join keeps ASVs which didn't get assigned, left_join would remove unassigned</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>SCH_asv_CO1_merged <span class="ot">&lt;-</span> <span class="fu">merge</span>(SCH_asv_fasta_CO1_df, SCH_asv_count_CO1, <span class="at">by.x =</span> <span class="st">"Seqs"</span>, <span class="at">by.y =</span> <span class="st">"row.names"</span>) <span class="co">#add ASVs to count data</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>SCH_taxa_CO1 <span class="ot">&lt;-</span> <span class="fu">full_join</span>(SCH_CO1_taxa, SCH_asv_CO1_merged, <span class="at">by =</span> <span class="st">"ASV"</span>) <span class="co">#full_join keeps ASVs which didn't get assigned, left_join would remove unassigned</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co">#North Wales----</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>NW_asv_18S_merged <span class="ot">&lt;-</span> <span class="fu">merge</span>(NW_asv_fasta_18S_df, NW_asv_count_18S, <span class="at">by.x =</span> <span class="st">"Seqs"</span>, <span class="at">by.y =</span> <span class="st">"row.names"</span>) <span class="co">#add ASVs to count data</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>NW_taxa_18S <span class="ot">&lt;-</span> <span class="fu">full_join</span>(NW_18S_taxa, NW_asv_18S_merged, <span class="at">by =</span> <span class="st">"ASV"</span>) <span class="co">#full_join keeps ASVs which didn't get assigned, left_join would remove unassigned</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>NW_asv_CO1_merged <span class="ot">&lt;-</span> <span class="fu">merge</span>(NW_asv_fasta_CO1_df, NW_asv_count_CO1, <span class="at">by.x =</span> <span class="st">"Seqs"</span>, <span class="at">by.y =</span> <span class="st">"row.names"</span>) <span class="co">#add ASVs to count data</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>NW_taxa_CO1 <span class="ot">&lt;-</span> <span class="fu">full_join</span>(NW_CO1_taxa, NW_asv_CO1_merged, <span class="at">by =</span> <span class="st">"ASV"</span>) <span class="co">#full_join keeps ASVs which didn't get assigned, left_join would remove unassigned</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a><span class="co">#South Wales----</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>SW_asv_18S_merged <span class="ot">&lt;-</span> <span class="fu">merge</span>(SW_asv_fasta_18S_df, SW_asv_count_18S, <span class="at">by.x =</span> <span class="st">"Seqs"</span>, <span class="at">by.y =</span> <span class="st">"row.names"</span>) <span class="co">#add ASVs to count data</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>SW_taxa_18S <span class="ot">&lt;-</span> <span class="fu">full_join</span>(SW_18S_taxa, SW_asv_18S_merged, <span class="at">by =</span> <span class="st">"ASV"</span>) <span class="co">#full_join keeps ASVs which didn't get assigned, left_join would remove unassigned</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>SW_asv_CO1_merged <span class="ot">&lt;-</span> <span class="fu">merge</span>(SW_asv_fasta_CO1_df, SW_asv_count_CO1, <span class="at">by.x =</span> <span class="st">"Seqs"</span>, <span class="at">by.y =</span> <span class="st">"row.names"</span>) <span class="co">#add ASVs to count data</span></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>SW_taxa_CO1 <span class="ot">&lt;-</span> <span class="fu">full_join</span>(SW_CO1_taxa, SW_asv_CO1_merged, <span class="at">by =</span> <span class="st">"ASV"</span>) <span class="co">#full_join keeps ASVs which didn't get assigned, left_join would remove unassigned</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a><span class="co">#Northeast----</span></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>NE_asv_18S_merged <span class="ot">&lt;-</span> <span class="fu">merge</span>(NE_asv_fasta_18S_df, NE_asv_count_18S, <span class="at">by.x =</span> <span class="st">"Seqs"</span>, <span class="at">by.y =</span> <span class="st">"row.names"</span>) <span class="co">#add ASVs to count data</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>NE_taxa_18S <span class="ot">&lt;-</span> <span class="fu">full_join</span>(NE_18S_taxa, NE_asv_18S_merged, <span class="at">by =</span> <span class="st">"ASV"</span>) <span class="co">#full_join keeps ASVs which didn't get assigned, left_join would remove unassigned</span></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>NE_asv_CO1_merged <span class="ot">&lt;-</span> <span class="fu">merge</span>(NE_asv_fasta_CO1_df, NE_asv_count_CO1, <span class="at">by.x =</span> <span class="st">"Seqs"</span>, <span class="at">by.y =</span> <span class="st">"row.names"</span>) <span class="co">#add ASVs to count data</span></span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>NE_taxa_CO1 <span class="ot">&lt;-</span> <span class="fu">full_join</span>(NE_CO1_taxa, NE_asv_CO1_merged, <span class="at">by =</span> <span class="st">"ASV"</span>) <span class="co">#full_join keeps ASVs which didn't get assigned, left_join would remove unassigned</span></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a><span class="co">#Cornwall----</span></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a>CN_asv_18S_merged <span class="ot">&lt;-</span> <span class="fu">merge</span>(CN_asv_fasta_18S_df, CN_asv_count_18S, <span class="at">by.x =</span> <span class="st">"Seqs"</span>, <span class="at">by.y =</span> <span class="st">"row.names"</span>) <span class="co">#add ASVs to count data</span></span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a>CN_taxa_18S <span class="ot">&lt;-</span> <span class="fu">full_join</span>(CN_18S_taxa, CN_asv_18S_merged, <span class="at">by =</span> <span class="st">"ASV"</span>) <span class="co">#full_join keeps ASVs which didn't get assigned, left_join would remove unassigned</span></span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a>CN_asv_CO1_merged <span class="ot">&lt;-</span> <span class="fu">merge</span>(CN_asv_fasta_CO1_df, CN_asv_count_CO1, <span class="at">by.x =</span> <span class="st">"Seqs"</span>, <span class="at">by.y =</span> <span class="st">"row.names"</span>) <span class="co">#add ASVs to count data</span></span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a>CN_taxa_CO1 <span class="ot">&lt;-</span> <span class="fu">full_join</span>(CN_CO1_taxa, CN_asv_CO1_merged, <span class="at">by =</span> <span class="st">"ASV"</span>) <span class="co">#full_join keeps ASVs which didn't get assigned, left_join would remove unassigned</span></span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a><span class="co">#Controls----</span></span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a>Controls_asv_18S_merged <span class="ot">&lt;-</span> <span class="fu">merge</span>(controls_PhD2_asv_fasta_18S_df, controls_PhD2_asv_count_18S, <span class="at">by.x =</span> <span class="st">"Seqs"</span>, <span class="at">by.y =</span> <span class="st">"row.names"</span>) <span class="co">#add ASVs to count data</span></span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a>Controls_taxa_18S <span class="ot">&lt;-</span> <span class="fu">full_join</span>(Controls_18S_taxa, Controls_asv_18S_merged, <span class="at">by =</span> <span class="st">"ASV"</span>) <span class="co">#full_join keeps ASVs which didn't get assigned, left_join would remove unassigned</span></span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a>Controls_asv_CO1_merged <span class="ot">&lt;-</span> <span class="fu">merge</span>(controls_PhD2_asv_fasta_CO1_df, controls_PhD2_asv_count_CO1, <span class="at">by.x =</span> <span class="st">"Seqs"</span>, <span class="at">by.y =</span> <span class="st">"row.names"</span>) <span class="co">#add ASVs to count data</span></span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a>Controls_taxa_CO1 <span class="ot">&lt;-</span> <span class="fu">full_join</span>(Controls_CO1_taxa, Controls_asv_CO1_merged, <span class="at">by =</span> <span class="st">"ASV"</span>) <span class="co">#full_join keeps ASVs which didn't get assigned, left_join would remove unassigned</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="remove-undetermined-sequences-in-18s-data" class="level3">
<h3 class="anchored" data-anchor-id="remove-undetermined-sequences-in-18s-data">Remove undetermined sequences in 18S data</h3>
<p>Sequences that were unable to be assigned to a specific sample in the 18S are stored under the variable ‘S0’. We want to remove those sequences from our data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>drop <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"S0"</span>) <span class="co"># set which variable we want to remove (undetermined in MiniSeq run)</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>SCH_taxa_18S <span class="ot">&lt;-</span> SCH_taxa_18S[,<span class="sc">!</span>(<span class="fu">names</span>(SCH_taxa_18S) <span class="sc">%in%</span> drop)]</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>NW_taxa_18S <span class="ot">&lt;-</span> NW_taxa_18S[,<span class="sc">!</span>(<span class="fu">names</span>(NW_taxa_18S) <span class="sc">%in%</span> drop)]</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>SW_taxa_18S <span class="ot">&lt;-</span> SW_taxa_18S[,<span class="sc">!</span>(<span class="fu">names</span>(SW_taxa_18S) <span class="sc">%in%</span> drop)] </span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>NE_taxa_18S <span class="ot">&lt;-</span> NE_taxa_18S[,<span class="sc">!</span>(<span class="fu">names</span>(NE_taxa_18S) <span class="sc">%in%</span> drop)]</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>CN_taxa_18S <span class="ot">&lt;-</span> CN_taxa_18S[,<span class="sc">!</span>(<span class="fu">names</span>(CN_taxa_18S) <span class="sc">%in%</span> drop)]</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>Controls_taxa_18S <span class="ot">&lt;-</span> Controls_taxa_18S[,<span class="sc">!</span>(<span class="fu">names</span>(Controls_taxa_18S) <span class="sc">%in%</span> drop)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="convert-to-long-format" class="level3">
<h3 class="anchored" data-anchor-id="convert-to-long-format">Convert to long format</h3>
<p>Here, we flip the data sets into a long format and add a region variable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Scotland----</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>flip_SCH_18S <span class="ot">&lt;-</span> <span class="fu">gather</span>(SCH_taxa_18S, sample, reads, S66<span class="sc">:</span>S132) </span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>flip_SCH_18S<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="st">"Scotland"</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>flip_SCH_CO1 <span class="ot">&lt;-</span> <span class="fu">gather</span>(SCH_taxa_CO1, sample, reads, Adapter3616.SCH01<span class="fl">.04</span>rep<span class="sc">:</span>Adapter453.SCHNC2803) <span class="co">#this is not number order, this is first and last column</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>flip_SCH_CO1<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="st">"Scotland"</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="co">#North Wales----</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>flip_NW_18S <span class="ot">&lt;-</span> <span class="fu">gather</span>(NW_taxa_18S, sample, reads, S1<span class="sc">:</span>S51) </span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>flip_NW_18S<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="st">"North Wales"</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>flip_NW_CO1 <span class="ot">&lt;-</span> <span class="fu">gather</span>(NW_taxa_CO1, sample, reads, X1.NW01<span class="fl">.01</span><span class="sc">:</span>X9.NW02<span class="fl">.06</span>) <span class="co">#this is not number order, this is first and last column</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>flip_NW_CO1<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="st">"North Wales"</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a><span class="co">#South Wales----</span></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>flip_SW_18S <span class="ot">&lt;-</span> <span class="fu">gather</span>(SW_taxa_18S, sample, reads, S60<span class="sc">:</span>S54) </span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>flip_SW_18S<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="st">"South Wales"</span></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>flip_SW_CO1 <span class="ot">&lt;-</span> <span class="fu">gather</span>(SW_taxa_CO1, sample, reads, Adapter3553.SW01<span class="fl">.01</span><span class="sc">:</span>Adapter3620.PCrep) <span class="co">#this is not number order, this is first and last column</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>flip_SW_CO1<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="st">"South Wales"</span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a><span class="co">#Northeast----</span></span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a>flip_NE_18S <span class="ot">&lt;-</span> <span class="fu">gather</span>(NE_taxa_18S, sample, reads, S73<span class="sc">:</span>S123) </span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a>flip_NE_18S<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="st">"Northeast England"</span></span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a>flip_NE_CO1 <span class="ot">&lt;-</span> <span class="fu">gather</span>(NE_taxa_CO1, sample, reads, X100.NE05<span class="fl">.02</span><span class="sc">:</span>X99.NE05<span class="fl">.01</span>) <span class="co">#this is not number order, this is first and last column</span></span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a>flip_NE_CO1<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="st">"Northeast England"</span></span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true" tabindex="-1"></a><span class="co">#Cornwall----</span></span>
<span id="cb17-30"><a href="#cb17-30" aria-hidden="true" tabindex="-1"></a>flip_CN_18S <span class="ot">&lt;-</span> <span class="fu">gather</span>(CN_taxa_18S, sample, reads, S133<span class="sc">:</span>S183) </span>
<span id="cb17-31"><a href="#cb17-31" aria-hidden="true" tabindex="-1"></a>flip_CN_18S<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="st">"Cornwall"</span></span>
<span id="cb17-32"><a href="#cb17-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-33"><a href="#cb17-33" aria-hidden="true" tabindex="-1"></a>flip_CN_CO1 <span class="ot">&lt;-</span> <span class="fu">gather</span>(CN_taxa_CO1, sample, reads, X115.CN01<span class="fl">.01</span><span class="sc">:</span>X192.CN01<span class="fl">.02</span>rep) <span class="co">#this is not number order, this is first and last column</span></span>
<span id="cb17-34"><a href="#cb17-34" aria-hidden="true" tabindex="-1"></a>flip_CN_CO1<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="st">"Cornwall"</span></span>
<span id="cb17-35"><a href="#cb17-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-36"><a href="#cb17-36" aria-hidden="true" tabindex="-1"></a><span class="co">#Controls----</span></span>
<span id="cb17-37"><a href="#cb17-37" aria-hidden="true" tabindex="-1"></a>flip_controls_18S <span class="ot">&lt;-</span> <span class="fu">gather</span>(Controls_taxa_18S, sample, reads, S34<span class="sc">:</span>S61) <span class="co">#this is not number order, this is first and last column</span></span>
<span id="cb17-38"><a href="#cb17-38" aria-hidden="true" tabindex="-1"></a>flip_controls_18S<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="st">"Varied"</span></span>
<span id="cb17-39"><a href="#cb17-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-40"><a href="#cb17-40" aria-hidden="true" tabindex="-1"></a>flip_controls_CO1 <span class="ot">&lt;-</span> <span class="fu">gather</span>(Controls_taxa_CO1, sample, reads, X173.SCH02<span class="fl">.03</span><span class="sc">:</span>X195.minusVEpcr) <span class="co">#this is not number order, this is first and last column</span></span>
<span id="cb17-41"><a href="#cb17-41" aria-hidden="true" tabindex="-1"></a>flip_controls_CO1<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="st">"Varied"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s take a look at the long format for Scotland CO1. We now have a long format data set with ASVs, taxonomy, sequences, sample names, number of reads and region variables.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(flip_SCH_CO1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        ASV                taxonomy
1   ASV_820 Paramoeba pemaquidensis
2 ASV_18576      Vexillifera kereti
3 ASV_22679      Vexillifera kereti
4   ASV_280       Parvamoeba rugata
5 ASV_10580       Parvamoeba rugata
6 ASV_20723       Parvamoeba rugata
                                                                                                                                                                                                                                                                                                                       Seqs
1 ATTATCTAGTGTAGTTGCACATTCAACTCCTTCCGTAGATTTAGCTATATTTAGTTTACATTTAGCAGGTATAGCATCAATTGCAGGATCTATTAATTATATTACAACTATTTTTAATATGAGAGTTGCTAAATATGATATGTTTAGATTACCTTTATTTATTTGAGGTATGCTTATTACTGCTTATTTATTAGTTTTTACTTTACCTGTTTTAGCTGGTGCTATCACTATGCTATTAACTGATAGAAATTTTAATACTTCTTTTTTTGATCCCGCTGGAGGTGGAGATCCTATACTATATCAACATTTATTT
2 ATTATCTAGTATAATTGCGCATTCTGGTGGTTCAGTAGATTTAGCTATTTTTAGTTTACATTTAGCAGGTATTTCGTCATTATTAGGTGCTATTAATTTTATTACAACTATTTTTAATATGAGAGCAAATAATTTTAGCATATATAAAATGCCTTTATTTGTTTGAGCAGTATTAATTACAGCTTTTTTATTATTATTATCATTACCTGTATTAGCAGGAGCTATTACTATGCTTTTAACTGATAGAAATTTTAATACTACTTTTTTTGACCCAGCAGGAGGTGGTGATCCTATATTGTATCAACATTTATTT
3 ATTATCTAGTATAATTGCGCATTCTGGTGGTTCAGTAGATTTAGCTATTTTTAGTTTACATTTAGCAGGTATTTCGTCATTATTAGGTGCTATTAATTTTATTACAACTATTTTTAATATGAGAGCAAATAATTTTAGTATATATAAAATGCCTTTATTTGTTTGAGCAGTATTAATTACAGCTTTTTTATTATTATTATCATTACCTGTATTAGCTGGAGCTATTACTATGCTTTTAACTGATAGAAATTTTAATACTACTTTTTTTGACCCAGCAGGAGGAGGTGATCCTATATTGTATCAACATTTATTT
4 TTTATCGGGTATAGAAGCGCATTCTGGTGGATCTGTAGATTTAGCAATTTTTAGTTTACATTTAGCAGGAGCATCTTCTTTATTAGGAGCAATAAATTTTATTACTACTGTCTTTAATATGAGAGCACCTAATTTAGAAATGCATAAATTACCTTTATTTGTATGAGCTATTTTAATTACAGCTTTTTTATTATTATTATCTTTACCTGTTTTAGCTGGTGCAATAACTATGCTTTTAACTGATCGTAATTTTAATACTACTTTTTTTGATCCAGCTGGTGGTGGTGATCCAATATTGTATCAGCATTTATTT
5 TTTATCTGGTATAGAAGCACATTCTGGTGGATCTGTAGATTTAGCAATTTTTAGTTTACATTTAGCAGGAGCATCTTCTTTATTAGGAGCAATAAATTTTATTACTACTGTTTTTAATATGAGAGCACCTAATTTAGAAATGCATAAATTACCTTTATTTGTTTGAGCTATTTTAATCACAGCTTTTTTATTATTATTATCTTTACCTGTTTTAGCTGGTGCAATAACTATGCTTTTAACTGATCGTAATTTTAATACTACTTTTTTTGATCCAGCTGGTGGTGGTGATCCAATATTGTATCAGCATTTGTTT
6 TTTATCGGGTATAGAAGCGCATTCTGGTGGATCTGTAGATTTAGCAATTTTTAGTTTACATTTAGCAGGAGCATCTTCTTTATTAGGAGCAATAAATTTTATTACTACTGTCTTTAATATGAGAGCACCTAATTTAGAAATGCATAAATTACCTTTATTTGTATGAGCTATTTTAATTACAGCTTTTTTATTATTATTATCTTTACCTGTTTTAGCAGGTGCAATAACTATGCTTTTAACTGATCGTAATTTTAATACTACTTTTTTTGATCCAGCTGGTGGTGGTGATCCAATATTGTATCAGCATTTATTT
                   sample reads   region
1 Adapter3616.SCH01.04rep    13 Scotland
2 Adapter3616.SCH01.04rep     2 Scotland
3 Adapter3616.SCH01.04rep     0 Scotland
4 Adapter3616.SCH01.04rep     0 Scotland
5 Adapter3616.SCH01.04rep    14 Scotland
6 Adapter3616.SCH01.04rep     0 Scotland</code></pre>
</div>
</div>
</section>
<section id="join-data-frames" class="level3">
<h3 class="anchored" data-anchor-id="join-data-frames">Join data frames</h3>
<p>We will merge our data in multiple stages: first by sequencing run, then by gene, and finally, we will perform the final join. Once we have our single data frame, we’ll add in our metadata.</p>
<p>Our data was derived from multiple sequences runs. Therefore, we are going to join the data in the same sequencing run first. This means we can correctly match IDs we want to sample names.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge stage-one locations (SCH and SW) and add IDs</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>full_stageone_18S <span class="ot">&lt;-</span><span class="fu">rbind</span>(flip_SCH_18S, flip_SW_18S) <span class="sc">%&gt;%</span> </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">sample_18S =</span> sample) <span class="sc">%&gt;%</span> <span class="co">#rename sample column to match</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">full_join</span>(IDs_stageone, <span class="at">by =</span> <span class="st">"sample_18S"</span>)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>full_stageone_18S<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(full_stageone_18S<span class="sc">$</span>region)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>full_stageone_CO1 <span class="ot">&lt;-</span><span class="fu">rbind</span>(flip_SCH_CO1, flip_SW_CO1) <span class="sc">%&gt;%</span> </span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">sample_CO1 =</span> sample) <span class="sc">%&gt;%</span> <span class="co">#rename sample column to match</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">full_join</span>(IDs_stageone, <span class="at">by =</span> <span class="st">"sample_CO1"</span>)</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>full_stageone_CO1<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(full_stageone_CO1<span class="sc">$</span>region)</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge stage-two locations (controls, CN, NE, NW)</span></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>full_stagetwo_18S <span class="ot">&lt;-</span><span class="fu">rbind</span>(flip_controls_18S, flip_CN_18S, flip_NE_18S, flip_NW_18S) <span class="sc">%&gt;%</span> </span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">sample_18S =</span> sample) <span class="sc">%&gt;%</span> </span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(IDs_stagetwo, <span class="at">by =</span> <span class="st">"sample_18S"</span>)</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>full_stagetwo_18S<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(full_stagetwo_18S<span class="sc">$</span>region)</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>full_stagetwo_CO1 <span class="ot">&lt;-</span><span class="fu">rbind</span>(flip_controls_CO1, flip_CN_CO1, flip_NE_CO1, flip_NW_CO1) <span class="sc">%&gt;%</span></span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">sample_CO1 =</span> sample) <span class="sc">%&gt;%</span> </span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">full_join</span>(IDs_stagetwo, <span class="at">by =</span> <span class="st">"sample_CO1"</span>)</span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a>full_stagetwo_CO1<span class="sc">$</span>region <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(full_stagetwo_CO1<span class="sc">$</span>region)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we join data sets by primer and add primer variable. This will leave us with two main data sets, one for 18S and one for CO1. We also want to make sure out wrangling worked as expected by checking there are no duplicate rows.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>full_18S <span class="ot">&lt;-</span> <span class="fu">rbind</span>(full_stageone_18S, full_stagetwo_18S) <span class="co">#18s</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>full_18S<span class="sc">$</span>primer <span class="ot">&lt;-</span> <span class="st">"18S"</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">duplicated</span>(full_18S)) <span class="co">#check for duplicates - should be false</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>full_CO1 <span class="ot">&lt;-</span> <span class="fu">rbind</span>(full_stageone_CO1, full_stagetwo_CO1) <span class="co">#CO1</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>full_CO1<span class="sc">$</span>primer <span class="ot">&lt;-</span> <span class="st">"CO1"</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">duplicated</span>(full_CO1)) <span class="co">#check for duplicates - should be false</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
</div>
<p>Finally, we join the two primer data sets into a single data set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>eDNA_ASVs_long <span class="ot">&lt;-</span> <span class="fu">rbind</span>(full_18S, full_CO1)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">duplicated</span>(eDNA_ASVs_long)) <span class="co">#check for duplicates - should be false (this may take a minute or so)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
</div>
<p>Let’s now add in our metadata and do some final tweaks to sample names.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>eDNA_long_data <span class="ot">&lt;-</span> eDNA_ASVs_long <span class="sc">%&gt;%</span> </span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset</span>(<span class="at">select =</span> <span class="sc">-</span><span class="fu">c</span>(sample_18S, sample_CO1)) <span class="sc">%&gt;%</span> </span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">fieldID =</span> sample_ID)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>eDNA_long_data <span class="ot">&lt;-</span> <span class="fu">full_join</span>(eDNA_long_data, sites, <span class="at">by =</span> <span class="st">"fieldID"</span>)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>eDNA_long_data<span class="sc">$</span>fullID <span class="ot">=</span> <span class="fu">paste</span>(eDNA_long_data<span class="sc">$</span>fieldID, eDNA_long_data<span class="sc">$</span>primer, <span class="at">sep=</span><span class="st">"_"</span>) <span class="co">#adds CO1 or 18S onto the end of the fieldID for further analyses</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Great - we now have a single long data set with all ASV across samples and genes, as well as associated metadata.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(eDNA_long_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   7012768 obs. of  70 variables:
 $ ASV                  : chr  "ASV_15" "ASV_16" "ASV_25" "ASV_32" ...
 $ taxonomy             : chr  "Eukaryota" "Eukaryota" "Eukaryota" "Eukaryota" ...
 $ Seqs                 : chr  "ATAACGAACGAGACCTTAACCTGCTAAATAGTCAGGCCGGCTTTGGCTGGTCGTCGACTTCTTAGAGGGACTATTGGCGTTTAGCCAATGGAAGTTTGAG" "TTAACGAACGAGACCTCAGCCTGCTAAATAGTTGGACCCTACTCTTAGGGCCACAACTTCTTAGAGGGACTATGTGCGTGTAGCACGTGGAAGTTTGAG" "TTAACGAACGAGACCTCAGCCTGCTAAATAGTTGTACGCTACTCTTAGCGCAGCAACTTCTTAGAGGGACTATTGGCGTTTAGCCAATGGAAGTTTGAG" "TTAACGAACGAGACCTTAACCTGCTAAATAGTTACACTTAACTTCGGTTATGTGGGCAACTTCTTAGAGGGACTTTGTGTGTTTAACGCAAGGAAGTTTGAG" ...
 $ reads                : int  12 69 1743 138 843 0 192 802 0 33 ...
 $ region               : Factor w/ 6 levels "Scotland","South Wales",..: 1 1 1 1 1 1 1 1 1 1 ...
 $ fieldID              : chr  "PHD1-SCH01-01" "PHD1-SCH01-01" "PHD1-SCH01-01" "PHD1-SCH01-01" ...
 $ primer               : chr  "18S" "18S" "18S" "18S" ...
 $ eventID              : chr  "PHD1-2022.06.12-SCH01-01" "PHD1-2022.06.12-SCH01-01" "PHD1-2022.06.12-SCH01-01" "PHD1-2022.06.12-SCH01-01" ...
 $ projectID            : chr  "PHD1" "PHD1" "PHD1" "PHD1" ...
 $ localityID           : chr  "SCH01" "SCH01" "SCH01" "SCH01" ...
 $ sampleID             : chr  "1" "1" "1" "1" ...
 $ sampleType           : chr  "sample" "sample" "sample" "sample" ...
 $ controlCheck         : chr  "sample" "sample" "sample" "sample" ...
 $ negCheckLogical      : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
 $ pairedRepLogical     : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
 $ batch_extraction     : int  1 1 1 1 1 1 1 1 1 1 ...
 $ batch_PCR_CO1        : int  2 2 2 2 2 2 2 2 2 2 ...
 $ plate_col_CO1        : int  1 1 1 1 1 1 1 1 1 1 ...
 $ plate_row_CO1        : chr  "A" "A" "A" "A" ...
 $ batch_PCR_18S        : int  4 4 4 4 4 4 4 4 4 4 ...
 $ plate_col_18S        : int  1 1 1 1 1 1 1 1 1 1 ...
 $ plate_row_18S        : chr  "A" "A" "A" "A" ...
 $ date_extraction      : chr  "24.02.2023" "24.02.2023" "24.02.2023" "24.02.2023" ...
 $ date_amplified_CO1   : chr  "28.03.2023" "28.03.2023" "28.03.2023" "28.03.2023" ...
 $ date_amplified_18S   : chr  "27.03.2023" "27.03.2023" "27.03.2023" "27.03.2023" ...
 $ sample_18S           : chr  "S66" "S66" "S66" "S66" ...
 $ sample_CO1           : chr  "Adapter385.SCH01.01" "Adapter385.SCH01.01" "Adapter385.SCH01.01" "Adapter385.SCH01.01" ...
 $ COUNT                : int  2 2 2 2 2 2 2 2 2 2 ...
 $ dna_concentration_CO1: num  82.5 82.5 82.5 82.5 82.5 ...
 $ dna_concentration_18S: num  55.4 55.4 55.4 55.4 55.4 ...
 $ continent            : chr  "Europe" "Europe" "Europe" "Europe" ...
 $ country              : chr  "Scotland" "Scotland" "Scotland" "Scotland" ...
 $ county               : chr  "Sutherland" "Sutherland" "Sutherland" "Sutherland" ...
 $ waterBody            : chr  "rocky" "rocky" "rocky" "rocky" ...
 $ verbatimLocality     : chr  "Scourie" "Scourie" "Scourie" "Scourie" ...
 $ exposure             : chr  "Exposed" "Exposed" "Exposed" "Exposed" ...
 $ weather              : chr  "moderate rain, strong winds, very cloudy" "moderate rain, strong winds, very cloudy" "moderate rain, strong winds, very cloudy" "moderate rain, strong winds, very cloudy" ...
 $ tide                 : chr  "outgoing" "outgoing" "outgoing" "outgoing" ...
 $ lowWater             : chr  "12:23" "12:23" "12:23" "12:23" ...
 $ decimalLongitude     : num  -5.17 -5.17 -5.17 -5.17 -5.17 ...
 $ decimalLatitude      : num  58.4 58.4 58.4 58.4 58.4 ...
 $ recordedBy           : chr  "Dina-Leigh Simons; Nova Mieszkowska" "Dina-Leigh Simons; Nova Mieszkowska" "Dina-Leigh Simons; Nova Mieszkowska" "Dina-Leigh Simons; Nova Mieszkowska" ...
 $ year                 : int  2022 2022 2022 2022 2022 2022 2022 2022 2022 2022 ...
 $ month                : int  6 6 6 6 6 6 6 6 6 6 ...
 $ day                  : int  12 12 12 12 12 12 12 12 12 12 ...
 $ eventTime            : chr  "09:31" "09:31" "09:31" "09:31" ...
 $ samplingProtocol     : chr  "Sterivex-0.22" "Sterivex-0.22" "Sterivex-0.22" "Sterivex-0.22" ...
 $ pumpType             : chr  "Syringe 60mL, sealant gun" "Syringe 60mL, sealant gun" "Syringe 60mL, sealant gun" "Syringe 60mL, sealant gun" ...
 $ repVol..mL.          : int  1000 1000 1000 1000 1000 1000 1000 1000 1000 1000 ...
 $ algaeCover.          : int  0 0 0 0 0 0 0 0 0 0 ...
 $ remarks              : chr  "Filtered on shore" "Filtered on shore" "Filtered on shore" "Filtered on shore" ...
 $ daysBeforefreeze     : int  9 9 9 9 9 9 9 9 9 9 ...
 $ preservation         : chr  "EtOH" "EtOH" "EtOH" "EtOH" ...
 $ type                 : chr  "Rockpool" "Rockpool" "Rockpool" "Rockpool" ...
 $ shorePosition        : chr  "High" "High" "High" "High" ...
 $ rockpoolarea.cm.2.   : num  7443 7443 7443 7443 7443 ...
 $ depth1               : num  NA NA NA NA NA NA NA NA NA NA ...
 $ depth2               : num  NA NA NA NA NA NA NA NA NA NA ...
 $ averageDepth..cm.    : num  25 25 25 25 25 25 25 25 25 25 ...
 $ rockpoolVol.m.3.     : num  0.186 0.186 0.186 0.186 0.186 ...
 $ pH1                  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ pH2                  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ averagePH            : num  8.59 8.59 8.59 8.59 8.59 8.59 8.59 8.59 8.59 8.59 ...
 $ temp1                : chr  NA NA NA NA ...
 $ temp2                : chr  NA NA NA NA ...
 $ averageTemp          : num  13 13 13 13 13 13 13 13 13 13 ...
 $ siteImage            : chr  "IMG_SCH01-01.jpeg" "IMG_SCH01-01.jpeg" "IMG_SCH01-01.jpeg" "IMG_SCH01-01.jpeg" ...
 $ rockType             : chr  "metamorphic" "metamorphic" "metamorphic" "metamorphic" ...
 $ rockFold             : chr  NA NA NA NA ...
 $ fullID               : chr  "PHD1-SCH01-01_18S" "PHD1-SCH01-01_18S" "PHD1-SCH01-01_18S" "PHD1-SCH01-01_18S" ...</code></pre>
</div>
</div>
</section>
</section>
<section id="cleaning-taxonomic-names" class="level2">
<h2 class="anchored" data-anchor-id="cleaning-taxonomic-names">Cleaning taxonomic names</h2>
<p>Now we have our data, we need to tidy up our taxonomic names and get more information on the taxa we’ve detected.</p>
<section id="remove-ambiguous-names" class="level3">
<h3 class="anchored" data-anchor-id="remove-ambiguous-names">Remove ambiguous names</h3>
<p>We first remove any occurrences where the taxonomic matches certain ambiguous phrases. These assignment tell us nothing about the organism we’ve detected, so it’s best to remove them.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>unassigned_phrases <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"uncultured"</span>, <span class="st">"Uncultured"</span>, <span class="st">"unclassified"</span>, <span class="st">"Unclassified"</span>, <span class="st">"NCBI"</span>, </span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>             <span class="st">"pseudo"</span>, <span class="st">"Pseudo"</span>, <span class="st">"Not assigned"</span>, <span class="st">"not assigned"</span>, <span class="st">"SAR"</span>, <span class="st">"sar"</span>, </span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>             <span class="st">"cellular"</span>, <span class="st">"Cellular"</span>, <span class="st">"environmental"</span>, <span class="st">"Environemental"</span>, </span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>             <span class="st">"eukaryote"</span>, <span class="st">"Eukaryote"</span>, <span class="st">"eukaryota"</span>, <span class="st">"Eukaryota"</span>)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>eDNA_long_data_cleaning <span class="ot">&lt;-</span> eDNA_long_data[<span class="sc">!</span><span class="fu">grepl</span>(<span class="fu">paste</span>(unassigned_phrases, <span class="at">collapse =</span> <span class="st">"|"</span>), eDNA_long_data<span class="sc">$</span>taxonomy),] <span class="sc">%&gt;%</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>(taxonomy)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="using-clean_names-in-janitor-package" class="level3">
<h3 class="anchored" data-anchor-id="using-clean_names-in-janitor-package">Using clean_names() in janitor package</h3>
<p>We often get random spaces or phrases we don’t want in our taxonomic names. We can use the <a href="https://cran.r-project.org/web/packages/janitor/vignettes/janitor.html">janitor package</a> to clean our taxonomic names.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> <span class="fu">unique</span>(eDNA_long_data_cleaning<span class="sc">$</span>taxonomy) <span class="co">#get all unique names from the data</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>cleaned_names <span class="ot">&lt;-</span> <span class="fu">clean_names</span>(names) <span class="co">#clean names using janitor package in tidyverse</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>names_together <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(names, cleaned_names) <span class="sc">%&gt;%</span> </span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">taxonomy =</span> names, <span class="at">taxonomy_clean =</span> cleaned_names) <span class="sc">%&gt;%</span> <span class="co">#join cleaned names next to original</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">taxonomy_clean =</span> <span class="fu">gsub</span>(<span class="st">' cf'</span>,<span class="st">''</span>, taxonomy_clean)) <span class="sc">%&gt;%</span> <span class="co">#remove cfs which function didn't remove</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">taxonomy_clean =</span> <span class="fu">gsub</span>(<span class="st">'cf '</span>,<span class="st">''</span>, taxonomy_clean)) <span class="co">#remove cfs which function didn't remove</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>eDNA_long_data_cleaned <span class="ot">&lt;-</span> <span class="fu">full_join</span>(names_together, eDNA_long_data_cleaning)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We should now have a variable added to our data set called taxonomy_clean. We will use this variable as the taxonomic name moving forward.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(eDNA_long_data_cleaned)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   841939 obs. of  71 variables:
 $ taxonomy             : chr  "Amoebozoa" "Amoebozoa" "Amoebozoa" "Amoebozoa" ...
 $ taxonomy_clean       : chr  "amoebozoa" "amoebozoa" "amoebozoa" "amoebozoa" ...
 $ ASV                  : chr  "ASV_3057" "ASV_3512" "ASV_3942" "ASV_4258" ...
 $ Seqs                 : chr  "TTAACGAACGAGACCTTAACCTATTAAATAGTTGCGCGAACTTGAATCATGCGTCATGTTATCTTCGGATAATGTTTCCTCCATGGTTCTATCGTTCTGTGAAAACTTCTT"| __truncated__ "ATAACGAACGAGACCTCAGCCTGCTAACTAGTATGCCTTAGCCTTGTTTTCTTGGCTAAGGAATTTATTGAATATACTTCTTAGAGGGACTATCCGTGTCTAGCGGGTGGAAGTTTGAG" "TTAACGAACGAGACCTTAACCTATTAAATAGTTGCGCGAACTTGAATCATGCGTCAGTTCGTCTTTTGGGCGGACTGTCCTCCATGGTTCTATCGTTCTGTGAAAACTTCT"| __truncated__ "TTAACGAACGAGACCTTAACCTATTAAATAGTTGCGCGAACTTGAATCATGCGTCAGTTCGTCCTTCGTGGGCGAGCTGTCCTCCATGGTTCTATCGTTCTGTGAAAACTT"| __truncated__ ...
 $ reads                : int  0 0 0 0 0 0 0 0 0 0 ...
 $ region               : Factor w/ 6 levels "Scotland","South Wales",..: 1 1 1 1 1 1 1 1 1 1 ...
 $ fieldID              : chr  "PHD1-SCH01-01" "PHD1-SCH01-01" "PHD1-SCH01-01" "PHD1-SCH01-01" ...
 $ primer               : chr  "18S" "18S" "18S" "18S" ...
 $ eventID              : chr  "PHD1-2022.06.12-SCH01-01" "PHD1-2022.06.12-SCH01-01" "PHD1-2022.06.12-SCH01-01" "PHD1-2022.06.12-SCH01-01" ...
 $ projectID            : chr  "PHD1" "PHD1" "PHD1" "PHD1" ...
 $ localityID           : chr  "SCH01" "SCH01" "SCH01" "SCH01" ...
 $ sampleID             : chr  "1" "1" "1" "1" ...
 $ sampleType           : chr  "sample" "sample" "sample" "sample" ...
 $ controlCheck         : chr  "sample" "sample" "sample" "sample" ...
 $ negCheckLogical      : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
 $ pairedRepLogical     : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
 $ batch_extraction     : int  1 1 1 1 1 1 1 1 1 1 ...
 $ batch_PCR_CO1        : int  2 2 2 2 2 2 2 2 2 2 ...
 $ plate_col_CO1        : int  1 1 1 1 1 1 1 1 1 1 ...
 $ plate_row_CO1        : chr  "A" "A" "A" "A" ...
 $ batch_PCR_18S        : int  4 4 4 4 4 4 4 4 4 4 ...
 $ plate_col_18S        : int  1 1 1 1 1 1 1 1 1 1 ...
 $ plate_row_18S        : chr  "A" "A" "A" "A" ...
 $ date_extraction      : chr  "24.02.2023" "24.02.2023" "24.02.2023" "24.02.2023" ...
 $ date_amplified_CO1   : chr  "28.03.2023" "28.03.2023" "28.03.2023" "28.03.2023" ...
 $ date_amplified_18S   : chr  "27.03.2023" "27.03.2023" "27.03.2023" "27.03.2023" ...
 $ sample_18S           : chr  "S66" "S66" "S66" "S66" ...
 $ sample_CO1           : chr  "Adapter385.SCH01.01" "Adapter385.SCH01.01" "Adapter385.SCH01.01" "Adapter385.SCH01.01" ...
 $ COUNT                : int  2 2 2 2 2 2 2 2 2 2 ...
 $ dna_concentration_CO1: num  82.5 82.5 82.5 82.5 82.5 ...
 $ dna_concentration_18S: num  55.4 55.4 55.4 55.4 55.4 ...
 $ continent            : chr  "Europe" "Europe" "Europe" "Europe" ...
 $ country              : chr  "Scotland" "Scotland" "Scotland" "Scotland" ...
 $ county               : chr  "Sutherland" "Sutherland" "Sutherland" "Sutherland" ...
 $ waterBody            : chr  "rocky" "rocky" "rocky" "rocky" ...
 $ verbatimLocality     : chr  "Scourie" "Scourie" "Scourie" "Scourie" ...
 $ exposure             : chr  "Exposed" "Exposed" "Exposed" "Exposed" ...
 $ weather              : chr  "moderate rain, strong winds, very cloudy" "moderate rain, strong winds, very cloudy" "moderate rain, strong winds, very cloudy" "moderate rain, strong winds, very cloudy" ...
 $ tide                 : chr  "outgoing" "outgoing" "outgoing" "outgoing" ...
 $ lowWater             : chr  "12:23" "12:23" "12:23" "12:23" ...
 $ decimalLongitude     : num  -5.17 -5.17 -5.17 -5.17 -5.17 ...
 $ decimalLatitude      : num  58.4 58.4 58.4 58.4 58.4 ...
 $ recordedBy           : chr  "Dina-Leigh Simons; Nova Mieszkowska" "Dina-Leigh Simons; Nova Mieszkowska" "Dina-Leigh Simons; Nova Mieszkowska" "Dina-Leigh Simons; Nova Mieszkowska" ...
 $ year                 : int  2022 2022 2022 2022 2022 2022 2022 2022 2022 2022 ...
 $ month                : int  6 6 6 6 6 6 6 6 6 6 ...
 $ day                  : int  12 12 12 12 12 12 12 12 12 12 ...
 $ eventTime            : chr  "09:31" "09:31" "09:31" "09:31" ...
 $ samplingProtocol     : chr  "Sterivex-0.22" "Sterivex-0.22" "Sterivex-0.22" "Sterivex-0.22" ...
 $ pumpType             : chr  "Syringe 60mL, sealant gun" "Syringe 60mL, sealant gun" "Syringe 60mL, sealant gun" "Syringe 60mL, sealant gun" ...
 $ repVol..mL.          : int  1000 1000 1000 1000 1000 1000 1000 1000 1000 1000 ...
 $ algaeCover.          : int  0 0 0 0 0 0 0 0 0 0 ...
 $ remarks              : chr  "Filtered on shore" "Filtered on shore" "Filtered on shore" "Filtered on shore" ...
 $ daysBeforefreeze     : int  9 9 9 9 9 9 9 9 9 9 ...
 $ preservation         : chr  "EtOH" "EtOH" "EtOH" "EtOH" ...
 $ type                 : chr  "Rockpool" "Rockpool" "Rockpool" "Rockpool" ...
 $ shorePosition        : chr  "High" "High" "High" "High" ...
 $ rockpoolarea.cm.2.   : num  7443 7443 7443 7443 7443 ...
 $ depth1               : num  NA NA NA NA NA NA NA NA NA NA ...
 $ depth2               : num  NA NA NA NA NA NA NA NA NA NA ...
 $ averageDepth..cm.    : num  25 25 25 25 25 25 25 25 25 25 ...
 $ rockpoolVol.m.3.     : num  0.186 0.186 0.186 0.186 0.186 ...
 $ pH1                  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ pH2                  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ averagePH            : num  8.59 8.59 8.59 8.59 8.59 8.59 8.59 8.59 8.59 8.59 ...
 $ temp1                : chr  NA NA NA NA ...
 $ temp2                : chr  NA NA NA NA ...
 $ averageTemp          : num  13 13 13 13 13 13 13 13 13 13 ...
 $ siteImage            : chr  "IMG_SCH01-01.jpeg" "IMG_SCH01-01.jpeg" "IMG_SCH01-01.jpeg" "IMG_SCH01-01.jpeg" ...
 $ rockType             : chr  "metamorphic" "metamorphic" "metamorphic" "metamorphic" ...
 $ rockFold             : chr  NA NA NA NA ...
 $ fullID               : chr  "PHD1-SCH01-01_18S" "PHD1-SCH01-01_18S" "PHD1-SCH01-01_18S" "PHD1-SCH01-01_18S" ...</code></pre>
</div>
</div>
<p>We can see the data set is a lot smaller after removal of rows with ambiguous names.</p>
</section>
</section>
<section id="extracting-information-from-worms" class="level2">
<h2 class="anchored" data-anchor-id="extracting-information-from-worms">Extracting information from WoRMS</h2>
<p>The <a href="https://www.marinespecies.org/">World Register of Marine Species</a> (WoRMS) is an authoritative classification and catalogue of marine names. It can provide a range of information. We will be extracting information on taxonomy and functional trait information using the taxize and worrms functions.</p>
<section id="defining-functions" class="level3">
<h3 class="anchored" data-anchor-id="defining-functions">Defining functions</h3>
<p>First, we’ve got to set-up some functions.</p>
<p>The first function called <code>get_wormsid_noerror</code> extracts unique identifiers for all taxon used in WoRMS called AphiaIDs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>get_wormsid_noerror <span class="ot">&lt;-</span> <span class="cf">function</span>(sp_name){</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  wm_id <span class="ot">&lt;-</span> <span class="fu">try</span>(taxize<span class="sc">::</span><span class="fu">get_wormsid</span>(sp_name,</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">accepted =</span> <span class="cn">TRUE</span>, </span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">searchtype=</span> <span class="st">"scientific"</span>, </span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">marine_only =</span> <span class="cn">FALSE</span>,</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">ask =</span> <span class="cn">FALSE</span>,</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">row =</span> <span class="dv">1</span>,</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">message =</span> <span class="cn">FALSE</span>),</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>               <span class="at">silent =</span> <span class="cn">TRUE</span>)</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>  <span class="co">#remove end word to attempt matching genus when previous no match</span></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">class</span>(wm_id) <span class="sc">==</span> <span class="st">"try-error"</span>){</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>    sp_name <span class="ot">&lt;-</span> <span class="fu">str_remove</span>(sp_name, <span class="st">"(</span><span class="sc">\\</span><span class="st">s+</span><span class="sc">\\</span><span class="st">w+)"</span>) </span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>    wm_id <span class="ot">&lt;-</span> <span class="fu">try</span>(taxize<span class="sc">::</span><span class="fu">get_wormsid</span>(sp_name,</span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">accepted =</span> <span class="cn">TRUE</span>, </span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">searchtype=</span> <span class="st">"scientific"</span>, </span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">marine_only =</span> <span class="cn">FALSE</span>,</span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">ask =</span> <span class="cn">FALSE</span>,</span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">row =</span> <span class="dv">1</span>,</span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">message =</span> <span class="cn">FALSE</span>),</span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a>                 <span class="at">silent =</span> <span class="cn">TRUE</span>)</span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a>  } </span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>  <span class="co">#convert non-matched species to NA AphiaID</span></span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb34-27"><a href="#cb34-27" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">class</span>(wm_id) <span class="sc">==</span> <span class="st">"try-error"</span>){</span>
<span id="cb34-28"><a href="#cb34-28" aria-hidden="true" tabindex="-1"></a>    wm_id <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb34-29"><a href="#cb34-29" aria-hidden="true" tabindex="-1"></a>  } </span>
<span id="cb34-30"><a href="#cb34-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">sciname =</span> sp_name, <span class="at">aphiaID =</span> <span class="fu">as.double</span>(<span class="fu">unlist</span>(wm_id)))</span>
<span id="cb34-31"><a href="#cb34-31" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The second function called <code>get_wormsmeta</code> accesses and formats any taxonomic meta data based on Aphia IDs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>get_wormsmeta <span class="ot">&lt;-</span> <span class="cf">function</span>(aphia_input){</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="co">#split into smaller chunks for wm_record() to work</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  taxadf_split <span class="ot">&lt;-</span> <span class="fu">split</span>(aphia_input, (<span class="fu">seq</span>(<span class="fu">nrow</span>(aphia_input))<span class="sc">-</span><span class="dv">1</span>) <span class="sc">%/%</span> <span class="dv">50</span>) <span class="co">#split into smaller groups for worms to work</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  temp_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>  <span class="co">#run wm_record() through split list</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> taxadf_split) {</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>    taxa_temp <span class="ot">&lt;-</span> <span class="fu">wm_record</span>(<span class="at">id =</span> i<span class="sc">$</span>aphiaID)</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>    temp_df <span class="ot">=</span> <span class="fu">rbind</span>(temp_df, taxa_temp)</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(temp_df)</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The third function called <code>get_worms_fgrp</code> (originally developed <a href="https://github.com/tomjwebb/WoRMS-functional-groups">here</a>) accesses and formats information on broad taxonomic groupings of taxa based on Aphia IDs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>get_worms_fgrp <span class="ot">&lt;-</span> <span class="cf">function</span>(AphiaID) {</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="co">#' First, test if the species has attribute data</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  attr_dat <span class="ot">&lt;-</span> <span class="fu">try</span>(<span class="fu">wm_attr_data</span>(</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>    AphiaID, <span class="at">include_inherited =</span> <span class="cn">TRUE</span>), <span class="at">silent =</span> <span class="cn">TRUE</span>)</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>  <span class="co">#' set up out as null for later use</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="sc">!</span><span class="fu">identical</span>(<span class="fu">class</span>(attr_dat), <span class="st">"try-error"</span>)){</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>    <span class="co">#' if attribute data exists, test if functional group is there</span></span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="st">"Functional group"</span> <span class="sc">%in%</span> attr_dat<span class="sc">$</span>measurementType){</span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>      fg_dat <span class="ot">&lt;-</span> attr_dat <span class="sc">%&gt;%</span> <span class="fu">filter</span>(measurementType <span class="sc">==</span> <span class="st">"Functional group"</span>)</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a>      <span class="co">#' Insert if statement here about $children empty</span></span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a>      <span class="co">#' assign the $children - so that it can be used </span></span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a>      children <span class="ot">&lt;-</span> fg_dat<span class="sc">$</span>children</span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span>(<span class="fu">max</span>(<span class="fu">lengths</span>(children)) <span class="sc">&gt;</span> <span class="dv">0</span> ){</span>
<span id="cb36-21"><a href="#cb36-21" aria-hidden="true" tabindex="-1"></a>        <span class="co">#' Extract the life stage information from the $children field</span></span>
<span id="cb36-22"><a href="#cb36-22" aria-hidden="true" tabindex="-1"></a>        life_stage <span class="ot">&lt;-</span> children <span class="sc">%&gt;%</span> <span class="fu">bind_rows</span>() <span class="sc">%&gt;%</span></span>
<span id="cb36-23"><a href="#cb36-23" aria-hidden="true" tabindex="-1"></a>          dplyr<span class="sc">::</span><span class="fu">select</span>(measurementValue) <span class="sc">%&gt;%</span></span>
<span id="cb36-24"><a href="#cb36-24" aria-hidden="true" tabindex="-1"></a>          dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">stage =</span> measurementValue)</span>
<span id="cb36-25"><a href="#cb36-25" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb36-26"><a href="#cb36-26" aria-hidden="true" tabindex="-1"></a>        <span class="co">#' add in rows for instances missing children</span></span>
<span id="cb36-27"><a href="#cb36-27" aria-hidden="true" tabindex="-1"></a>        idx <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">lengths</span>(children) <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb36-28"><a href="#cb36-28" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span>(<span class="fu">length</span>(idx) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb36-29"><a href="#cb36-29" aria-hidden="true" tabindex="-1"></a>          life_stage_null <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">stage =</span> <span class="fu">rep</span>(<span class="fu">as.character</span>(<span class="cn">NA</span>), <span class="fu">length</span>(children)))</span>
<span id="cb36-30"><a href="#cb36-30" aria-hidden="true" tabindex="-1"></a>          idy <span class="ot">&lt;-</span> (<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(children))[<span class="sc">-</span>idx]</span>
<span id="cb36-31"><a href="#cb36-31" aria-hidden="true" tabindex="-1"></a>          life_stage_null[idy,] <span class="ot">&lt;-</span> life_stage</span>
<span id="cb36-32"><a href="#cb36-32" aria-hidden="true" tabindex="-1"></a>          life_stage <span class="ot">&lt;-</span> life_stage_null</span>
<span id="cb36-33"><a href="#cb36-33" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb36-34"><a href="#cb36-34" aria-hidden="true" tabindex="-1"></a>        life_stage <span class="ot">&lt;-</span> life_stage <span class="sc">%&gt;%</span> <span class="fu">bind_cols</span>(., fg_dat)</span>
<span id="cb36-35"><a href="#cb36-35" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb36-36"><a href="#cb36-36" aria-hidden="true" tabindex="-1"></a>        <span class="co">#' deal with cases where multiple records are returned for the same life stage:</span></span>
<span id="cb36-37"><a href="#cb36-37" aria-hidden="true" tabindex="-1"></a>        <span class="co">#' add a suffix to subsequent identical stages (adult_2, etc.)        </span></span>
<span id="cb36-38"><a href="#cb36-38" aria-hidden="true" tabindex="-1"></a>        life_stage <span class="ot">&lt;-</span> life_stage <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(stage) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">nth_stage_val =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">n</span>()) <span class="sc">%&gt;%</span> </span>
<span id="cb36-39"><a href="#cb36-39" aria-hidden="true" tabindex="-1"></a>          <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb36-40"><a href="#cb36-40" aria-hidden="true" tabindex="-1"></a>          <span class="fu">mutate</span>(<span class="at">stage =</span> <span class="fu">case_when</span>(</span>
<span id="cb36-41"><a href="#cb36-41" aria-hidden="true" tabindex="-1"></a>            nth_stage_val <span class="sc">==</span> <span class="dv">1</span> <span class="sc">~</span> stage,</span>
<span id="cb36-42"><a href="#cb36-42" aria-hidden="true" tabindex="-1"></a>            <span class="cn">TRUE</span> <span class="sc">~</span> <span class="fu">paste</span>(stage, nth_stage_val, <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb36-43"><a href="#cb36-43" aria-hidden="true" tabindex="-1"></a>          )</span>
<span id="cb36-44"><a href="#cb36-44" aria-hidden="true" tabindex="-1"></a>          ) <span class="sc">%&gt;%</span> </span>
<span id="cb36-45"><a href="#cb36-45" aria-hidden="true" tabindex="-1"></a>          <span class="fu">select</span>(<span class="sc">-</span>nth_stage_val)</span>
<span id="cb36-46"><a href="#cb36-46" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb36-47"><a href="#cb36-47" aria-hidden="true" tabindex="-1"></a>        <span class="co">#' create the output to return</span></span>
<span id="cb36-48"><a href="#cb36-48" aria-hidden="true" tabindex="-1"></a>        out <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">AphiaID =</span> <span class="fu">as.numeric</span>(life_stage<span class="sc">$</span>AphiaID),</span>
<span id="cb36-49"><a href="#cb36-49" aria-hidden="true" tabindex="-1"></a>                      <span class="at">stage =</span> life_stage<span class="sc">$</span>stage, <span class="at">fun_grp =</span> life_stage<span class="sc">$</span>measurementValue)</span>
<span id="cb36-50"><a href="#cb36-50" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb36-51"><a href="#cb36-51" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span>{</span>
<span id="cb36-52"><a href="#cb36-52" aria-hidden="true" tabindex="-1"></a>        <span class="co">#' If no life stage info, assume stage is adult</span></span>
<span id="cb36-53"><a href="#cb36-53" aria-hidden="true" tabindex="-1"></a>        out <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">AphiaID =</span> <span class="fu">as.numeric</span>(fg_dat<span class="sc">$</span>AphiaID),</span>
<span id="cb36-54"><a href="#cb36-54" aria-hidden="true" tabindex="-1"></a>                      <span class="at">stage =</span> <span class="st">"adult"</span>, <span class="at">fun_grp =</span> fg_dat<span class="sc">$</span>measurementValue)</span>
<span id="cb36-55"><a href="#cb36-55" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb36-56"><a href="#cb36-56" aria-hidden="true" tabindex="-1"></a>        <span class="co">#' Deal with cases where multiple records are returned (i.e. 2 or more adult fun_grps)</span></span>
<span id="cb36-57"><a href="#cb36-57" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span>(<span class="fu">nrow</span>(out) <span class="sc">&gt;</span> <span class="dv">1</span>){</span>
<span id="cb36-58"><a href="#cb36-58" aria-hidden="true" tabindex="-1"></a>          out <span class="ot">&lt;-</span> out <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(stage) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">nth_stage_val =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">n</span>()) <span class="sc">%&gt;%</span> </span>
<span id="cb36-59"><a href="#cb36-59" aria-hidden="true" tabindex="-1"></a>            <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb36-60"><a href="#cb36-60" aria-hidden="true" tabindex="-1"></a>            <span class="fu">mutate</span>(<span class="at">stage =</span> <span class="fu">case_when</span>(</span>
<span id="cb36-61"><a href="#cb36-61" aria-hidden="true" tabindex="-1"></a>              nth_stage_val <span class="sc">==</span> <span class="dv">1</span> <span class="sc">~</span> stage,</span>
<span id="cb36-62"><a href="#cb36-62" aria-hidden="true" tabindex="-1"></a>              <span class="cn">TRUE</span> <span class="sc">~</span> <span class="fu">paste</span>(stage, nth_stage_val, <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb36-63"><a href="#cb36-63" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb36-64"><a href="#cb36-64" aria-hidden="true" tabindex="-1"></a>            ) <span class="sc">%&gt;%</span> </span>
<span id="cb36-65"><a href="#cb36-65" aria-hidden="true" tabindex="-1"></a>            <span class="fu">select</span>(<span class="sc">-</span>nth_stage_val)</span>
<span id="cb36-66"><a href="#cb36-66" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb36-67"><a href="#cb36-67" aria-hidden="true" tabindex="-1"></a>      } </span>
<span id="cb36-68"><a href="#cb36-68" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb36-69"><a href="#cb36-69" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb36-70"><a href="#cb36-70" aria-hidden="true" tabindex="-1"></a>    <span class="co">#' add Pisces, from paraphyletic group: this takes priority over the above</span></span>
<span id="cb36-71"><a href="#cb36-71" aria-hidden="true" tabindex="-1"></a>    <span class="co">#' (e.g. class something as Pisces if it is a fish even if it is also listed as benthos)</span></span>
<span id="cb36-72"><a href="#cb36-72" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="st">"Paraphyletic group"</span> <span class="sc">%in%</span> attr_dat<span class="sc">$</span>measurementType) {</span>
<span id="cb36-73"><a href="#cb36-73" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span>(<span class="fu">first</span>(</span>
<span id="cb36-74"><a href="#cb36-74" aria-hidden="true" tabindex="-1"></a>        attr_dat<span class="sc">$</span>measurementValue[attr_dat<span class="sc">$</span>measurementType <span class="sc">==</span> <span class="st">"Paraphyletic group"</span>] <span class="sc">==</span> <span class="st">"Pisces"</span>)){</span>
<span id="cb36-75"><a href="#cb36-75" aria-hidden="true" tabindex="-1"></a>        out <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">AphiaID =</span> AphiaID, <span class="at">stage =</span> <span class="st">"adult"</span>,</span>
<span id="cb36-76"><a href="#cb36-76" aria-hidden="true" tabindex="-1"></a>                      <span class="at">fun_grp =</span> <span class="fu">first</span>(attr_dat<span class="sc">$</span>measurementValue[</span>
<span id="cb36-77"><a href="#cb36-77" aria-hidden="true" tabindex="-1"></a>                        attr_dat<span class="sc">$</span>measurementType <span class="sc">==</span> <span class="st">"Paraphyletic group"</span>]))</span>
<span id="cb36-78"><a href="#cb36-78" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb36-79"><a href="#cb36-79" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb36-80"><a href="#cb36-80" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb36-81"><a href="#cb36-81" aria-hidden="true" tabindex="-1"></a>  <span class="co">#' add other paraphyletic groups: Algae, Algae &gt; Macroalgae, Mangroves this DOES NOT takes priority over the above; i.e. only use if a functional group (e.g. phytoplankton) is not available</span></span>
<span id="cb36-82"><a href="#cb36-82" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">is.null</span>(out)){</span>
<span id="cb36-83"><a href="#cb36-83" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">identical</span>(<span class="fu">class</span>(attr_dat), <span class="st">"try-error"</span>) <span class="sc">&amp;&amp;</span> <span class="st">"Paraphyletic group"</span> <span class="sc">%in%</span> attr_dat<span class="sc">$</span>measurementType) {</span>
<span id="cb36-84"><a href="#cb36-84" aria-hidden="true" tabindex="-1"></a>      out <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">AphiaID =</span> AphiaID, <span class="at">stage =</span> <span class="st">"adult"</span>,</span>
<span id="cb36-85"><a href="#cb36-85" aria-hidden="true" tabindex="-1"></a>                    <span class="at">fun_grp =</span> <span class="fu">first</span>(attr_dat<span class="sc">$</span>measurementValue[</span>
<span id="cb36-86"><a href="#cb36-86" aria-hidden="true" tabindex="-1"></a>                      attr_dat<span class="sc">$</span>measurementType <span class="sc">==</span> <span class="st">"Paraphyletic group"</span>]))</span>
<span id="cb36-87"><a href="#cb36-87" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb36-88"><a href="#cb36-88" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb36-89"><a href="#cb36-89" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-90"><a href="#cb36-90" aria-hidden="true" tabindex="-1"></a>  <span class="co">#' check taxonomy for other groups</span></span>
<span id="cb36-91"><a href="#cb36-91" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">is.null</span>(out)){</span>
<span id="cb36-92"><a href="#cb36-92" aria-hidden="true" tabindex="-1"></a>    taxo_dat <span class="ot">&lt;-</span> <span class="fu">try</span>(<span class="fu">wm_classification</span>(AphiaID), <span class="at">silent =</span> <span class="cn">TRUE</span>)</span>
<span id="cb36-93"><a href="#cb36-93" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">identical</span>(<span class="fu">class</span>(taxo_dat), <span class="st">"try-error"</span>)){</span>
<span id="cb36-94"><a href="#cb36-94" aria-hidden="true" tabindex="-1"></a>      fg <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="cn">NA</span>)</span>
<span id="cb36-95"><a href="#cb36-95" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb36-96"><a href="#cb36-96" aria-hidden="true" tabindex="-1"></a>      fg <span class="ot">&lt;-</span> <span class="fu">case_when</span>(</span>
<span id="cb36-97"><a href="#cb36-97" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Aves"</span> <span class="sc">%in%</span> taxo_dat<span class="sc">$</span>scientificname <span class="sc">~</span> <span class="st">"birds"</span>,</span>
<span id="cb36-98"><a href="#cb36-98" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Mammalia"</span> <span class="sc">%in%</span> taxo_dat<span class="sc">$</span>scientificname <span class="sc">~</span> <span class="st">"mammals"</span>,</span>
<span id="cb36-99"><a href="#cb36-99" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Reptilia"</span> <span class="sc">%in%</span> taxo_dat<span class="sc">$</span>scientificname <span class="sc">~</span> <span class="st">"reptiles"</span>,</span>
<span id="cb36-100"><a href="#cb36-100" aria-hidden="true" tabindex="-1"></a>        <span class="cn">TRUE</span> <span class="sc">~</span> <span class="fu">as.character</span>(<span class="cn">NA</span>)</span>
<span id="cb36-101"><a href="#cb36-101" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb36-102"><a href="#cb36-102" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb36-103"><a href="#cb36-103" aria-hidden="true" tabindex="-1"></a>    out <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">AphiaID =</span> AphiaID, <span class="at">stage =</span> <span class="st">"adult"</span>, <span class="at">fun_grp =</span> fg)</span>
<span id="cb36-104"><a href="#cb36-104" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb36-105"><a href="#cb36-105" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-106"><a href="#cb36-106" aria-hidden="true" tabindex="-1"></a>  <span class="co">#' what if there are duplicate rows?</span></span>
<span id="cb36-107"><a href="#cb36-107" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> out[<span class="sc">!</span><span class="fu">duplicated</span>(out), ]</span>
<span id="cb36-108"><a href="#cb36-108" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> out[<span class="sc">!</span>(<span class="fu">is.na</span>(out<span class="sc">$</span>stage)),] <span class="co">#gets rid of NA stages</span></span>
<span id="cb36-109"><a href="#cb36-109" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-110"><a href="#cb36-110" aria-hidden="true" tabindex="-1"></a>  <span class="co">#' Replace 'not applicable' with NA</span></span>
<span id="cb36-111"><a href="#cb36-111" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> out <span class="sc">%&gt;%</span> <span class="fu">mutate_all</span>(<span class="sc">~</span><span class="fu">ifelse</span>(. <span class="sc">==</span> <span class="st">'not applicable'</span>, <span class="cn">NA</span>, .))</span>
<span id="cb36-112"><a href="#cb36-112" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-113"><a href="#cb36-113" aria-hidden="true" tabindex="-1"></a>  <span class="co">#' Keep only the life stage with the highest number (if applicable)</span></span>
<span id="cb36-114"><a href="#cb36-114" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> out <span class="sc">%&gt;%</span></span>
<span id="cb36-115"><a href="#cb36-115" aria-hidden="true" tabindex="-1"></a>    <span class="fu">separate</span>(stage, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"stage"</span>, <span class="st">"number"</span>), <span class="at">sep =</span> <span class="st">"_"</span>, <span class="at">remove =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb36-116"><a href="#cb36-116" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(AphiaID, stage) <span class="sc">%&gt;%</span></span>
<span id="cb36-117"><a href="#cb36-117" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(<span class="fu">desc</span>(number)) <span class="sc">%&gt;%</span></span>
<span id="cb36-118"><a href="#cb36-118" aria-hidden="true" tabindex="-1"></a>    <span class="fu">slice</span>(<span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb36-119"><a href="#cb36-119" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb36-120"><a href="#cb36-120" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">-</span>number)</span>
<span id="cb36-121"><a href="#cb36-121" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-122"><a href="#cb36-122" aria-hidden="true" tabindex="-1"></a>  <span class="co">#' Convert specific life stages to 'larva_other'</span></span>
<span id="cb36-123"><a href="#cb36-123" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> out <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">stage =</span> <span class="fu">case_when</span>(</span>
<span id="cb36-124"><a href="#cb36-124" aria-hidden="true" tabindex="-1"></a>    stage <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">'cerinula'</span>, <span class="st">'larva &gt; planula'</span>, <span class="st">'medusa'</span>, <span class="st">'zoea'</span>, <span class="st">'nauplius'</span>, <span class="st">'polyp'</span>, <span class="st">'medusoid'</span>, <span class="st">'ephyra'</span>, <span class="st">'colony'</span>) <span class="sc">~</span> <span class="st">'larva_other'</span>,</span>
<span id="cb36-125"><a href="#cb36-125" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> stage</span>
<span id="cb36-126"><a href="#cb36-126" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb36-127"><a href="#cb36-127" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-128"><a href="#cb36-128" aria-hidden="true" tabindex="-1"></a>  <span class="co">#' do some tidying of the output: tidy up functional_group text</span></span>
<span id="cb36-129"><a href="#cb36-129" aria-hidden="true" tabindex="-1"></a>  <span class="co">#' and spread to give one column per life stage</span></span>
<span id="cb36-130"><a href="#cb36-130" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> out <span class="sc">%&gt;%</span> </span>
<span id="cb36-131"><a href="#cb36-131" aria-hidden="true" tabindex="-1"></a>    <span class="fu">distinct</span>(AphiaID, stage, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span>  <span class="co"># Add this line to remove duplicates</span></span>
<span id="cb36-132"><a href="#cb36-132" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">functional_group =</span> <span class="fu">case_when</span>(</span>
<span id="cb36-133"><a href="#cb36-133" aria-hidden="true" tabindex="-1"></a>      <span class="fu">str_detect</span>(fun_grp, <span class="st">"&gt;"</span>) <span class="sc">~</span> <span class="fu">tolower</span>(<span class="fu">word</span>(fun_grp, <span class="sc">-</span><span class="dv">1</span>)),</span>
<span id="cb36-134"><a href="#cb36-134" aria-hidden="true" tabindex="-1"></a>      fun_grp <span class="sc">==</span> <span class="st">"Pisces"</span> <span class="sc">~</span> <span class="st">"fish"</span>,</span>
<span id="cb36-135"><a href="#cb36-135" aria-hidden="true" tabindex="-1"></a>      fun_grp <span class="sc">==</span> <span class="st">"Algae &gt; Macroalgae"</span> <span class="sc">~</span> <span class="st">"macroalgae"</span>,</span>
<span id="cb36-136"><a href="#cb36-136" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span> <span class="sc">~</span> <span class="fu">tolower</span>(fun_grp)</span>
<span id="cb36-137"><a href="#cb36-137" aria-hidden="true" tabindex="-1"></a>    )) <span class="sc">%&gt;%</span></span>
<span id="cb36-138"><a href="#cb36-138" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>fun_grp) <span class="sc">%&gt;%</span></span>
<span id="cb36-139"><a href="#cb36-139" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spread</span>(stage, functional_group)</span>
<span id="cb36-140"><a href="#cb36-140" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-141"><a href="#cb36-141" aria-hidden="true" tabindex="-1"></a>  <span class="co">#' Change column name NA to other</span></span>
<span id="cb36-142"><a href="#cb36-142" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(out)[<span class="fu">colnames</span>(out) <span class="sc">==</span> <span class="st">"NA"</span>] <span class="ot">&lt;-</span> <span class="st">"other"</span></span>
<span id="cb36-143"><a href="#cb36-143" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-144"><a href="#cb36-144" aria-hidden="true" tabindex="-1"></a>  <span class="co">#' output the fg data</span></span>
<span id="cb36-145"><a href="#cb36-145" aria-hidden="true" tabindex="-1"></a>  out</span>
<span id="cb36-146"><a href="#cb36-146" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Before we run anything, we need to extract the taxa names we want to run through the functions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>eDNA_taxa_unique_list <span class="ot">&lt;-</span> <span class="fu">unique</span>(eDNA_long_data_cleaned<span class="sc">$</span>taxonomy_clean) <span class="co">#Get unique taxa for WoRMs input as list</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>eDNA_taxa_unique_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">taxonomy =</span> <span class="fu">tolower</span>(<span class="fu">unique</span>(eDNA_long_data_cleaned<span class="sc">$</span>taxonomy_clean)), </span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>) <span class="do">##Get unique taxa for WoRMs input as df</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(eDNA_taxa_unique_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                taxonomy
1              amoebozoa
2               discosea
3            flabellinia
4           neoparamoeba
5               vannella
6 protostelium nocturnum</code></pre>
</div>
</div>
</section>
<section id="get-aphia-ids" class="level3">
<h3 class="anchored" data-anchor-id="get-aphia-ids">Get Aphia IDs</h3>
<p>Let’s use the function <code>get_wormsid()</code> to obtain Aphia IDs for our taxa.</p>
<p>The duration of this task is approximately 15-minutes for this data set, but can vary depending on the size of the data. Any non-matched species will produce a NA.</p>
<p>To prevent the document taking hours to render, we’ve put # in front of the code for now. We’ve saved the output and reloaded the data back in.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co">#aphiaID_worms_output &lt;- eDNA_taxa_unique_list %&gt;%</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="co">#purrr::map(get_wormsid_noerror, .progress = TRUE) %&gt;%</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="co">#enframe() %&gt;%</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="co">#unnest(cols = everything()) %&gt;%</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="co">#select(-name)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We’ll save this output to reduce time when rerunning.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co">#write.csv(aphiaID_worms_output, "Processed_Data/aphiaID_worms_output.csv")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s reload the data back in and do some formatting.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>aphiaID_worms_output <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"Processed_Data/aphiaID_worms_output.csv"</span>, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>aphiaID_worms_output<span class="sc">$</span>taxonomy_clean <span class="ot">=</span> eDNA_taxa_unique_df<span class="sc">$</span>taxonomy <span class="co">#add previous taxonomy name for joining later (order remains the same)</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>aphiaID_worms_output<span class="sc">$</span>aphiaID <span class="ot">=</span> <span class="fu">as.integer</span>(aphiaID_worms_output<span class="sc">$</span>aphiaID) <span class="co">#convert to integer for joining</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(aphiaID_worms_output)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 sciname aphiaID         taxonomy_clean
1              amoebozoa  391848              amoebozoa
2               discosea  582189               discosea
3            flabellinia  582205            flabellinia
4           neoparamoeba  120567           neoparamoeba
5               vannella  120566               vannella
6 protostelium nocturnum      NA protostelium nocturnum</code></pre>
</div>
</div>
<p>Let’s check how many matches we made.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "You matched 1781 / 2199 taxa with AphiaIDs. Therefore, 418 entries could not be matched and do not have AphiaIDs."</code></pre>
</div>
</div>
<p>Now let’s add the Aphia IDs to our main data set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>eDNA_long_data_final <span class="ot">&lt;-</span> <span class="fu">full_join</span>(eDNA_long_data_cleaned, aphiaID_worms_output, <span class="at">by =</span> <span class="fu">join_by</span>(taxonomy_clean))</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(eDNA_long_data_final)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   841939 obs. of  73 variables:
 $ taxonomy             : chr  "Amoebozoa" "Amoebozoa" "Amoebozoa" "Amoebozoa" ...
 $ taxonomy_clean       : chr  "amoebozoa" "amoebozoa" "amoebozoa" "amoebozoa" ...
 $ ASV                  : chr  "ASV_3057" "ASV_3512" "ASV_3942" "ASV_4258" ...
 $ Seqs                 : chr  "TTAACGAACGAGACCTTAACCTATTAAATAGTTGCGCGAACTTGAATCATGCGTCATGTTATCTTCGGATAATGTTTCCTCCATGGTTCTATCGTTCTGTGAAAACTTCTT"| __truncated__ "ATAACGAACGAGACCTCAGCCTGCTAACTAGTATGCCTTAGCCTTGTTTTCTTGGCTAAGGAATTTATTGAATATACTTCTTAGAGGGACTATCCGTGTCTAGCGGGTGGAAGTTTGAG" "TTAACGAACGAGACCTTAACCTATTAAATAGTTGCGCGAACTTGAATCATGCGTCAGTTCGTCTTTTGGGCGGACTGTCCTCCATGGTTCTATCGTTCTGTGAAAACTTCT"| __truncated__ "TTAACGAACGAGACCTTAACCTATTAAATAGTTGCGCGAACTTGAATCATGCGTCAGTTCGTCCTTCGTGGGCGAGCTGTCCTCCATGGTTCTATCGTTCTGTGAAAACTT"| __truncated__ ...
 $ reads                : int  0 0 0 0 0 0 0 0 0 0 ...
 $ region               : Factor w/ 6 levels "Scotland","South Wales",..: 1 1 1 1 1 1 1 1 1 1 ...
 $ fieldID              : chr  "PHD1-SCH01-01" "PHD1-SCH01-01" "PHD1-SCH01-01" "PHD1-SCH01-01" ...
 $ primer               : chr  "18S" "18S" "18S" "18S" ...
 $ eventID              : chr  "PHD1-2022.06.12-SCH01-01" "PHD1-2022.06.12-SCH01-01" "PHD1-2022.06.12-SCH01-01" "PHD1-2022.06.12-SCH01-01" ...
 $ projectID            : chr  "PHD1" "PHD1" "PHD1" "PHD1" ...
 $ localityID           : chr  "SCH01" "SCH01" "SCH01" "SCH01" ...
 $ sampleID             : chr  "1" "1" "1" "1" ...
 $ sampleType           : chr  "sample" "sample" "sample" "sample" ...
 $ controlCheck         : chr  "sample" "sample" "sample" "sample" ...
 $ negCheckLogical      : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
 $ pairedRepLogical     : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
 $ batch_extraction     : int  1 1 1 1 1 1 1 1 1 1 ...
 $ batch_PCR_CO1        : int  2 2 2 2 2 2 2 2 2 2 ...
 $ plate_col_CO1        : int  1 1 1 1 1 1 1 1 1 1 ...
 $ plate_row_CO1        : chr  "A" "A" "A" "A" ...
 $ batch_PCR_18S        : int  4 4 4 4 4 4 4 4 4 4 ...
 $ plate_col_18S        : int  1 1 1 1 1 1 1 1 1 1 ...
 $ plate_row_18S        : chr  "A" "A" "A" "A" ...
 $ date_extraction      : chr  "24.02.2023" "24.02.2023" "24.02.2023" "24.02.2023" ...
 $ date_amplified_CO1   : chr  "28.03.2023" "28.03.2023" "28.03.2023" "28.03.2023" ...
 $ date_amplified_18S   : chr  "27.03.2023" "27.03.2023" "27.03.2023" "27.03.2023" ...
 $ sample_18S           : chr  "S66" "S66" "S66" "S66" ...
 $ sample_CO1           : chr  "Adapter385.SCH01.01" "Adapter385.SCH01.01" "Adapter385.SCH01.01" "Adapter385.SCH01.01" ...
 $ COUNT                : int  2 2 2 2 2 2 2 2 2 2 ...
 $ dna_concentration_CO1: num  82.5 82.5 82.5 82.5 82.5 ...
 $ dna_concentration_18S: num  55.4 55.4 55.4 55.4 55.4 ...
 $ continent            : chr  "Europe" "Europe" "Europe" "Europe" ...
 $ country              : chr  "Scotland" "Scotland" "Scotland" "Scotland" ...
 $ county               : chr  "Sutherland" "Sutherland" "Sutherland" "Sutherland" ...
 $ waterBody            : chr  "rocky" "rocky" "rocky" "rocky" ...
 $ verbatimLocality     : chr  "Scourie" "Scourie" "Scourie" "Scourie" ...
 $ exposure             : chr  "Exposed" "Exposed" "Exposed" "Exposed" ...
 $ weather              : chr  "moderate rain, strong winds, very cloudy" "moderate rain, strong winds, very cloudy" "moderate rain, strong winds, very cloudy" "moderate rain, strong winds, very cloudy" ...
 $ tide                 : chr  "outgoing" "outgoing" "outgoing" "outgoing" ...
 $ lowWater             : chr  "12:23" "12:23" "12:23" "12:23" ...
 $ decimalLongitude     : num  -5.17 -5.17 -5.17 -5.17 -5.17 ...
 $ decimalLatitude      : num  58.4 58.4 58.4 58.4 58.4 ...
 $ recordedBy           : chr  "Dina-Leigh Simons; Nova Mieszkowska" "Dina-Leigh Simons; Nova Mieszkowska" "Dina-Leigh Simons; Nova Mieszkowska" "Dina-Leigh Simons; Nova Mieszkowska" ...
 $ year                 : int  2022 2022 2022 2022 2022 2022 2022 2022 2022 2022 ...
 $ month                : int  6 6 6 6 6 6 6 6 6 6 ...
 $ day                  : int  12 12 12 12 12 12 12 12 12 12 ...
 $ eventTime            : chr  "09:31" "09:31" "09:31" "09:31" ...
 $ samplingProtocol     : chr  "Sterivex-0.22" "Sterivex-0.22" "Sterivex-0.22" "Sterivex-0.22" ...
 $ pumpType             : chr  "Syringe 60mL, sealant gun" "Syringe 60mL, sealant gun" "Syringe 60mL, sealant gun" "Syringe 60mL, sealant gun" ...
 $ repVol..mL.          : int  1000 1000 1000 1000 1000 1000 1000 1000 1000 1000 ...
 $ algaeCover.          : int  0 0 0 0 0 0 0 0 0 0 ...
 $ remarks              : chr  "Filtered on shore" "Filtered on shore" "Filtered on shore" "Filtered on shore" ...
 $ daysBeforefreeze     : int  9 9 9 9 9 9 9 9 9 9 ...
 $ preservation         : chr  "EtOH" "EtOH" "EtOH" "EtOH" ...
 $ type                 : chr  "Rockpool" "Rockpool" "Rockpool" "Rockpool" ...
 $ shorePosition        : chr  "High" "High" "High" "High" ...
 $ rockpoolarea.cm.2.   : num  7443 7443 7443 7443 7443 ...
 $ depth1               : num  NA NA NA NA NA NA NA NA NA NA ...
 $ depth2               : num  NA NA NA NA NA NA NA NA NA NA ...
 $ averageDepth..cm.    : num  25 25 25 25 25 25 25 25 25 25 ...
 $ rockpoolVol.m.3.     : num  0.186 0.186 0.186 0.186 0.186 ...
 $ pH1                  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ pH2                  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ averagePH            : num  8.59 8.59 8.59 8.59 8.59 8.59 8.59 8.59 8.59 8.59 ...
 $ temp1                : chr  NA NA NA NA ...
 $ temp2                : chr  NA NA NA NA ...
 $ averageTemp          : num  13 13 13 13 13 13 13 13 13 13 ...
 $ siteImage            : chr  "IMG_SCH01-01.jpeg" "IMG_SCH01-01.jpeg" "IMG_SCH01-01.jpeg" "IMG_SCH01-01.jpeg" ...
 $ rockType             : chr  "metamorphic" "metamorphic" "metamorphic" "metamorphic" ...
 $ rockFold             : chr  NA NA NA NA ...
 $ fullID               : chr  "PHD1-SCH01-01_18S" "PHD1-SCH01-01_18S" "PHD1-SCH01-01_18S" "PHD1-SCH01-01_18S" ...
 $ sciname              : chr  "amoebozoa" "amoebozoa" "amoebozoa" "amoebozoa" ...
 $ aphiaID              : int  391848 391848 391848 391848 391848 391848 391848 391848 391848 391848 ...</code></pre>
</div>
</div>
</section>
<section id="get-taxonomic-metadata" class="level3">
<h3 class="anchored" data-anchor-id="get-taxonomic-metadata">Get taxonomic metadata</h3>
<p>Next, let’s use the function <code>get_wormsmeta()</code> to obtain taxonomic information for our Aphia IDs. This command takes a few minutes to run.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>worms_meta_output <span class="ot">&lt;-</span> <span class="fu">get_wormsmeta</span>(aphiaID_worms_output)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s look at the output. We can see we now have lots of information for each taxon.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(worms_meta_output)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>tibble [2,199 × 27] (S3: tbl_df/tbl/data.frame)
 $ AphiaID          : int [1:2199] 391848 582189 582205 120567 120566 NA NA 1317 NA NA ...
 $ url              : chr [1:2199] "https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=391848" "https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=582189" "https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=582205" "https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=120567" ...
 $ scientificname   : chr [1:2199] "Amoebozoa" "Discosea" "Flabellinia" "Neoparamoeba" ...
 $ authority        : chr [1:2199] "Lühe, 1913, emend. Cavalier-Smith, 1998" "Cavalier-Smith et al., 2004" "Smirnov et al., 2005" "F.C.Page, 1987" ...
 $ status           : chr [1:2199] "accepted" "accepted" "accepted" "accepted" ...
 $ unacceptreason   : logi [1:2199] NA NA NA NA NA NA ...
 $ taxonRankID      : int [1:2199] 30 60 70 180 180 NA NA 100 NA NA ...
 $ rank             : chr [1:2199] "Phylum" "Class" "Subclass" "Genus" ...
 $ valid_AphiaID    : int [1:2199] 391848 582189 582205 120567 120566 NA NA 1317 NA NA ...
 $ valid_name       : chr [1:2199] "Amoebozoa" "Discosea" "Flabellinia" "Neoparamoeba" ...
 $ valid_authority  : chr [1:2199] "Lühe, 1913, emend. Cavalier-Smith, 1998" "Cavalier-Smith et al., 2004" "Smirnov et al., 2005" "F.C.Page, 1987" ...
 $ parentNameUsageID: int [1:2199] 582160 582168 582189 22725 22724 NA NA 582188 NA NA ...
 $ kingdom          : chr [1:2199] "Protozoa" "Protozoa" "Protozoa" "Protozoa" ...
 $ phylum           : chr [1:2199] "Amoebozoa" "Amoebozoa" "Amoebozoa" "Amoebozoa" ...
 $ class            : chr [1:2199] NA "Discosea" "Discosea" "Discosea" ...
 $ order            : chr [1:2199] NA NA NA "Dactylopodida" ...
 $ family           : chr [1:2199] NA NA NA "Vexilliferidae" ...
 $ genus            : chr [1:2199] NA NA NA "Neoparamoeba" ...
 $ citation         : chr [1:2199] "WoRMS (2025). Amoebozoa. Accessed at: https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=391848 on 2025-03-11" "Guiry, M.D. &amp; Guiry, G.M. (2025). AlgaeBase. World-wide electronic publication, National University of Ireland,"| __truncated__ "Guiry, M.D. &amp; Guiry, G.M. (2025). AlgaeBase. World-wide electronic publication, National University of Ireland,"| __truncated__ "Guiry, M.D. &amp; Guiry, G.M. (2025). AlgaeBase. World-wide electronic publication, National University of Ireland,"| __truncated__ ...
 $ lsid             : chr [1:2199] "urn:lsid:marinespecies.org:taxname:391848" "urn:lsid:marinespecies.org:taxname:582189" "urn:lsid:marinespecies.org:taxname:582205" "urn:lsid:marinespecies.org:taxname:120567" ...
 $ isMarine         : int [1:2199] 1 1 1 1 1 NA NA 1 NA NA ...
 $ isBrackish       : int [1:2199] 1 1 NA NA NA NA NA 1 NA NA ...
 $ isFreshwater     : int [1:2199] 1 1 1 NA 1 NA NA 1 NA NA ...
 $ isTerrestrial    : int [1:2199] 1 1 NA NA NA NA NA 1 NA NA ...
 $ isExtinct        : int [1:2199] NA NA NA NA NA NA NA NA NA NA ...
 $ match_type       : chr [1:2199] "exact" "exact" "exact" "exact" ...
 $ modified         : chr [1:2199] "2011-10-14T10:07:31.253Z" "2011-10-14T10:07:31.253Z" "2011-10-14T10:07:31.253Z" "2024-02-21T07:10:53.720Z" ...</code></pre>
</div>
</div>
<p>A bit of tidying of this output is required to be able to join it with our own.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>worms_meta_output_tidy <span class="ot">&lt;-</span> worms_meta_output <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">aphiaID =</span> AphiaID) <span class="sc">%&gt;%</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(aphiaID)) <span class="sc">%&gt;%</span> <span class="co">#remove NAs</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>() <span class="co">#remove duplicates</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s add the taxonomic metadata to our main data set and check the joining of data sets worked.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>eDNA_long_data_final <span class="ot">&lt;-</span> <span class="fu">right_join</span>(eDNA_long_data_final, worms_meta_output_tidy, <span class="at">by =</span> <span class="st">"aphiaID"</span>) <span class="co">#worms meta</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(<span class="fu">is.na</span>(eDNA_long_data_final<span class="sc">$</span>aphiaID), <span class="at">arr.ind=</span><span class="cn">TRUE</span>) <span class="co">#check NAs in IDs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>integer(0)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">duplicated</span>(eDNA_long_data_final)) <span class="co">#check for duplicates</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
</div>
</section>
<section id="get-broad-functional-groups" class="level3">
<h3 class="anchored" data-anchor-id="get-broad-functional-groups">Get broad functional groups</h3>
<p>Finally, let’s use the function&nbsp;<code>get_worms_fgrp()</code> to obtain the broad taxonomic groups of our taxa.</p>
<p>Similar to above, this command takes a long time (approximately 30 minutes with this data set). We have put # in front of this code to reduce time to render.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>species <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">AphiaID =</span> <span class="fu">c</span>(<span class="fu">unique</span>(eDNA_long_data_final<span class="sc">$</span>aphiaID))) <span class="co">#list aphiaIDs</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="co">#worms_fgrp_output &lt;- species %&gt;%</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>  <span class="co">#group_by(AphiaID) %&gt;%</span></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>  <span class="co">#do(get_worms_fgrp(AphiaID = .$AphiaID))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We’ll save this output to reduce time when rerunning.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co">#write.csv(worms_fgrp_output, "Processed_Data/worms_fgrp_output.csv")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s reload the data back in.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>worms_fgrp_output <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"Processed_Data/worms_fgrp_output.csv"</span>, <span class="at">row.names =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s see how many matches we got.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "You found function groups for 1435 / 1780 adult taxa. Therefore, 345 have no functional group."</code></pre>
</div>
</div>
<p>Now let’s add the functional group information to our main data set and check the joining of data sets worked.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>worms_fgrp_output_tidy<span class="ot">&lt;-</span> worms_fgrp_output <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">aphiaID =</span> AphiaID)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>eDNA_long_data_final <span class="ot">&lt;-</span> <span class="fu">left_join</span>(eDNA_long_data_final, worms_fgrp_output_tidy)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Quick look at the final long data set (remember this has not been decontaminated yet!).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(eDNA_long_data_final)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   715430 obs. of  104 variables:
 $ taxonomy             : chr  "Amoebozoa" "Amoebozoa" "Amoebozoa" "Amoebozoa" ...
 $ taxonomy_clean       : chr  "amoebozoa" "amoebozoa" "amoebozoa" "amoebozoa" ...
 $ ASV                  : chr  "ASV_3057" "ASV_3512" "ASV_3942" "ASV_4258" ...
 $ Seqs                 : chr  "TTAACGAACGAGACCTTAACCTATTAAATAGTTGCGCGAACTTGAATCATGCGTCATGTTATCTTCGGATAATGTTTCCTCCATGGTTCTATCGTTCTGTGAAAACTTCTT"| __truncated__ "ATAACGAACGAGACCTCAGCCTGCTAACTAGTATGCCTTAGCCTTGTTTTCTTGGCTAAGGAATTTATTGAATATACTTCTTAGAGGGACTATCCGTGTCTAGCGGGTGGAAGTTTGAG" "TTAACGAACGAGACCTTAACCTATTAAATAGTTGCGCGAACTTGAATCATGCGTCAGTTCGTCTTTTGGGCGGACTGTCCTCCATGGTTCTATCGTTCTGTGAAAACTTCT"| __truncated__ "TTAACGAACGAGACCTTAACCTATTAAATAGTTGCGCGAACTTGAATCATGCGTCAGTTCGTCCTTCGTGGGCGAGCTGTCCTCCATGGTTCTATCGTTCTGTGAAAACTT"| __truncated__ ...
 $ reads                : int  0 0 0 0 0 0 0 0 0 0 ...
 $ region               : Factor w/ 6 levels "Scotland","South Wales",..: 1 1 1 1 1 1 1 1 1 1 ...
 $ fieldID              : chr  "PHD1-SCH01-01" "PHD1-SCH01-01" "PHD1-SCH01-01" "PHD1-SCH01-01" ...
 $ primer               : chr  "18S" "18S" "18S" "18S" ...
 $ eventID              : chr  "PHD1-2022.06.12-SCH01-01" "PHD1-2022.06.12-SCH01-01" "PHD1-2022.06.12-SCH01-01" "PHD1-2022.06.12-SCH01-01" ...
 $ projectID            : chr  "PHD1" "PHD1" "PHD1" "PHD1" ...
 $ localityID           : chr  "SCH01" "SCH01" "SCH01" "SCH01" ...
 $ sampleID             : chr  "1" "1" "1" "1" ...
 $ sampleType           : chr  "sample" "sample" "sample" "sample" ...
 $ controlCheck         : chr  "sample" "sample" "sample" "sample" ...
 $ negCheckLogical      : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
 $ pairedRepLogical     : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
 $ batch_extraction     : int  1 1 1 1 1 1 1 1 1 1 ...
 $ batch_PCR_CO1        : int  2 2 2 2 2 2 2 2 2 2 ...
 $ plate_col_CO1        : int  1 1 1 1 1 1 1 1 1 1 ...
 $ plate_row_CO1        : chr  "A" "A" "A" "A" ...
 $ batch_PCR_18S        : int  4 4 4 4 4 4 4 4 4 4 ...
 $ plate_col_18S        : int  1 1 1 1 1 1 1 1 1 1 ...
 $ plate_row_18S        : chr  "A" "A" "A" "A" ...
 $ date_extraction      : chr  "24.02.2023" "24.02.2023" "24.02.2023" "24.02.2023" ...
 $ date_amplified_CO1   : chr  "28.03.2023" "28.03.2023" "28.03.2023" "28.03.2023" ...
 $ date_amplified_18S   : chr  "27.03.2023" "27.03.2023" "27.03.2023" "27.03.2023" ...
 $ sample_18S           : chr  "S66" "S66" "S66" "S66" ...
 $ sample_CO1           : chr  "Adapter385.SCH01.01" "Adapter385.SCH01.01" "Adapter385.SCH01.01" "Adapter385.SCH01.01" ...
 $ COUNT                : int  2 2 2 2 2 2 2 2 2 2 ...
 $ dna_concentration_CO1: num  82.5 82.5 82.5 82.5 82.5 ...
 $ dna_concentration_18S: num  55.4 55.4 55.4 55.4 55.4 ...
 $ continent            : chr  "Europe" "Europe" "Europe" "Europe" ...
 $ country              : chr  "Scotland" "Scotland" "Scotland" "Scotland" ...
 $ county               : chr  "Sutherland" "Sutherland" "Sutherland" "Sutherland" ...
 $ waterBody            : chr  "rocky" "rocky" "rocky" "rocky" ...
 $ verbatimLocality     : chr  "Scourie" "Scourie" "Scourie" "Scourie" ...
 $ exposure             : chr  "Exposed" "Exposed" "Exposed" "Exposed" ...
 $ weather              : chr  "moderate rain, strong winds, very cloudy" "moderate rain, strong winds, very cloudy" "moderate rain, strong winds, very cloudy" "moderate rain, strong winds, very cloudy" ...
 $ tide                 : chr  "outgoing" "outgoing" "outgoing" "outgoing" ...
 $ lowWater             : chr  "12:23" "12:23" "12:23" "12:23" ...
 $ decimalLongitude     : num  -5.17 -5.17 -5.17 -5.17 -5.17 ...
 $ decimalLatitude      : num  58.4 58.4 58.4 58.4 58.4 ...
 $ recordedBy           : chr  "Dina-Leigh Simons; Nova Mieszkowska" "Dina-Leigh Simons; Nova Mieszkowska" "Dina-Leigh Simons; Nova Mieszkowska" "Dina-Leigh Simons; Nova Mieszkowska" ...
 $ year                 : int  2022 2022 2022 2022 2022 2022 2022 2022 2022 2022 ...
 $ month                : int  6 6 6 6 6 6 6 6 6 6 ...
 $ day                  : int  12 12 12 12 12 12 12 12 12 12 ...
 $ eventTime            : chr  "09:31" "09:31" "09:31" "09:31" ...
 $ samplingProtocol     : chr  "Sterivex-0.22" "Sterivex-0.22" "Sterivex-0.22" "Sterivex-0.22" ...
 $ pumpType             : chr  "Syringe 60mL, sealant gun" "Syringe 60mL, sealant gun" "Syringe 60mL, sealant gun" "Syringe 60mL, sealant gun" ...
 $ repVol..mL.          : int  1000 1000 1000 1000 1000 1000 1000 1000 1000 1000 ...
 $ algaeCover.          : int  0 0 0 0 0 0 0 0 0 0 ...
 $ remarks              : chr  "Filtered on shore" "Filtered on shore" "Filtered on shore" "Filtered on shore" ...
 $ daysBeforefreeze     : int  9 9 9 9 9 9 9 9 9 9 ...
 $ preservation         : chr  "EtOH" "EtOH" "EtOH" "EtOH" ...
 $ type                 : chr  "Rockpool" "Rockpool" "Rockpool" "Rockpool" ...
 $ shorePosition        : chr  "High" "High" "High" "High" ...
 $ rockpoolarea.cm.2.   : num  7443 7443 7443 7443 7443 ...
 $ depth1               : num  NA NA NA NA NA NA NA NA NA NA ...
 $ depth2               : num  NA NA NA NA NA NA NA NA NA NA ...
 $ averageDepth..cm.    : num  25 25 25 25 25 25 25 25 25 25 ...
 $ rockpoolVol.m.3.     : num  0.186 0.186 0.186 0.186 0.186 ...
 $ pH1                  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ pH2                  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ averagePH            : num  8.59 8.59 8.59 8.59 8.59 8.59 8.59 8.59 8.59 8.59 ...
 $ temp1                : chr  NA NA NA NA ...
 $ temp2                : chr  NA NA NA NA ...
 $ averageTemp          : num  13 13 13 13 13 13 13 13 13 13 ...
 $ siteImage            : chr  "IMG_SCH01-01.jpeg" "IMG_SCH01-01.jpeg" "IMG_SCH01-01.jpeg" "IMG_SCH01-01.jpeg" ...
 $ rockType             : chr  "metamorphic" "metamorphic" "metamorphic" "metamorphic" ...
 $ rockFold             : chr  NA NA NA NA ...
 $ fullID               : chr  "PHD1-SCH01-01_18S" "PHD1-SCH01-01_18S" "PHD1-SCH01-01_18S" "PHD1-SCH01-01_18S" ...
 $ sciname              : chr  "amoebozoa" "amoebozoa" "amoebozoa" "amoebozoa" ...
 $ aphiaID              : int  391848 391848 391848 391848 391848 391848 391848 391848 391848 391848 ...
 $ url                  : chr  "https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=391848" "https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=391848" "https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=391848" "https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=391848" ...
 $ scientificname       : chr  "Amoebozoa" "Amoebozoa" "Amoebozoa" "Amoebozoa" ...
 $ authority            : chr  "Lühe, 1913, emend. Cavalier-Smith, 1998" "Lühe, 1913, emend. Cavalier-Smith, 1998" "Lühe, 1913, emend. Cavalier-Smith, 1998" "Lühe, 1913, emend. Cavalier-Smith, 1998" ...
 $ status               : chr  "accepted" "accepted" "accepted" "accepted" ...
 $ unacceptreason       : logi  NA NA NA NA NA NA ...
 $ taxonRankID          : int  30 30 30 30 30 30 30 30 30 30 ...
 $ rank                 : chr  "Phylum" "Phylum" "Phylum" "Phylum" ...
 $ valid_AphiaID        : int  391848 391848 391848 391848 391848 391848 391848 391848 391848 391848 ...
 $ valid_name           : chr  "Amoebozoa" "Amoebozoa" "Amoebozoa" "Amoebozoa" ...
 $ valid_authority      : chr  "Lühe, 1913, emend. Cavalier-Smith, 1998" "Lühe, 1913, emend. Cavalier-Smith, 1998" "Lühe, 1913, emend. Cavalier-Smith, 1998" "Lühe, 1913, emend. Cavalier-Smith, 1998" ...
 $ parentNameUsageID    : int  582160 582160 582160 582160 582160 582160 582160 582160 582160 582160 ...
 $ kingdom              : chr  "Protozoa" "Protozoa" "Protozoa" "Protozoa" ...
 $ phylum               : chr  "Amoebozoa" "Amoebozoa" "Amoebozoa" "Amoebozoa" ...
 $ class                : chr  NA NA NA NA ...
 $ order                : chr  NA NA NA NA ...
 $ family               : chr  NA NA NA NA ...
 $ genus                : chr  NA NA NA NA ...
 $ citation             : chr  "WoRMS (2025). Amoebozoa. Accessed at: https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=391848 on 2025-03-11" "WoRMS (2025). Amoebozoa. Accessed at: https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=391848 on 2025-03-11" "WoRMS (2025). Amoebozoa. Accessed at: https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=391848 on 2025-03-11" "WoRMS (2025). Amoebozoa. Accessed at: https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=391848 on 2025-03-11" ...
 $ lsid                 : chr  "urn:lsid:marinespecies.org:taxname:391848" "urn:lsid:marinespecies.org:taxname:391848" "urn:lsid:marinespecies.org:taxname:391848" "urn:lsid:marinespecies.org:taxname:391848" ...
 $ isMarine             : int  1 1 1 1 1 1 1 1 1 1 ...
 $ isBrackish           : int  1 1 1 1 1 1 1 1 1 1 ...
 $ isFreshwater         : int  1 1 1 1 1 1 1 1 1 1 ...
 $ isTerrestrial        : int  1 1 1 1 1 1 1 1 1 1 ...
 $ isExtinct            : int  NA NA NA NA NA NA NA NA NA NA ...
 $ match_type           : chr  "exact" "exact" "exact" "exact" ...
 $ modified             : chr  "2011-10-14T10:07:31.253Z" "2011-10-14T10:07:31.253Z" "2011-10-14T10:07:31.253Z" "2011-10-14T10:07:31.253Z" ...
  [list output truncated]</code></pre>
</div>
</div>
</section>
</section>
<section id="create-phyloseq-object" class="level2">
<h2 class="anchored" data-anchor-id="create-phyloseq-object">Create phyloseq object</h2>
<p>We’ve now got a tidied long data set with lots of meta data. To aid further check and analyses, we can convert our long data into a <code>phyloseq</code> object to be used in the <code>phyloseq</code> package.</p>
<p>First, we need to document and remove any failed samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>failed_samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"PHD1-SCH02-03_CO1"</span>,</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"PHD1-SCH02-05_CO1"</span>,</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"PHD1-SCH03-05rep_CO1"</span>,</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"PHD1-SCH04-03_CO1"</span>,</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"PHD1-SCH06-05_18S"</span>,</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"PHD1-SW04-01_CO1"</span>,</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"PHD1-SW06-02_18S"</span>,</span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"PHD2-CN02-03_18S"</span>,</span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"PHD2-CN05-02_18S"</span>,</span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"PHD2-NE01-09_CO1"</span>,</span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"PHD2-NE02-03_CO1"</span>,</span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"PHD2-NE04-03_18S"</span>,</span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"PHD2-NW03-01_CO1"</span>)</span>
<span id="cb61-14"><a href="#cb61-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-15"><a href="#cb61-15" aria-hidden="true" tabindex="-1"></a>eDNA_long_data_removefailed <span class="ot">&lt;-</span> <span class="fu">filter</span>(eDNA_long_data_final, <span class="sc">!</span>fullID <span class="sc">%in%</span> failed_samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="create-tax_table" class="level3">
<h3 class="anchored" data-anchor-id="create-tax_table">Create tax_table()</h3>
<p>We need to extract all the taxonomic information from our long data and do some tidying.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>taxa <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">subset</span>(eDNA_long_data_removefailed, <span class="at">select =</span> <span class="fu">c</span>(<span class="st">"kingdom"</span>, </span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"phylum"</span>, </span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"class"</span>, </span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"order"</span>, </span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"family"</span>, </span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"genus"</span>, </span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"valid_name"</span>,</span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"adult"</span>, <span class="co">#adult is functional group</span></span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"valid_AphiaID"</span>,</span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"rank"</span>))) </span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a>taxa<span class="sc">$</span>valid_name <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(taxa<span class="sc">$</span>valid_name)</span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a>taxa <span class="ot">&lt;-</span> taxa[<span class="fu">order</span>(taxa<span class="sc">$</span>valid_name),] <span class="co">#order alphabetically to match other ps elements</span></span>
<span id="cb62-14"><a href="#cb62-14" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(taxa) <span class="ot">&lt;-</span>taxa<span class="sc">$</span>valid_name</span>
<span id="cb62-15"><a href="#cb62-15" aria-hidden="true" tabindex="-1"></a>taxa <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(taxa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, save the taxa matrix as a tax_table() object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>phylo_tax <span class="ot">&lt;-</span> <span class="fu">tax_table</span>(taxa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="create-sample_data" class="level3">
<h3 class="anchored" data-anchor-id="create-sample_data">Create sample_data()</h3>
<p>Similarly, need to extract all the relavent sample information from our long data and do some tidying.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">subset</span>(eDNA_long_data_removefailed, <span class="at">select =</span> <span class="fu">c</span>(<span class="st">"fieldID"</span>,</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"projectID"</span>,</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"eventID"</span>,</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"type"</span>, </span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"verbatimLocality"</span>,</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"localityID"</span>,</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"shorePosition"</span>, </span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"fullID"</span>,</span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"year"</span>,</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"month"</span>,</span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"day"</span>,</span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"eventTime"</span>,</span>
<span id="cb64-13"><a href="#cb64-13" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"dna_concentration_CO1"</span>,</span>
<span id="cb64-14"><a href="#cb64-14" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"dna_concentration_18S"</span>,</span>
<span id="cb64-15"><a href="#cb64-15" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"negCheckLogical"</span>,</span>
<span id="cb64-16"><a href="#cb64-16" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"pairedRepLogical"</span>,</span>
<span id="cb64-17"><a href="#cb64-17" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"country"</span>,</span>
<span id="cb64-18"><a href="#cb64-18" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"controlCheck"</span>,</span>
<span id="cb64-19"><a href="#cb64-19" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"sampleType"</span>,</span>
<span id="cb64-20"><a href="#cb64-20" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"primer"</span>,</span>
<span id="cb64-21"><a href="#cb64-21" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"date_amplified_CO1"</span>,</span>
<span id="cb64-22"><a href="#cb64-22" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"date_amplified_18S"</span>,</span>
<span id="cb64-23"><a href="#cb64-23" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"batch_extraction"</span>,</span>
<span id="cb64-24"><a href="#cb64-24" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"batch_PCR_CO1"</span>,</span>
<span id="cb64-25"><a href="#cb64-25" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"plate_col_CO1"</span>,</span>
<span id="cb64-26"><a href="#cb64-26" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"plate_row_CO1"</span>,</span>
<span id="cb64-27"><a href="#cb64-27" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"batch_PCR_18S"</span>,</span>
<span id="cb64-28"><a href="#cb64-28" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"plate_col_18S"</span>,</span>
<span id="cb64-29"><a href="#cb64-29" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"plate_row_18S"</span>,</span>
<span id="cb64-30"><a href="#cb64-30" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"date_extraction"</span>,</span>
<span id="cb64-31"><a href="#cb64-31" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"exposure"</span>,</span>
<span id="cb64-32"><a href="#cb64-32" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"weather"</span>,</span>
<span id="cb64-33"><a href="#cb64-33" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"tide"</span>,</span>
<span id="cb64-34"><a href="#cb64-34" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"lowWater"</span>,</span>
<span id="cb64-35"><a href="#cb64-35" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"decimalLongitude"</span>,</span>
<span id="cb64-36"><a href="#cb64-36" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"decimalLatitude"</span>,</span>
<span id="cb64-37"><a href="#cb64-37" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"eventTime"</span>,</span>
<span id="cb64-38"><a href="#cb64-38" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"rockpoolarea.cm.2."</span>,</span>
<span id="cb64-39"><a href="#cb64-39" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"averageDepth..cm."</span>,</span>
<span id="cb64-40"><a href="#cb64-40" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"rockpoolVol.m.3."</span>,</span>
<span id="cb64-41"><a href="#cb64-41" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"averagePH"</span>,</span>
<span id="cb64-42"><a href="#cb64-42" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"averageTemp"</span></span>
<span id="cb64-43"><a href="#cb64-43" aria-hidden="true" tabindex="-1"></a>)))</span>
<span id="cb64-44"><a href="#cb64-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-45"><a href="#cb64-45" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> meta[<span class="fu">order</span>(meta<span class="sc">$</span>fullID),] <span class="co">#order rows alphabetically</span></span>
<span id="cb64-46"><a href="#cb64-46" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(meta) <span class="ot">&lt;-</span>meta<span class="sc">$</span>fullID</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, save the sample matrix as a sample_data() object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>phylo_samples <span class="ot">&lt;-</span> <span class="fu">sample_data</span>(meta)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="create-otu_table" class="level3">
<h3 class="anchored" data-anchor-id="create-otu_table">Create otu_table()</h3>
<p>Since our data are currently in long format, we need to recreate the ASV matrix through some data wrangling.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>species_reads_long <span class="ot">&lt;-</span> eDNA_long_data_removefailed <span class="sc">%&gt;%</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(ASV, reads, fullID) <span class="sc">%&gt;%</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(ASV, fullID), as.factor)) <span class="sc">%&gt;%</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>()</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(species_reads_long)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   712877 obs. of  3 variables:
 $ ASV   : Factor w/ 6321 levels "ASV_1","ASV_10",..: 3051 3391 3692 3892 4707 4838 4863 5068 5286 5749 ...
 $ reads : int  0 0 0 0 0 0 0 0 0 0 ...
 $ fullID: Factor w/ 655 levels "PHD1-NCbead_CO1",..: 6 6 6 6 6 6 6 6 6 6 ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(<span class="fu">is.na</span>(species_reads_long)) <span class="co">#check for NAs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>integer(0)</code></pre>
</div>
</div>
<p>Turning our long data into a wide format takes a long time due to the size of the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>species_reads_wide <span class="ot">&lt;-</span> species_reads_long <span class="sc">%&gt;%</span> </span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> fullID, </span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">values_from =</span> reads, </span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">values_fill =</span> <span class="fu">list</span>(<span class="at">reads =</span> <span class="dv">0</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A bit of formatting is needed for joining elements together later on. We can see that we have 6321 ASVs and 655 samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co">#order names alphabetically</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>species_reads_wide<span class="ot">&lt;-</span> species_reads_wide[<span class="fu">order</span>(species_reads_wide<span class="sc">$</span>ASV),] <span class="co">#order rows alphabetically</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>species_reads_wide <span class="ot">&lt;-</span> species_reads_wide[,<span class="fu">order</span>(<span class="fu">colnames</span>(species_reads_wide))] <span class="co">#order cols alphabetically</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a><span class="co">#update row names to ASVs</span></span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(species_reads_wide) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a>species_reads_wide <span class="ot">&lt;-</span> species_reads_wide <span class="sc">%&gt;%</span> <span class="fu">column_to_rownames</span>(<span class="at">var=</span><span class="st">"ASV"</span>)</span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(species_reads_wide)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6321  655</code></pre>
</div>
</div>
<p>We can also repeat the last few steps but group by <code>valid_names</code> to get a taxa-level <code>otu_table</code> . In my further analyses, taxa-level detection are appropriate and so we will move forward using this <code>otu_table</code>.</p>
<p>We can see that we have 1780 ASVs and 655 samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: format-otu-table-long-taxa</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>species_reads_long_taxa <span class="ot">&lt;-</span> eDNA_long_data_removefailed <span class="sc">%&gt;%</span></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(ASV, valid_name, reads, fullID) <span class="sc">%&gt;%</span></span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(ASV, valid_name, fullID), as.factor)) <span class="sc">%&gt;%</span></span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>()</span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a>species_reads_long_taxa_grouped <span class="ot">&lt;-</span> species_reads_long_taxa <span class="sc">%&gt;%</span> </span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(valid_name, fullID) <span class="sc">%&gt;%</span> </span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">reads =</span> <span class="fu">sum</span>(reads, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">.groups =</span> <span class="st">"drop"</span>)</span>
<span id="cb73-11"><a href="#cb73-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-12"><a href="#cb73-12" aria-hidden="true" tabindex="-1"></a>species_reads_wide_taxa <span class="ot">&lt;-</span> species_reads_long_taxa_grouped <span class="sc">%&gt;%</span> </span>
<span id="cb73-13"><a href="#cb73-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> fullID, </span>
<span id="cb73-14"><a href="#cb73-14" aria-hidden="true" tabindex="-1"></a>              <span class="at">values_from =</span> reads, </span>
<span id="cb73-15"><a href="#cb73-15" aria-hidden="true" tabindex="-1"></a>              <span class="at">values_fill =</span> <span class="fu">list</span>(<span class="at">reads =</span> <span class="dv">0</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb73-16"><a href="#cb73-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>()</span>
<span id="cb73-17"><a href="#cb73-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-18"><a href="#cb73-18" aria-hidden="true" tabindex="-1"></a><span class="co">#order names alphabetically</span></span>
<span id="cb73-19"><a href="#cb73-19" aria-hidden="true" tabindex="-1"></a>species_reads_wide_taxa<span class="ot">&lt;-</span> species_reads_wide_taxa[<span class="fu">order</span>(species_reads_wide_taxa<span class="sc">$</span>valid_name),] <span class="co">#order rows alphabetically</span></span>
<span id="cb73-20"><a href="#cb73-20" aria-hidden="true" tabindex="-1"></a>species_reads_wide_taxa <span class="ot">&lt;-</span> species_reads_wide_taxa[,<span class="fu">order</span>(<span class="fu">colnames</span>(species_reads_wide_taxa))] <span class="co">#order cols alphabetically</span></span>
<span id="cb73-21"><a href="#cb73-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-22"><a href="#cb73-22" aria-hidden="true" tabindex="-1"></a><span class="co">#update row names to ASVs</span></span>
<span id="cb73-23"><a href="#cb73-23" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(species_reads_wide_taxa) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb73-24"><a href="#cb73-24" aria-hidden="true" tabindex="-1"></a>species_reads_wide_taxa <span class="ot">&lt;-</span> species_reads_wide_taxa <span class="sc">%&gt;%</span> <span class="fu">column_to_rownames</span>(<span class="at">var=</span><span class="st">"valid_name"</span>)</span>
<span id="cb73-25"><a href="#cb73-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-26"><a href="#cb73-26" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(species_reads_wide_taxa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1780  655</code></pre>
</div>
</div>
<p>Now let’s create our <code>otu_table</code> elements - one at a ASV level (in case we need it later on) and one at a taxa-level (which we’ll use for compiling the <code>phyloseq</code> object).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>phylo_asv <span class="ot">&lt;-</span> <span class="fu">otu_table</span>(species_reads_wide, <span class="at">taxa_are_rows=</span><span class="cn">TRUE</span>)</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>phylo_asv_taxa <span class="ot">&lt;-</span> <span class="fu">otu_table</span>(species_reads_wide_taxa, <span class="at">taxa_are_rows=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It’s useful to do some checks befor we compile our taxa-level <code>phyloseq</code> object to make sure everything aligns correctly. This is why we’ve performed the alphabetical ordering in the above steps.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if colnames in asv object match meta rownames</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>list1<span class="ot">&lt;-</span> <span class="fu">colnames</span>(phylo_asv_taxa)</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>list2 <span class="ot">&lt;-</span> <span class="fu">rownames</span>(phylo_samples)</span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(list1 <span class="sc">!=</span> list2) <span class="co">#should be NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>integer(0)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if species names match in asv object and taxa object</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>list1<span class="ot">&lt;-</span> <span class="fu">rownames</span>(phylo_asv_taxa)</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>list2 <span class="ot">&lt;-</span> <span class="fu">rownames</span>(phylo_tax)</span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(list1 <span class="sc">!=</span> list2) <span class="co">#should be NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>integer(0)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>phylo_eDNA <span class="ot">&lt;-</span> <span class="fu">phyloseq</span>(phylo_asv_taxa, phylo_tax, phylo_samples)</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove taxa wihout occurances from the data</span></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>phylo_eDNA <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(phylo_eDNA, <span class="fu">taxa_sums</span>(phylo_eDNA) <span class="sc">&gt;</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s take a look at what the <code>phyloseq</code> object looks like.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sample_names</span>(phylo_eDNA)) <span class="co">#check sample names</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "PHD1-NCbead_CO1"   "PHD1-NCrep_18S"    "PHD1-NCrep_CO1"   
[4] "PHD1-PCrep_18S"    "PHD1-PCrep_CO1"    "PHD1-SCH01-01_18S"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rank_names</span>(phylo_eDNA)) <span class="co">#check taxonomic information</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "kingdom" "phylum"  "class"   "order"   "family"  "genus"  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sample_variables</span>(phylo_eDNA)) <span class="co">#check meta data </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "fieldID"          "projectID"        "eventID"          "type"            
[5] "verbatimLocality" "localityID"      </code></pre>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>phylo_eDNA</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 1725 taxa and 655 samples ]
sample_data() Sample Data:       [ 655 samples by 42 sample variables ]
tax_table()   Taxonomy Table:    [ 1725 taxa by 10 taxonomic ranks ]</code></pre>
</div>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">summarize_phyloseq</span>(phylo_eDNA) <span class="co">#full summary</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Compositional = NO2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>1] Min. number of reads = 02] Max. number of reads = 3520013] Total number of reads = 257099254] Average number of reads = 39251.79389312985] Median number of reads = 245707] Sparsity = 0.96441287753076] Any OTU sum to 1 or less? NO8] Number of singletons = 09] Percent of OTUs that are singletons 
        (i.e. exactly one read detected across all samples)010] Number of sample variables are: 42fieldIDprojectIDeventIDtypeverbatimLocalitylocalityIDshorePositionfullIDyearmonthdayeventTimedna_concentration_CO1dna_concentration_18SnegCheckLogicalpairedRepLogicalcountrycontrolChecksampleTypeprimerdate_amplified_CO1date_amplified_18Sbatch_extractionbatch_PCR_CO1plate_col_CO1plate_row_CO1batch_PCR_18Splate_col_18Splate_row_18Sdate_extractionexposureweathertidelowWaterdecimalLongitudedecimalLatitudeeventTime.1rockpoolarea.cm.2.averageDepth..cm.rockpoolVol.m.3.averagePHaverageTemp2</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
[1] "1] Min. number of reads = 0"

[[2]]
[1] "2] Max. number of reads = 352001"

[[3]]
[1] "3] Total number of reads = 25709925"

[[4]]
[1] "4] Average number of reads = 39251.7938931298"

[[5]]
[1] "5] Median number of reads = 24570"

[[6]]
[1] "7] Sparsity = 0.9644128775307"

[[7]]
[1] "6] Any OTU sum to 1 or less? NO"

[[8]]
[1] "8] Number of singletons = 0"

[[9]]
[1] "9] Percent of OTUs that are singletons \n        (i.e. exactly one read detected across all samples)0"

[[10]]
[1] "10] Number of sample variables are: 42"

[[11]]
 [1] "fieldID"               "projectID"             "eventID"              
 [4] "type"                  "verbatimLocality"      "localityID"           
 [7] "shorePosition"         "fullID"                "year"                 
[10] "month"                 "day"                   "eventTime"            
[13] "dna_concentration_CO1" "dna_concentration_18S" "negCheckLogical"      
[16] "pairedRepLogical"      "country"               "controlCheck"         
[19] "sampleType"            "primer"                "date_amplified_CO1"   
[22] "date_amplified_18S"    "batch_extraction"      "batch_PCR_CO1"        
[25] "plate_col_CO1"         "plate_row_CO1"         "batch_PCR_18S"        
[28] "plate_col_18S"         "plate_row_18S"         "date_extraction"      
[31] "exposure"              "weather"               "tide"                 
[34] "lowWater"              "decimalLongitude"      "decimalLatitude"      
[37] "eventTime.1"           "rockpoolarea.cm.2."    "averageDepth..cm."    
[40] "rockpoolVol.m.3."      "averagePH"             "averageTemp"          </code></pre>
</div>
</div>
</section>
</section>
<section id="decontamination" class="level2">
<h2 class="anchored" data-anchor-id="decontamination">Decontamination</h2>
<section id="visualizing-contaminants" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-contaminants">Visualizing contaminants</h3>
<p>Let’s produce some figures to visualize potential contaminates present in our negative controls.</p>
<p>We need to subset the negative control samples first.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>phylo_neg_controls <span class="ot">=</span> <span class="fu">subset_samples</span>(phylo_eDNA, sampleType <span class="sc">==</span> <span class="st">"field-control"</span> <span class="sc">|</span> sampleType <span class="sc">==</span> <span class="st">"lab-negative-control"</span>) </span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Clean out taxa/ASV columns that are no longer present.</span></span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>phylo_neg_controls <span class="ot">&lt;-</span> <span class="fu">prune_taxa</span>(<span class="fu">taxa_sums</span>(phylo_neg_controls) <span class="sc">&gt;</span> <span class="dv">0</span>, phylo_neg_controls)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s check the distribution of taxa in the negative controls.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">sort</span>(<span class="fu">taxa_sums</span>(phylo_neg_controls), <span class="cn">TRUE</span>), <span class="at">type=</span><span class="st">"h"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="intertidal-eDNA-formatting-and-quality-control_files/figure-html/abundance-negatives-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It might be useful to view contaminates by primer. Let’s look at 18S first.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>phylo_neg_18S <span class="ot">&lt;-</span> <span class="fu">subset_samples</span>(phylo_neg_controls, primer <span class="sc">==</span> <span class="st">"18S"</span>)</span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>(contam_18S_plot<span class="ot">&lt;-</span> <span class="fu">plot_bar</span>(phylo_neg_18S, <span class="st">"fullID"</span>, <span class="at">fill =</span> <span class="st">"phylum"</span>) <span class="sc">+</span></span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(.<span class="sc">~</span> primer)<span class="sc">+</span></span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">'none'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="intertidal-eDNA-formatting-and-quality-control_files/figure-html/18S-negatives-abundance-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>And now for CO1.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>phylo_neg_CO1 <span class="ot">&lt;-</span> <span class="fu">subset_samples</span>(phylo_neg_controls, primer <span class="sc">==</span> <span class="st">"CO1"</span>)</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>(contam_CO1_plot<span class="ot">&lt;-</span> <span class="fu">plot_bar</span>(phylo_neg_CO1, <span class="st">"fullID"</span>, <span class="at">fill =</span> <span class="st">"phylum"</span>) <span class="sc">+</span></span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(.<span class="sc">~</span> primer) <span class="sc">+</span></span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">'none'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="intertidal-eDNA-formatting-and-quality-control_files/figure-html/CO1-negatives-abundance-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can also find taxa that are shared among a certain number of negative controls. As an example, we check which taxa are present in more than five samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>controls_shared_taxa <span class="ot">=</span> <span class="fu">filter_taxa</span>(phylo_neg_controls, </span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>                                   <span class="cf">function</span>(x) <span class="fu">sum</span>(x <span class="sc">&gt;=</span> <span class="dv">1</span>) <span class="sc">==</span> (<span class="dv">5</span>), <span class="cn">TRUE</span>)</span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_bar</span>(controls_shared_taxa, <span class="st">"fullID"</span>, <span class="at">fill =</span> <span class="st">"valid_name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="intertidal-eDNA-formatting-and-quality-control_files/figure-html/negatives-shared-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can also visualize our original PCR plates (if we have variables in our metadata which state the position of samples on the plates).</p>
<p>First, calculate overall read depth across samples and this to our sample data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Calculate read depths</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>sample_depths <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">readcount</span>(phylo_eDNA)</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>sample_depths_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(sample_depths)</span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>sample_depths_df <span class="ot">&lt;-</span> tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(sample_depths_df, <span class="st">"fullID"</span>)</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sample_depths_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             fullID sample_depths
1   PHD1-NCbead_CO1            48
2    PHD1-NCrep_18S          1205
3    PHD1-NCrep_CO1             0
4    PHD1-PCrep_18S             0
5    PHD1-PCrep_CO1             0
6 PHD1-SCH01-01_18S         43943</code></pre>
</div>
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add depths to sample data</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>sample_data <span class="ot">&lt;-</span> <span class="fu">sample_data</span>(phylo_eDNA)</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>sample_data[,<span class="st">"depth"</span>] <span class="ot">&lt;-</span> sample_depths</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s look at CO1 first.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter out NA values</span></span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>sample_data_CO1 <span class="ot">&lt;-</span> sample_data[<span class="sc">!</span><span class="fu">is.na</span>(sample_data<span class="sc">$</span>plate_row_CO1), ]</span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder levels of plate_row_CO1</span></span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a>sample_data_CO1<span class="sc">$</span>plate_row_CO1 <span class="ot">&lt;-</span> <span class="fu">factor</span>(sample_data_CO1<span class="sc">$</span>plate_row_CO1, </span>
<span id="cb101-6"><a href="#cb101-6" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">levels =</span> <span class="fu">rev</span>(LETTERS[<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]))</span>
<span id="cb101-7"><a href="#cb101-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-8"><a href="#cb101-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Relabel PCR plates</span></span>
<span id="cb101-9"><a href="#cb101-9" aria-hidden="true" tabindex="-1"></a>sample_data_CO1<span class="sc">$</span>batch_PCR_CO1 <span class="ot">&lt;-</span> <span class="fu">recode</span>(sample_data_CO1<span class="sc">$</span>batch_PCR_CO1,</span>
<span id="cb101-10"><a href="#cb101-10" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"SW-CO1"</span>,</span>
<span id="cb101-11"><a href="#cb101-11" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"2"</span> <span class="ot">=</span> <span class="st">"SCH-CO1"</span>,</span>
<span id="cb101-12"><a href="#cb101-12" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"3"</span> <span class="ot">=</span> <span class="st">"SW-18S"</span>,</span>
<span id="cb101-13"><a href="#cb101-13" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"4"</span> <span class="ot">=</span> <span class="st">"SCH-18S"</span>,</span>
<span id="cb101-14"><a href="#cb101-14" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"5"</span> <span class="ot">=</span> <span class="st">"NW-18S"</span>,</span>
<span id="cb101-15"><a href="#cb101-15" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"6"</span> <span class="ot">=</span> <span class="st">"NE-18S"</span>,</span>
<span id="cb101-16"><a href="#cb101-16" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"7"</span> <span class="ot">=</span> <span class="st">"CN-18S"</span>,</span>
<span id="cb101-17"><a href="#cb101-17" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"8"</span> <span class="ot">=</span> <span class="st">"NW-CO1"</span>,</span>
<span id="cb101-18"><a href="#cb101-18" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"9"</span> <span class="ot">=</span> <span class="st">"NE-CO1"</span>,</span>
<span id="cb101-19"><a href="#cb101-19" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"10"</span> <span class="ot">=</span> <span class="st">"CN-CO1"</span>)</span>
<span id="cb101-20"><a href="#cb101-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-21"><a href="#cb101-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot with facet_wrap and circles representing concentrations - CO1 first</span></span>
<span id="cb101-22"><a href="#cb101-22" aria-hidden="true" tabindex="-1"></a>pcr_plate_CO1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(sample_data_CO1, <span class="fu">aes</span>(<span class="at">x =</span> plate_col_CO1, <span class="at">y =</span> plate_row_CO1)) <span class="sc">+</span></span>
<span id="cb101-23"><a href="#cb101-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">fill =</span> sampleType), <span class="at">color =</span> <span class="st">"white"</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb101-24"><a href="#cb101-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">size =</span> depth), <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">fill =</span> <span class="st">"black"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span> <span class="co"># Fill circles with black</span></span>
<span id="cb101-25"><a href="#cb101-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"field-control"</span> <span class="ot">=</span> <span class="st">"brown"</span>, </span>
<span id="cb101-26"><a href="#cb101-26" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"lab-extraction-control"</span> <span class="ot">=</span> <span class="st">"red"</span>,</span>
<span id="cb101-27"><a href="#cb101-27" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"lab-negative-control"</span> <span class="ot">=</span> <span class="st">"pink"</span>, </span>
<span id="cb101-28"><a href="#cb101-28" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"lab-positive-control"</span> <span class="ot">=</span> <span class="st">"seagreen"</span>,</span>
<span id="cb101-29"><a href="#cb101-29" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"sample"</span> <span class="ot">=</span><span class="st">"white"</span>, </span>
<span id="cb101-30"><a href="#cb101-30" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"plate-replicate"</span> <span class="ot">=</span> <span class="st">"lightblue"</span>, </span>
<span id="cb101-31"><a href="#cb101-31" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"failed-repeat"</span> <span class="ot">=</span> <span class="st">"white"</span>, </span>
<span id="cb101-32"><a href="#cb101-32" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"run-replicate"</span> <span class="ot">=</span> <span class="st">"lightblue4"</span>)) <span class="sc">+</span> </span>
<span id="cb101-33"><a href="#cb101-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_size_continuous</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">10</span>)) <span class="sc">+</span>  <span class="co"># Adjust circle sizes as needed</span></span>
<span id="cb101-34"><a href="#cb101-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Virtual PCR Plates"</span>,</span>
<span id="cb101-35"><a href="#cb101-35" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="cn">NULL</span>,</span>
<span id="cb101-36"><a href="#cb101-36" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="cn">NULL</span>,</span>
<span id="cb101-37"><a href="#cb101-37" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="st">"Control Type"</span>,</span>
<span id="cb101-38"><a href="#cb101-38" aria-hidden="true" tabindex="-1"></a>       <span class="at">size =</span> <span class="st">"# of reads per sample"</span>) <span class="sc">+</span></span>
<span id="cb101-39"><a href="#cb101-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb101-40"><a href="#cb101-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>) <span class="sc">+</span> <span class="co"># Adjust legend position if needed</span></span>
<span id="cb101-41"><a href="#cb101-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> batch_PCR_CO1, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span> <span class="co"># Facet by plate variable, 2 plates per row</span></span>
<span id="cb101-42"><a href="#cb101-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>)  <span class="co"># Discrete x-axis from 1 to 12</span></span>
<span id="cb101-43"><a href="#cb101-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-44"><a href="#cb101-44" aria-hidden="true" tabindex="-1"></a>pcr_plate_CO1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="intertidal-eDNA-formatting-and-quality-control_files/figure-html/PCR-plate-CO1-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>And now for 18S.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter out NA values</span></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>sample_data_18S <span class="ot">&lt;-</span> sample_data[<span class="sc">!</span><span class="fu">is.na</span>(sample_data<span class="sc">$</span>plate_row_18S), ]</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder levels of plate_row_18S</span></span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a>sample_data_18S<span class="sc">$</span>plate_row_18S <span class="ot">&lt;-</span> <span class="fu">factor</span>(sample_data_18S<span class="sc">$</span>plate_row_18S, <span class="at">levels =</span> <span class="fu">rev</span>(LETTERS[<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]))</span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-7"><a href="#cb102-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Relabel PCR plates</span></span>
<span id="cb102-8"><a href="#cb102-8" aria-hidden="true" tabindex="-1"></a>sample_data_18S<span class="sc">$</span>batch_PCR_18S <span class="ot">&lt;-</span> <span class="fu">recode</span>(sample_data_18S<span class="sc">$</span>batch_PCR_18S,</span>
<span id="cb102-9"><a href="#cb102-9" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"SW-CO1"</span>,</span>
<span id="cb102-10"><a href="#cb102-10" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"2"</span> <span class="ot">=</span> <span class="st">"SCH-CO1"</span>,</span>
<span id="cb102-11"><a href="#cb102-11" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"3"</span> <span class="ot">=</span> <span class="st">"SW-18S"</span>,</span>
<span id="cb102-12"><a href="#cb102-12" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"4"</span> <span class="ot">=</span> <span class="st">"SCH-18S"</span>,</span>
<span id="cb102-13"><a href="#cb102-13" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"5"</span> <span class="ot">=</span> <span class="st">"NW-18S"</span>,</span>
<span id="cb102-14"><a href="#cb102-14" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"6"</span> <span class="ot">=</span> <span class="st">"NE-18S"</span>,</span>
<span id="cb102-15"><a href="#cb102-15" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"7"</span> <span class="ot">=</span> <span class="st">"CN-18S"</span>,</span>
<span id="cb102-16"><a href="#cb102-16" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"8"</span> <span class="ot">=</span> <span class="st">"NW-CO1"</span>,</span>
<span id="cb102-17"><a href="#cb102-17" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"9"</span> <span class="ot">=</span> <span class="st">"NE-CO1"</span>,</span>
<span id="cb102-18"><a href="#cb102-18" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"10"</span> <span class="ot">=</span> <span class="st">"CN-CO1"</span>)</span>
<span id="cb102-19"><a href="#cb102-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-20"><a href="#cb102-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot with facet_wrap and circles representing concentrations - 18S first</span></span>
<span id="cb102-21"><a href="#cb102-21" aria-hidden="true" tabindex="-1"></a>pcr_plate_18S <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(sample_data_18S, <span class="fu">aes</span>(<span class="at">x =</span> plate_col_18S, <span class="at">y =</span> plate_row_18S)) <span class="sc">+</span></span>
<span id="cb102-22"><a href="#cb102-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">fill =</span> sampleType), <span class="at">color =</span> <span class="st">"white"</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb102-23"><a href="#cb102-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">size =</span> depth), <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">fill =</span> <span class="st">"black"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span> <span class="co"># Fill circles with black</span></span>
<span id="cb102-24"><a href="#cb102-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"field-control"</span> <span class="ot">=</span> <span class="st">"brown"</span>, </span>
<span id="cb102-25"><a href="#cb102-25" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"lab-extraction-control"</span> <span class="ot">=</span> <span class="st">"red"</span>,</span>
<span id="cb102-26"><a href="#cb102-26" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"lab-negative-control"</span> <span class="ot">=</span> <span class="st">"pink"</span>, </span>
<span id="cb102-27"><a href="#cb102-27" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"lab-positive-control"</span> <span class="ot">=</span> <span class="st">"seagreen"</span>,</span>
<span id="cb102-28"><a href="#cb102-28" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"sample"</span> <span class="ot">=</span><span class="st">"white"</span>, </span>
<span id="cb102-29"><a href="#cb102-29" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"plate-replicate"</span> <span class="ot">=</span> <span class="st">"lightblue"</span>, </span>
<span id="cb102-30"><a href="#cb102-30" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"failed-repeat"</span> <span class="ot">=</span> <span class="st">"white"</span>, </span>
<span id="cb102-31"><a href="#cb102-31" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"run-replicate"</span> <span class="ot">=</span> <span class="st">"lightblue4"</span>)) <span class="sc">+</span> </span>
<span id="cb102-32"><a href="#cb102-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_size_continuous</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">10</span>)) <span class="sc">+</span>  <span class="co"># Adjust circle sizes as needed</span></span>
<span id="cb102-33"><a href="#cb102-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Virtual PCR Plates"</span>,</span>
<span id="cb102-34"><a href="#cb102-34" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="cn">NULL</span>,</span>
<span id="cb102-35"><a href="#cb102-35" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="cn">NULL</span>,</span>
<span id="cb102-36"><a href="#cb102-36" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="st">"Control Type"</span>,</span>
<span id="cb102-37"><a href="#cb102-37" aria-hidden="true" tabindex="-1"></a>       <span class="at">size =</span> <span class="st">"# of reads per sample"</span>) <span class="sc">+</span></span>
<span id="cb102-38"><a href="#cb102-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb102-39"><a href="#cb102-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>) <span class="sc">+</span> <span class="co"># Adjust legend position if needed</span></span>
<span id="cb102-40"><a href="#cb102-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> batch_PCR_18S, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span> <span class="co"># Facet by plate variable, 2 plates per row</span></span>
<span id="cb102-41"><a href="#cb102-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>)  <span class="co"># Discrete x-axis from 1 to 12</span></span>
<span id="cb102-42"><a href="#cb102-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-43"><a href="#cb102-43" aria-hidden="true" tabindex="-1"></a>pcr_plate_18S</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="intertidal-eDNA-formatting-and-quality-control_files/figure-html/PCR-plate-18S-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="removing-contaminants" class="level3">
<h3 class="anchored" data-anchor-id="removing-contaminants">Removing contaminants</h3>
<p>We can see from our plots we have some contamination. We deal with this next.</p>
<p>There are multiple ways to deal with contamination. Here, we take the approach that if the abundance of a taxa in control sample is larger than abundance in main sample, it will be treated as contaminant and be removed.</p>
<p>We’ll remove taxa in stages: from sites (using the field controls), extraction batch (using the extraction controls) then PCR batch (using the PCR controls). This prevents unnecessary blanket removal across the whole data set.</p>
<p>Let’s set-up the site level decontamination function called <code>remove_field_control_taxa_abund</code> first.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>remove_field_control_taxa_abund <span class="ot">&lt;-</span> <span class="cf">function</span>(phylo_object) {</span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extract sample data</span></span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a>  sample_data <span class="ot">&lt;-</span> <span class="fu">sample_data</span>(phylo_object)</span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get unique localityIDs</span></span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a>  unique_localityIDs <span class="ot">&lt;-</span> <span class="fu">unique</span>(sample_data<span class="sc">$</span>localityID[<span class="sc">!</span><span class="fu">is.na</span>(sample_data<span class="sc">$</span>localityID)])</span>
<span id="cb103-8"><a href="#cb103-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-9"><a href="#cb103-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check if unique_localityIDs is empty</span></span>
<span id="cb103-10"><a href="#cb103-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(unique_localityIDs) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb103-11"><a href="#cb103-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"No localityIDs found in sample data."</span>)</span>
<span id="cb103-12"><a href="#cb103-12" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb103-13"><a href="#cb103-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-14"><a href="#cb103-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize variable to count OTUs removed</span></span>
<span id="cb103-15"><a href="#cb103-15" aria-hidden="true" tabindex="-1"></a>  otus_removed <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">localityID =</span> <span class="fu">character</span>(), <span class="at">OTU =</span> <span class="fu">character</span>(), <span class="at">reads_removed =</span> <span class="fu">integer</span>(), <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb103-16"><a href="#cb103-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-17"><a href="#cb103-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize filtered OTU table</span></span>
<span id="cb103-18"><a href="#cb103-18" aria-hidden="true" tabindex="-1"></a>  otu_table_filtered <span class="ot">&lt;-</span> <span class="fu">otu_table</span>(phylo_object)</span>
<span id="cb103-19"><a href="#cb103-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-20"><a href="#cb103-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check if the OTU table is empty</span></span>
<span id="cb103-21"><a href="#cb103-21" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">dim</span>(otu_table_filtered)[<span class="dv">1</span>] <span class="sc">==</span> <span class="dv">0</span> <span class="sc">||</span> <span class="fu">dim</span>(otu_table_filtered)[<span class="dv">2</span>] <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb103-22"><a href="#cb103-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"OTU abundance data is empty."</span>)</span>
<span id="cb103-23"><a href="#cb103-23" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb103-24"><a href="#cb103-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-25"><a href="#cb103-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"Original dimensions of otu_table_filtered:"</span>)</span>
<span id="cb103-26"><a href="#cb103-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">dim</span>(otu_table_filtered))</span>
<span id="cb103-27"><a href="#cb103-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-28"><a href="#cb103-28" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize list to store removed OTUs for each sample</span></span>
<span id="cb103-29"><a href="#cb103-29" aria-hidden="true" tabindex="-1"></a>  otus_removed_reads <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb103-30"><a href="#cb103-30" aria-hidden="true" tabindex="-1"></a>  otus_removed_reads_total <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb103-31"><a href="#cb103-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-32"><a href="#cb103-32" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Loop through each localityID</span></span>
<span id="cb103-33"><a href="#cb103-33" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (locality <span class="cf">in</span> unique_localityIDs) {</span>
<span id="cb103-34"><a href="#cb103-34" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-35"><a href="#cb103-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get field control sample fullIDs for the current locality</span></span>
<span id="cb103-36"><a href="#cb103-36" aria-hidden="true" tabindex="-1"></a>    field_control_samples <span class="ot">&lt;-</span> <span class="fu">unique</span>(sample_data<span class="sc">$</span>fullID[sample_data<span class="sc">$</span>localityID <span class="sc">==</span> locality <span class="sc">&amp;</span> <span class="fu">grepl</span>(<span class="st">"-C_CO1$|-C_18S$"</span>, sample_data<span class="sc">$</span>fullID)])</span>
<span id="cb103-37"><a href="#cb103-37" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-38"><a href="#cb103-38" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If there are more than 2 field control samples for the current locality, print a message and skip</span></span>
<span id="cb103-39"><a href="#cb103-39" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(field_control_samples) <span class="sc">&gt;</span> <span class="dv">2</span>) {</span>
<span id="cb103-40"><a href="#cb103-40" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"More than 1 field control samples found for site:"</span>, locality, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb103-41"><a href="#cb103-41" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb103-42"><a href="#cb103-42" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb103-43"><a href="#cb103-43" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-44"><a href="#cb103-44" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If there are no field control samples for the current locality, print a message and skip</span></span>
<span id="cb103-45"><a href="#cb103-45" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(field_control_samples) <span class="sc">&lt;</span> <span class="dv">2</span>) {</span>
<span id="cb103-46"><a href="#cb103-46" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No field control samples found for site:"</span>, locality, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb103-47"><a href="#cb103-47" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb103-48"><a href="#cb103-48" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb103-49"><a href="#cb103-49" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-50"><a href="#cb103-50" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get indices of field control and samples</span></span>
<span id="cb103-51"><a href="#cb103-51" aria-hidden="true" tabindex="-1"></a>    control_indices <span class="ot">&lt;-</span> <span class="fu">which</span>(sample_data<span class="sc">$</span>fullID <span class="sc">%in%</span> field_control_samples)</span>
<span id="cb103-52"><a href="#cb103-52" aria-hidden="true" tabindex="-1"></a>    sample_indices <span class="ot">&lt;-</span> <span class="fu">which</span>(sample_data<span class="sc">$</span>localityID <span class="sc">==</span> locality <span class="sc">&amp;</span> <span class="sc">!</span>sample_data<span class="sc">$</span>fullID <span class="sc">%in%</span> field_control_samples)</span>
<span id="cb103-53"><a href="#cb103-53" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-54"><a href="#cb103-54" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If no control indices found</span></span>
<span id="cb103-55"><a href="#cb103-55" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(control_indices) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb103-56"><a href="#cb103-56" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No control indices found for site:"</span>, locality, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb103-57"><a href="#cb103-57" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb103-58"><a href="#cb103-58" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb103-59"><a href="#cb103-59" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-60"><a href="#cb103-60" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If no sample indices found</span></span>
<span id="cb103-61"><a href="#cb103-61" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(sample_indices) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb103-62"><a href="#cb103-62" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No sample indices found for site:"</span>, locality, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb103-63"><a href="#cb103-63" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb103-64"><a href="#cb103-64" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb103-65"><a href="#cb103-65" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-66"><a href="#cb103-66" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Calculate mean abundance of each OTU in the control and samples</span></span>
<span id="cb103-67"><a href="#cb103-67" aria-hidden="true" tabindex="-1"></a>    control_abundance <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(otu_table_filtered[, control_indices])</span>
<span id="cb103-68"><a href="#cb103-68" aria-hidden="true" tabindex="-1"></a>    sample_abundance <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(otu_table_filtered[, sample_indices])</span>
<span id="cb103-69"><a href="#cb103-69" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-70"><a href="#cb103-70" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Identify potential contaminants based on higher abundance in control than samples</span></span>
<span id="cb103-71"><a href="#cb103-71" aria-hidden="true" tabindex="-1"></a>    contaminants <span class="ot">&lt;-</span> <span class="fu">rownames</span>(otu_table_filtered)[control_abundance <span class="sc">&gt;</span> sample_abundance]</span>
<span id="cb103-72"><a href="#cb103-72" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-73"><a href="#cb103-73" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(contaminants) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb103-74"><a href="#cb103-74" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="fu">length</span>(contaminants), <span class="st">"contaminant taxa found for site:"</span>, locality, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb103-75"><a href="#cb103-75" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb103-76"><a href="#cb103-76" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No contaminants found for site:"</span>, locality, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb103-77"><a href="#cb103-77" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb103-78"><a href="#cb103-78" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb103-79"><a href="#cb103-79" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-80"><a href="#cb103-80" aria-hidden="true" tabindex="-1"></a>    <span class="co">#Get original number of reads in samples for taxa with above minReads reads in controls</span></span>
<span id="cb103-81"><a href="#cb103-81" aria-hidden="true" tabindex="-1"></a>    <span class="co">#bind the previous output</span></span>
<span id="cb103-82"><a href="#cb103-82" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(otus_removed_reads) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb103-83"><a href="#cb103-83" aria-hidden="true" tabindex="-1"></a>      otus_removed_reads_total <span class="ot">&lt;-</span> <span class="fu">rbind</span>(otus_removed_reads, otus_removed_reads_total)</span>
<span id="cb103-84"><a href="#cb103-84" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb103-85"><a href="#cb103-85" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-86"><a href="#cb103-86" aria-hidden="true" tabindex="-1"></a>    <span class="co">#get reads</span></span>
<span id="cb103-87"><a href="#cb103-87" aria-hidden="true" tabindex="-1"></a>    otus_removed_reads <span class="ot">&lt;-</span> otu_table_filtered[contaminants,  sample_indices] <span class="sc">%&gt;%</span> reshape2<span class="sc">::</span><span class="fu">melt</span>()</span>
<span id="cb103-88"><a href="#cb103-88" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(otus_removed_reads) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"taxa"</span>, <span class="st">"fullID"</span>, <span class="st">"reads_removed"</span>)</span>
<span id="cb103-89"><a href="#cb103-89" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-90"><a href="#cb103-90" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Set read count of OTUs in real samples to zero if reads in field control samples exceed 100</span></span>
<span id="cb103-91"><a href="#cb103-91" aria-hidden="true" tabindex="-1"></a>    otu_table_filtered[contaminants, sample_indices] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb103-92"><a href="#cb103-92" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-93"><a href="#cb103-93" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb103-94"><a href="#cb103-94" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-95"><a href="#cb103-95" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print OTUs removed</span></span>
<span id="cb103-96"><a href="#cb103-96" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Reads removed:"</span>, <span class="fu">sum</span>(otus_removed_reads_total<span class="sc">$</span>reads_removed),<span class="st">"; Samples filtered:"</span>, <span class="fu">length</span>(<span class="fu">unique</span>(otus_removed_reads_total<span class="sc">$</span>fullID)), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb103-97"><a href="#cb103-97" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-98"><a href="#cb103-98" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print new dimensions</span></span>
<span id="cb103-99"><a href="#cb103-99" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"New dimensions of otu_table_filtered:"</span>)</span>
<span id="cb103-100"><a href="#cb103-100" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">dim</span>(otu_table_filtered))</span>
<span id="cb103-101"><a href="#cb103-101" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-102"><a href="#cb103-102" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove controls from the dataset</span></span>
<span id="cb103-103"><a href="#cb103-103" aria-hidden="true" tabindex="-1"></a>  control_samples<span class="ot">&lt;-</span> sample_data<span class="sc">$</span>fullID[<span class="fu">grepl</span>(<span class="st">"-C_CO1$|-C_18S$"</span>, sample_data<span class="sc">$</span>fullID)] <span class="co"># Build list of names of control samples</span></span>
<span id="cb103-104"><a href="#cb103-104" aria-hidden="true" tabindex="-1"></a>  otu_table_filtered <span class="ot">&lt;-</span> otu_table_filtered[, <span class="sc">-</span><span class="fu">which</span>(<span class="fu">colnames</span>(otu_table_filtered) <span class="sc">%in%</span> control_samples)] <span class="co"># Remove control samples from otu_table</span></span>
<span id="cb103-105"><a href="#cb103-105" aria-hidden="true" tabindex="-1"></a>  sample_data_filtered <span class="ot">&lt;-</span> sample_data[<span class="sc">!</span>sample_data<span class="sc">$</span>fullID <span class="sc">%in%</span> control_samples, ] <span class="co"># Remove control samples from sample_data</span></span>
<span id="cb103-106"><a href="#cb103-106" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-107"><a href="#cb103-107" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Return the modified phyloseq object</span></span>
<span id="cb103-108"><a href="#cb103-108" aria-hidden="true" tabindex="-1"></a>  modified_phylo <span class="ot">&lt;-</span> <span class="fu">phyloseq</span>(otu_table_filtered, <span class="fu">tax_table</span>(phylo_object), sample_data_filtered)</span>
<span id="cb103-109"><a href="#cb103-109" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-110"><a href="#cb103-110" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove taxa with zero reads removed in otus_removed_reads_total</span></span>
<span id="cb103-111"><a href="#cb103-111" aria-hidden="true" tabindex="-1"></a>  otus_removed_reads_total <span class="ot">&lt;-</span> otus_removed_reads_total <span class="sc">%&gt;%</span> <span class="fu">subset</span>(otus_removed_reads_total<span class="sc">$</span>reads_removed <span class="sc">!=</span> <span class="dv">0</span>)</span>
<span id="cb103-112"><a href="#cb103-112" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-113"><a href="#cb103-113" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Return the modified phyloseq object and otus_removed dataframe</span></span>
<span id="cb103-114"><a href="#cb103-114" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(modified_phylo, otus_removed_reads_total))</span>
<span id="cb103-115"><a href="#cb103-115" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, the function for extraction level decontamination called <code>remove_extract_control_taxa_abund</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>remove_extract_control_taxa_abund <span class="ot">&lt;-</span> <span class="cf">function</span>(phylo_object) {</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extract sample data</span></span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>  sample_data <span class="ot">&lt;-</span> <span class="fu">sample_data</span>(phylo_object)</span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get unique batches</span></span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a>  unique_batches <span class="ot">&lt;-</span> <span class="fu">unique</span>(sample_data<span class="sc">$</span>batch_extraction[<span class="sc">!</span><span class="fu">is.na</span>(sample_data<span class="sc">$</span>batch_extraction)])</span>
<span id="cb104-8"><a href="#cb104-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-9"><a href="#cb104-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check if unique_localityIDs is empty</span></span>
<span id="cb104-10"><a href="#cb104-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(unique_batches) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb104-11"><a href="#cb104-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"No DNA extraction batches found in sample data."</span>)</span>
<span id="cb104-12"><a href="#cb104-12" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb104-13"><a href="#cb104-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-14"><a href="#cb104-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize variable to count OTUs removed</span></span>
<span id="cb104-15"><a href="#cb104-15" aria-hidden="true" tabindex="-1"></a>  otus_removed <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">batch =</span> <span class="fu">character</span>(), <span class="at">OTU =</span> <span class="fu">character</span>(), <span class="at">reads_removed =</span> <span class="fu">integer</span>(), <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb104-16"><a href="#cb104-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-17"><a href="#cb104-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize filtered OTU table</span></span>
<span id="cb104-18"><a href="#cb104-18" aria-hidden="true" tabindex="-1"></a>  otu_table_filtered <span class="ot">&lt;-</span> <span class="fu">otu_table</span>(phylo_object)</span>
<span id="cb104-19"><a href="#cb104-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-20"><a href="#cb104-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check if the OTU table is empty</span></span>
<span id="cb104-21"><a href="#cb104-21" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">dim</span>(otu_table_filtered)[<span class="dv">1</span>] <span class="sc">==</span> <span class="dv">0</span> <span class="sc">||</span> <span class="fu">dim</span>(otu_table_filtered)[<span class="dv">2</span>] <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb104-22"><a href="#cb104-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"OTU abundance data is empty."</span>)</span>
<span id="cb104-23"><a href="#cb104-23" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb104-24"><a href="#cb104-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-25"><a href="#cb104-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"Original dimensions of otu_table_filtered:"</span>)</span>
<span id="cb104-26"><a href="#cb104-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">dim</span>(otu_table_filtered))</span>
<span id="cb104-27"><a href="#cb104-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-28"><a href="#cb104-28" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize list to store removed OTUs for each sample</span></span>
<span id="cb104-29"><a href="#cb104-29" aria-hidden="true" tabindex="-1"></a>  otus_removed_reads <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb104-30"><a href="#cb104-30" aria-hidden="true" tabindex="-1"></a>  otus_removed_reads_total <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb104-31"><a href="#cb104-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-32"><a href="#cb104-32" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Loop through each localityID</span></span>
<span id="cb104-33"><a href="#cb104-33" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (batch <span class="cf">in</span> unique_batches) {</span>
<span id="cb104-34"><a href="#cb104-34" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-35"><a href="#cb104-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get extract control sample for the current batch</span></span>
<span id="cb104-36"><a href="#cb104-36" aria-hidden="true" tabindex="-1"></a>    dna_extract_control_samples <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">sub</span>(<span class="st">"_[^_]*$"</span>, <span class="st">""</span>, sample_data<span class="sc">$</span>fullID[sample_data<span class="sc">$</span>sampleType <span class="sc">==</span> <span class="st">"lab-extraction-control"</span> </span>
<span id="cb104-37"><a href="#cb104-37" aria-hidden="true" tabindex="-1"></a>                                                                                <span class="sc">&amp;</span> sample_data<span class="sc">$</span>batch_extraction <span class="sc">==</span> batch]))</span>
<span id="cb104-38"><a href="#cb104-38" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-39"><a href="#cb104-39" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If there are more than 2 field control samples for the current locality, print a message and skip</span></span>
<span id="cb104-40"><a href="#cb104-40" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(dna_extract_control_samples) <span class="sc">&gt;</span> <span class="dv">2</span>) {</span>
<span id="cb104-41"><a href="#cb104-41" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"More than 1 DNA extract control sample found in batch:"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-42"><a href="#cb104-42" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb104-43"><a href="#cb104-43" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb104-44"><a href="#cb104-44" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-45"><a href="#cb104-45" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If there are no field control samples for the current locality, print a message and skip</span></span>
<span id="cb104-46"><a href="#cb104-46" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(dna_extract_control_samples) <span class="sc">&lt;</span> <span class="dv">1</span>) {</span>
<span id="cb104-47"><a href="#cb104-47" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No DNA extract control samples found in batch:"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-48"><a href="#cb104-48" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb104-49"><a href="#cb104-49" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb104-50"><a href="#cb104-50" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-51"><a href="#cb104-51" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get indices of dna extract control and samples</span></span>
<span id="cb104-52"><a href="#cb104-52" aria-hidden="true" tabindex="-1"></a>    control_indices <span class="ot">&lt;-</span> <span class="fu">which</span>(sample_data<span class="sc">$</span>fieldID <span class="sc">%in%</span> dna_extract_control_samples)</span>
<span id="cb104-53"><a href="#cb104-53" aria-hidden="true" tabindex="-1"></a>    sample_indices <span class="ot">&lt;-</span> <span class="fu">which</span>(sample_data<span class="sc">$</span>batch_extraction <span class="sc">==</span> batch <span class="sc">&amp;</span> <span class="sc">!</span>sample_data<span class="sc">$</span>fullID <span class="sc">%in%</span> dna_extract_control_samples)</span>
<span id="cb104-54"><a href="#cb104-54" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-55"><a href="#cb104-55" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If no control indices found</span></span>
<span id="cb104-56"><a href="#cb104-56" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(control_indices) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb104-57"><a href="#cb104-57" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No control indices found for batch:"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-58"><a href="#cb104-58" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb104-59"><a href="#cb104-59" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb104-60"><a href="#cb104-60" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-61"><a href="#cb104-61" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If no sample indices found</span></span>
<span id="cb104-62"><a href="#cb104-62" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(sample_indices) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb104-63"><a href="#cb104-63" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No sample indices found for batch:"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-64"><a href="#cb104-64" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb104-65"><a href="#cb104-65" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb104-66"><a href="#cb104-66" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-67"><a href="#cb104-67" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Calculate mean abundance of each OTU in the control and samples</span></span>
<span id="cb104-68"><a href="#cb104-68" aria-hidden="true" tabindex="-1"></a>    control_abundance <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(otu_table_filtered[, control_indices])</span>
<span id="cb104-69"><a href="#cb104-69" aria-hidden="true" tabindex="-1"></a>    sample_abundance <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(otu_table_filtered[, sample_indices])</span>
<span id="cb104-70"><a href="#cb104-70" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-71"><a href="#cb104-71" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Identify potential contaminants based on higher abundance in control than samples</span></span>
<span id="cb104-72"><a href="#cb104-72" aria-hidden="true" tabindex="-1"></a>    contaminants <span class="ot">&lt;-</span> <span class="fu">rownames</span>(otu_table_filtered)[control_abundance <span class="sc">&gt;</span> sample_abundance]</span>
<span id="cb104-73"><a href="#cb104-73" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-74"><a href="#cb104-74" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(contaminants) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb104-75"><a href="#cb104-75" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="fu">length</span>(contaminants), <span class="st">"contaminant taxa found for site:"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-76"><a href="#cb104-76" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb104-77"><a href="#cb104-77" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No contaminants found for site:"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-78"><a href="#cb104-78" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb104-79"><a href="#cb104-79" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb104-80"><a href="#cb104-80" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-81"><a href="#cb104-81" aria-hidden="true" tabindex="-1"></a>    <span class="co">#Get original number of reads in samples for taxa with above minReads reads in controls</span></span>
<span id="cb104-82"><a href="#cb104-82" aria-hidden="true" tabindex="-1"></a>    <span class="co">#bind the previous output</span></span>
<span id="cb104-83"><a href="#cb104-83" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(otus_removed_reads) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb104-84"><a href="#cb104-84" aria-hidden="true" tabindex="-1"></a>      otus_removed_reads_total <span class="ot">&lt;-</span> <span class="fu">rbind</span>(otus_removed_reads, otus_removed_reads_total)</span>
<span id="cb104-85"><a href="#cb104-85" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb104-86"><a href="#cb104-86" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-87"><a href="#cb104-87" aria-hidden="true" tabindex="-1"></a>    <span class="co">#get reads</span></span>
<span id="cb104-88"><a href="#cb104-88" aria-hidden="true" tabindex="-1"></a>    otus_removed_reads <span class="ot">&lt;-</span> otu_table_filtered[contaminants,  sample_indices] <span class="sc">%&gt;%</span> reshape2<span class="sc">::</span><span class="fu">melt</span>()</span>
<span id="cb104-89"><a href="#cb104-89" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(otus_removed_reads) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"taxa"</span>, <span class="st">"fullID"</span>, <span class="st">"reads_removed"</span>)</span>
<span id="cb104-90"><a href="#cb104-90" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-91"><a href="#cb104-91" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Set read count of OTUs in real samples to zero if reads in field control samples exceed 100</span></span>
<span id="cb104-92"><a href="#cb104-92" aria-hidden="true" tabindex="-1"></a>    otu_table_filtered[contaminants, sample_indices] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb104-93"><a href="#cb104-93" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-94"><a href="#cb104-94" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Remove the control samples from the phylo object</span></span>
<span id="cb104-95"><a href="#cb104-95" aria-hidden="true" tabindex="-1"></a>    otu_table_filtered <span class="ot">&lt;-</span> otu_table_filtered[<span class="sc">-</span>control_indices]</span>
<span id="cb104-96"><a href="#cb104-96" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-97"><a href="#cb104-97" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb104-98"><a href="#cb104-98" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-99"><a href="#cb104-99" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print OTUs removed</span></span>
<span id="cb104-100"><a href="#cb104-100" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Reads removed:"</span>, <span class="fu">sum</span>(otus_removed_reads_total<span class="sc">$</span>reads_removed),<span class="st">"; Samples filtered:"</span>, <span class="fu">length</span>(<span class="fu">unique</span>(otus_removed_reads_total<span class="sc">$</span>fullID)), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-101"><a href="#cb104-101" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-102"><a href="#cb104-102" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove controls from the dataset</span></span>
<span id="cb104-103"><a href="#cb104-103" aria-hidden="true" tabindex="-1"></a>  control_samples <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">sub</span>(<span class="st">"_[^_]*$"</span>, <span class="st">""</span>, sample_data<span class="sc">$</span>fullID[sample_data<span class="sc">$</span>sampleType <span class="sc">==</span> <span class="st">"lab-extraction-control"</span>])) <span class="co"># Build list of names of control samples</span></span>
<span id="cb104-104"><a href="#cb104-104" aria-hidden="true" tabindex="-1"></a>  control_samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">paste0</span>(control_samples, <span class="st">"_CO1"</span>), <span class="fu">paste0</span>(control_samples, <span class="st">"_18S"</span>))  <span class="co"># Create a list of control samples by appending "_CO1" and "_18S" to each ID</span></span>
<span id="cb104-105"><a href="#cb104-105" aria-hidden="true" tabindex="-1"></a>  otu_table_filtered <span class="ot">&lt;-</span> otu_table_filtered[, <span class="sc">-</span><span class="fu">which</span>(<span class="fu">colnames</span>(otu_table_filtered) <span class="sc">%in%</span> control_samples)] <span class="co"># Remove control samples from otu_table</span></span>
<span id="cb104-106"><a href="#cb104-106" aria-hidden="true" tabindex="-1"></a>  sample_data_filtered <span class="ot">&lt;-</span> sample_data[<span class="sc">!</span>sample_data<span class="sc">$</span>fullID <span class="sc">%in%</span> control_samples, ] <span class="co"># Remove control samples from sample_data</span></span>
<span id="cb104-107"><a href="#cb104-107" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-108"><a href="#cb104-108" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print new dimensions</span></span>
<span id="cb104-109"><a href="#cb104-109" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"New dimensions of otu_table_filtered:"</span>)</span>
<span id="cb104-110"><a href="#cb104-110" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">dim</span>(otu_table_filtered))</span>
<span id="cb104-111"><a href="#cb104-111" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-112"><a href="#cb104-112" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Return the modified phyloseq object</span></span>
<span id="cb104-113"><a href="#cb104-113" aria-hidden="true" tabindex="-1"></a>  modified_phylo <span class="ot">&lt;-</span> <span class="fu">phyloseq</span>(otu_table_filtered, <span class="fu">tax_table</span>(phylo_object), sample_data_filtered)</span>
<span id="cb104-114"><a href="#cb104-114" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-115"><a href="#cb104-115" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove taxa with zero reads removed in otus_removed_reads_total</span></span>
<span id="cb104-116"><a href="#cb104-116" aria-hidden="true" tabindex="-1"></a>  otus_removed_reads_total <span class="ot">&lt;-</span> otus_removed_reads_total <span class="sc">%&gt;%</span> <span class="fu">subset</span>(otus_removed_reads_total<span class="sc">$</span>reads_removed <span class="sc">!=</span> <span class="dv">0</span>)</span>
<span id="cb104-117"><a href="#cb104-117" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-118"><a href="#cb104-118" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Return the modified phyloseq object and otus_removed data frame</span></span>
<span id="cb104-119"><a href="#cb104-119" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(modified_phylo, otus_removed_reads_total))</span>
<span id="cb104-120"><a href="#cb104-120" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And finally, the function for PCR level decontamination. We have to do this by marker, considering the PCRs are different for each.</p>
<p>The 18S function is called <code>remove_PCR_control_taxa_18S_abund</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>remove_PCR_control_taxa_18S_abund <span class="ot">&lt;-</span> <span class="cf">function</span>(phylo_object) {</span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extract sample data</span></span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>  sample_data <span class="ot">&lt;-</span> <span class="fu">sample_data</span>(phylo_object)</span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-6"><a href="#cb105-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get unique batches</span></span>
<span id="cb105-7"><a href="#cb105-7" aria-hidden="true" tabindex="-1"></a>  unique_batches <span class="ot">&lt;-</span> <span class="fu">unique</span>(sample_data<span class="sc">$</span>batch_PCR_18S[<span class="sc">!</span><span class="fu">is.na</span>(sample_data<span class="sc">$</span>batch_PCR_18S)])</span>
<span id="cb105-8"><a href="#cb105-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-9"><a href="#cb105-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check if unique_localityIDs is empty</span></span>
<span id="cb105-10"><a href="#cb105-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(unique_batches) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb105-11"><a href="#cb105-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"No PCR (18S) batches found in sample data."</span>)</span>
<span id="cb105-12"><a href="#cb105-12" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb105-13"><a href="#cb105-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-14"><a href="#cb105-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize variable to count OTUs removed</span></span>
<span id="cb105-15"><a href="#cb105-15" aria-hidden="true" tabindex="-1"></a>  otus_removed <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">batch =</span> <span class="fu">character</span>(), <span class="at">OTU =</span> <span class="fu">character</span>(), <span class="at">reads_removed =</span> <span class="fu">integer</span>(), <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb105-16"><a href="#cb105-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-17"><a href="#cb105-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize filtered OTU table</span></span>
<span id="cb105-18"><a href="#cb105-18" aria-hidden="true" tabindex="-1"></a>  otu_table_filtered <span class="ot">&lt;-</span> <span class="fu">otu_table</span>(phylo_object)</span>
<span id="cb105-19"><a href="#cb105-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-20"><a href="#cb105-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check if the OTU table is empty</span></span>
<span id="cb105-21"><a href="#cb105-21" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">dim</span>(otu_table_filtered)[<span class="dv">1</span>] <span class="sc">==</span> <span class="dv">0</span> <span class="sc">||</span> <span class="fu">dim</span>(otu_table_filtered)[<span class="dv">2</span>] <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb105-22"><a href="#cb105-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"OTU abundance data is empty."</span>)</span>
<span id="cb105-23"><a href="#cb105-23" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb105-24"><a href="#cb105-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-25"><a href="#cb105-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"Original dimensions of otu_table_filtered:"</span>)</span>
<span id="cb105-26"><a href="#cb105-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">dim</span>(otu_table_filtered))</span>
<span id="cb105-27"><a href="#cb105-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-28"><a href="#cb105-28" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize list to store removed OTUs for each sample</span></span>
<span id="cb105-29"><a href="#cb105-29" aria-hidden="true" tabindex="-1"></a>  otus_removed_reads <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb105-30"><a href="#cb105-30" aria-hidden="true" tabindex="-1"></a>  otus_removed_reads_total <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb105-31"><a href="#cb105-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-32"><a href="#cb105-32" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Loop through each batch_PCR_18S</span></span>
<span id="cb105-33"><a href="#cb105-33" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (batch <span class="cf">in</span> unique_batches) {</span>
<span id="cb105-34"><a href="#cb105-34" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb105-35"><a href="#cb105-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get extract control samples for the current batch</span></span>
<span id="cb105-36"><a href="#cb105-36" aria-hidden="true" tabindex="-1"></a>    pcr_18S_control_samples <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">sub</span>(<span class="st">"_[^_]*$"</span>, <span class="st">""</span>, sample_data<span class="sc">$</span>fullID[sample_data<span class="sc">$</span>sampleType <span class="sc">==</span> <span class="st">"lab-negative-control"</span> </span>
<span id="cb105-37"><a href="#cb105-37" aria-hidden="true" tabindex="-1"></a>                                                                            <span class="sc">&amp;</span> sample_data<span class="sc">$</span>batch_PCR_18S <span class="sc">==</span> batch]))</span>
<span id="cb105-38"><a href="#cb105-38" aria-hidden="true" tabindex="-1"></a>    pcr_18S_control_samples <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(pcr_18S_control_samples)</span>
<span id="cb105-39"><a href="#cb105-39" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb105-40"><a href="#cb105-40" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If there are no PCR 18S control samples for the current batch, print a message and skip</span></span>
<span id="cb105-41"><a href="#cb105-41" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(pcr_18S_control_samples) <span class="sc">&lt;</span> <span class="dv">1</span>) {</span>
<span id="cb105-42"><a href="#cb105-42" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No PCR (18S) control samples found in batch:"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb105-43"><a href="#cb105-43" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb105-44"><a href="#cb105-44" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb105-45"><a href="#cb105-45" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb105-46"><a href="#cb105-46" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get indices of PCR (18S) control and samples</span></span>
<span id="cb105-47"><a href="#cb105-47" aria-hidden="true" tabindex="-1"></a>    control_indices <span class="ot">&lt;-</span> <span class="fu">which</span>(sample_data<span class="sc">$</span>fieldID <span class="sc">%in%</span> pcr_18S_control_samples)</span>
<span id="cb105-48"><a href="#cb105-48" aria-hidden="true" tabindex="-1"></a>    sample_indices <span class="ot">&lt;-</span> <span class="fu">which</span>(sample_data<span class="sc">$</span>batch_PCR_18S <span class="sc">==</span> batch <span class="sc">&amp;</span> <span class="sc">!</span>sample_data<span class="sc">$</span>fullID <span class="sc">%in%</span> pcr_18S_control_samples)</span>
<span id="cb105-49"><a href="#cb105-49" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb105-50"><a href="#cb105-50" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If no control indices found</span></span>
<span id="cb105-51"><a href="#cb105-51" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(control_indices) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb105-52"><a href="#cb105-52" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No control indices found for batch:"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb105-53"><a href="#cb105-53" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb105-54"><a href="#cb105-54" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb105-55"><a href="#cb105-55" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb105-56"><a href="#cb105-56" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If no sample indices found</span></span>
<span id="cb105-57"><a href="#cb105-57" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(sample_indices) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb105-58"><a href="#cb105-58" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No sample indices found for batch:"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb105-59"><a href="#cb105-59" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb105-60"><a href="#cb105-60" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb105-61"><a href="#cb105-61" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb105-62"><a href="#cb105-62" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If there are more than PCR 18S control samples for the current locality, loop through the controls</span></span>
<span id="cb105-63"><a href="#cb105-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(pcr_18S_control_samples) <span class="sc">&gt;</span> <span class="dv">2</span>) {</span>
<span id="cb105-64"><a href="#cb105-64" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"More than 1 PCR (18S) control sample found in batch"</span>, batch, <span class="st">". Now printing individual controls:"</span>, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb105-65"><a href="#cb105-65" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb105-66"><a href="#cb105-66" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> (control <span class="cf">in</span> control_indices) {</span>
<span id="cb105-67"><a href="#cb105-67" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb105-68"><a href="#cb105-68" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Calculate mean abundance of each OTU in the control and samples</span></span>
<span id="cb105-69"><a href="#cb105-69" aria-hidden="true" tabindex="-1"></a>        control_abundance <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(otu_table_filtered[, control])</span>
<span id="cb105-70"><a href="#cb105-70" aria-hidden="true" tabindex="-1"></a>        sample_abundance <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(otu_table_filtered[, sample_indices])</span>
<span id="cb105-71"><a href="#cb105-71" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb105-72"><a href="#cb105-72" aria-hidden="true" tabindex="-1"></a>        <span class="co">#check there are OTUs with reads above minReads in the field control samples</span></span>
<span id="cb105-73"><a href="#cb105-73" aria-hidden="true" tabindex="-1"></a>        control_sum_above_minReads <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(otu_table_filtered[, control]) <span class="sc">&gt;</span> <span class="dv">100</span></span>
<span id="cb105-74"><a href="#cb105-74" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb105-75"><a href="#cb105-75" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="fu">any</span>(control_sum_above_minReads)) {</span>
<span id="cb105-76"><a href="#cb105-76" aria-hidden="true" tabindex="-1"></a>          contaminants <span class="ot">&lt;-</span> <span class="fu">rownames</span>(otu_table_filtered[control_abundance <span class="sc">&gt;</span> sample_abundance])</span>
<span id="cb105-77"><a href="#cb105-77" aria-hidden="true" tabindex="-1"></a>          <span class="fu">cat</span>(<span class="st">"Batch"</span>, batch, <span class="st">":"</span>, <span class="fu">sum</span>(control_sum_above_minReads <span class="sc">==</span> T), <span class="st">"contaminant taxa found for control indices"</span>, control, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb105-78"><a href="#cb105-78" aria-hidden="true" tabindex="-1"></a>        } <span class="cf">else</span> {</span>
<span id="cb105-79"><a href="#cb105-79" aria-hidden="true" tabindex="-1"></a>          <span class="fu">cat</span>(<span class="st">"Batch"</span>, batch, <span class="st">": No contaminants found for control indices"</span>, control, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb105-80"><a href="#cb105-80" aria-hidden="true" tabindex="-1"></a>          <span class="cf">next</span></span>
<span id="cb105-81"><a href="#cb105-81" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb105-82"><a href="#cb105-82" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb105-83"><a href="#cb105-83" aria-hidden="true" tabindex="-1"></a>        <span class="co">#Get original number of reads in samples for taxa with above minReads reads in controls</span></span>
<span id="cb105-84"><a href="#cb105-84" aria-hidden="true" tabindex="-1"></a>        <span class="co">#bind the previous output</span></span>
<span id="cb105-85"><a href="#cb105-85" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="fu">length</span>(otus_removed_reads) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb105-86"><a href="#cb105-86" aria-hidden="true" tabindex="-1"></a>          otus_removed_reads_total <span class="ot">&lt;-</span> <span class="fu">rbind</span>(otus_removed_reads, otus_removed_reads_total)</span>
<span id="cb105-87"><a href="#cb105-87" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb105-88"><a href="#cb105-88" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb105-89"><a href="#cb105-89" aria-hidden="true" tabindex="-1"></a>        <span class="co">#get reads</span></span>
<span id="cb105-90"><a href="#cb105-90" aria-hidden="true" tabindex="-1"></a>        otus_removed_reads <span class="ot">&lt;-</span> otu_table_filtered[contaminants,  sample_indices] <span class="sc">%&gt;%</span> reshape2<span class="sc">::</span><span class="fu">melt</span>()</span>
<span id="cb105-91"><a href="#cb105-91" aria-hidden="true" tabindex="-1"></a>        <span class="fu">colnames</span>(otus_removed_reads) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"taxa"</span>, <span class="st">"fullID"</span>, <span class="st">"reads_removed"</span>)</span>
<span id="cb105-92"><a href="#cb105-92" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb105-93"><a href="#cb105-93" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Set read count of OTUs in real samples to zero if reads in field control samples exceed 100</span></span>
<span id="cb105-94"><a href="#cb105-94" aria-hidden="true" tabindex="-1"></a>        otu_table_filtered[contaminants, sample_indices] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb105-95"><a href="#cb105-95" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb105-96"><a href="#cb105-96" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Remove the control samples from the phylo object</span></span>
<span id="cb105-97"><a href="#cb105-97" aria-hidden="true" tabindex="-1"></a>        otu_table_filtered <span class="ot">&lt;-</span> otu_table_filtered[<span class="sc">-</span>control]</span>
<span id="cb105-98"><a href="#cb105-98" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb105-99"><a href="#cb105-99" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb105-100"><a href="#cb105-100" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb105-101"><a href="#cb105-101" aria-hidden="true" tabindex="-1"></a>      <span class="co">#check there are OTUs with reads above minReads in the field control samples</span></span>
<span id="cb105-102"><a href="#cb105-102" aria-hidden="true" tabindex="-1"></a>      control_sum_above_minReads <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(otu_table_filtered[, control_indices]) <span class="sc">&gt;</span> minReads</span>
<span id="cb105-103"><a href="#cb105-103" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb105-104"><a href="#cb105-104" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="fu">any</span>(control_sum_above_minReads)) {</span>
<span id="cb105-105"><a href="#cb105-105" aria-hidden="true" tabindex="-1"></a>        contaminants <span class="ot">&lt;-</span> <span class="fu">rownames</span>(otu_table_filtered[<span class="fu">rowSums</span>(otu_table_filtered[, control_indices]) <span class="sc">&gt;</span> minReads])</span>
<span id="cb105-106"><a href="#cb105-106" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="fu">sum</span>(control_sum_above_minReads <span class="sc">==</span> T), <span class="st">"contaminant taxa found for batch"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb105-107"><a href="#cb105-107" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb105-108"><a href="#cb105-108" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"No contaminants found for batch"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb105-109"><a href="#cb105-109" aria-hidden="true" tabindex="-1"></a>        <span class="cf">next</span></span>
<span id="cb105-110"><a href="#cb105-110" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb105-111"><a href="#cb105-111" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb105-112"><a href="#cb105-112" aria-hidden="true" tabindex="-1"></a>      <span class="co">#Get original number of reads in samples for taxa with above minReads reads in controls</span></span>
<span id="cb105-113"><a href="#cb105-113" aria-hidden="true" tabindex="-1"></a>      <span class="co">#bind the previous output</span></span>
<span id="cb105-114"><a href="#cb105-114" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="fu">length</span>(otus_removed_reads) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb105-115"><a href="#cb105-115" aria-hidden="true" tabindex="-1"></a>        otus_removed_reads_total <span class="ot">&lt;-</span> <span class="fu">rbind</span>(otus_removed_reads, otus_removed_reads_total)</span>
<span id="cb105-116"><a href="#cb105-116" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb105-117"><a href="#cb105-117" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb105-118"><a href="#cb105-118" aria-hidden="true" tabindex="-1"></a>      <span class="co">#get reads</span></span>
<span id="cb105-119"><a href="#cb105-119" aria-hidden="true" tabindex="-1"></a>      otus_removed_reads <span class="ot">&lt;-</span> otu_table_filtered[contaminants,  sample_indices] <span class="sc">%&gt;%</span> reshape2<span class="sc">::</span><span class="fu">melt</span>()</span>
<span id="cb105-120"><a href="#cb105-120" aria-hidden="true" tabindex="-1"></a>      <span class="fu">colnames</span>(otus_removed_reads) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"taxa"</span>, <span class="st">"fullID"</span>, <span class="st">"reads_removed"</span>)</span>
<span id="cb105-121"><a href="#cb105-121" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb105-122"><a href="#cb105-122" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Set read count of OTUs in real samples to zero if reads in field control samples exceed 100</span></span>
<span id="cb105-123"><a href="#cb105-123" aria-hidden="true" tabindex="-1"></a>      otu_table_filtered[contaminants, sample_indices] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb105-124"><a href="#cb105-124" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb105-125"><a href="#cb105-125" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Remove the control samples from the phylo object</span></span>
<span id="cb105-126"><a href="#cb105-126" aria-hidden="true" tabindex="-1"></a>      otu_table_filtered <span class="ot">&lt;-</span> otu_table_filtered[<span class="sc">-</span>control_indices]</span>
<span id="cb105-127"><a href="#cb105-127" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb105-128"><a href="#cb105-128" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb105-129"><a href="#cb105-129" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-130"><a href="#cb105-130" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print OTUs removed</span></span>
<span id="cb105-131"><a href="#cb105-131" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Reads removed:"</span>, <span class="fu">sum</span>(otus_removed_reads_total<span class="sc">$</span>reads_removed),<span class="st">"; Samples filtered:"</span>, <span class="fu">length</span>(<span class="fu">unique</span>(otus_removed_reads_total<span class="sc">$</span>fullID)), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb105-132"><a href="#cb105-132" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-133"><a href="#cb105-133" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get list of control samples</span></span>
<span id="cb105-134"><a href="#cb105-134" aria-hidden="true" tabindex="-1"></a>  control_samples <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">sub</span>(<span class="st">"_[^_]*$"</span>, <span class="st">""</span>, sample_data<span class="sc">$</span>fullID[sample_data<span class="sc">$</span>sampleType <span class="sc">==</span> <span class="st">"lab-negative-control"</span>])) <span class="co"># Build list of names of control samples</span></span>
<span id="cb105-135"><a href="#cb105-135" aria-hidden="true" tabindex="-1"></a>  control_samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">paste0</span>(control_samples, <span class="st">"_18S"</span>)) <span class="co"># Create a list of control samples by appending "_CO1" to each ID</span></span>
<span id="cb105-136"><a href="#cb105-136" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-137"><a href="#cb105-137" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check control samples to filtered otu table (will be more in the list than reality because some aren't _CO1)</span></span>
<span id="cb105-138"><a href="#cb105-138" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(control_samples)</span>
<span id="cb105-139"><a href="#cb105-139" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">which</span>(control_samples <span class="sc">%in%</span> <span class="fu">colnames</span>(otu_table_filtered)))</span>
<span id="cb105-140"><a href="#cb105-140" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-141"><a href="#cb105-141" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove control samples from otu and sample table</span></span>
<span id="cb105-142"><a href="#cb105-142" aria-hidden="true" tabindex="-1"></a>  otu_table_filtered <span class="ot">&lt;-</span> otu_table_filtered[, <span class="sc">-</span><span class="fu">which</span>(<span class="fu">colnames</span>(otu_table_filtered) <span class="sc">%in%</span> control_samples)] <span class="co"># Remove control samples from otu_table</span></span>
<span id="cb105-143"><a href="#cb105-143" aria-hidden="true" tabindex="-1"></a>  sample_data_filtered <span class="ot">&lt;-</span> sample_data[<span class="sc">!</span>sample_data<span class="sc">$</span>fullID <span class="sc">%in%</span> control_samples, ] <span class="co"># Remove control samples from sample_data</span></span>
<span id="cb105-144"><a href="#cb105-144" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-145"><a href="#cb105-145" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print new dimensions</span></span>
<span id="cb105-146"><a href="#cb105-146" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"New dimensions of otu_table_filtered:"</span>)</span>
<span id="cb105-147"><a href="#cb105-147" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">dim</span>(otu_table_filtered))</span>
<span id="cb105-148"><a href="#cb105-148" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-149"><a href="#cb105-149" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Return the modified phyloseq object</span></span>
<span id="cb105-150"><a href="#cb105-150" aria-hidden="true" tabindex="-1"></a>  modified_phylo <span class="ot">&lt;-</span> <span class="fu">phyloseq</span>(otu_table_filtered, <span class="fu">tax_table</span>(phylo_object), sample_data_filtered)</span>
<span id="cb105-151"><a href="#cb105-151" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-152"><a href="#cb105-152" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove taxa with zero reads removed in otus_removed_reads_total</span></span>
<span id="cb105-153"><a href="#cb105-153" aria-hidden="true" tabindex="-1"></a>  otus_removed_reads_total <span class="ot">&lt;-</span> otus_removed_reads_total <span class="sc">%&gt;%</span> <span class="fu">subset</span>(otus_removed_reads_total<span class="sc">$</span>reads_removed <span class="sc">!=</span> <span class="dv">0</span>)</span>
<span id="cb105-154"><a href="#cb105-154" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb105-155"><a href="#cb105-155" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Return the modified phyloseq object and otus_removed dataframe</span></span>
<span id="cb105-156"><a href="#cb105-156" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(modified_phylo, otus_removed_reads_total))</span>
<span id="cb105-157"><a href="#cb105-157" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And the CO1 function is called <code>remove_PCR_control_taxa_CO1_abund</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="do">### PCR negative decontamination (CO1)</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>remove_PCR_control_taxa_CO1_abund <span class="ot">&lt;-</span> <span class="cf">function</span>(phylo_object) {</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extract sample data</span></span>
<span id="cb106-5"><a href="#cb106-5" aria-hidden="true" tabindex="-1"></a>  sample_data <span class="ot">&lt;-</span> <span class="fu">sample_data</span>(phylo_object)</span>
<span id="cb106-6"><a href="#cb106-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-7"><a href="#cb106-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get unique batches</span></span>
<span id="cb106-8"><a href="#cb106-8" aria-hidden="true" tabindex="-1"></a>  unique_batches <span class="ot">&lt;-</span> <span class="fu">unique</span>(sample_data<span class="sc">$</span>batch_PCR_CO1[<span class="sc">!</span><span class="fu">is.na</span>(sample_data<span class="sc">$</span>batch_PCR_CO1)])</span>
<span id="cb106-9"><a href="#cb106-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-10"><a href="#cb106-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check if unique_localityIDs is empty</span></span>
<span id="cb106-11"><a href="#cb106-11" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(unique_batches) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb106-12"><a href="#cb106-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"No PCR (CO1) batches found in sample data."</span>)</span>
<span id="cb106-13"><a href="#cb106-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb106-14"><a href="#cb106-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-15"><a href="#cb106-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize variable to count OTUs removed</span></span>
<span id="cb106-16"><a href="#cb106-16" aria-hidden="true" tabindex="-1"></a>  otus_removed <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">batch =</span> <span class="fu">character</span>(), <span class="at">OTU =</span> <span class="fu">character</span>(), <span class="at">reads_removed =</span> <span class="fu">integer</span>(), <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb106-17"><a href="#cb106-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-18"><a href="#cb106-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize filtered OTU table</span></span>
<span id="cb106-19"><a href="#cb106-19" aria-hidden="true" tabindex="-1"></a>  otu_table_filtered <span class="ot">&lt;-</span> <span class="fu">otu_table</span>(phylo_object)</span>
<span id="cb106-20"><a href="#cb106-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-21"><a href="#cb106-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check if the OTU table is empty</span></span>
<span id="cb106-22"><a href="#cb106-22" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">dim</span>(otu_table_filtered)[<span class="dv">1</span>] <span class="sc">==</span> <span class="dv">0</span> <span class="sc">||</span> <span class="fu">dim</span>(otu_table_filtered)[<span class="dv">2</span>] <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb106-23"><a href="#cb106-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"OTU abundance data is empty."</span>)</span>
<span id="cb106-24"><a href="#cb106-24" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb106-25"><a href="#cb106-25" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-26"><a href="#cb106-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"Original dimensions of otu_table_filtered:"</span>)</span>
<span id="cb106-27"><a href="#cb106-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">dim</span>(otu_table_filtered))</span>
<span id="cb106-28"><a href="#cb106-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-29"><a href="#cb106-29" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize list to store removed OTUs for each sample</span></span>
<span id="cb106-30"><a href="#cb106-30" aria-hidden="true" tabindex="-1"></a>  otus_removed_reads <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb106-31"><a href="#cb106-31" aria-hidden="true" tabindex="-1"></a>  otus_removed_reads_total <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb106-32"><a href="#cb106-32" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-33"><a href="#cb106-33" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Loop through each batch_PCR_CO1</span></span>
<span id="cb106-34"><a href="#cb106-34" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (batch <span class="cf">in</span> unique_batches) {</span>
<span id="cb106-35"><a href="#cb106-35" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb106-36"><a href="#cb106-36" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get extract control samples for the current batch</span></span>
<span id="cb106-37"><a href="#cb106-37" aria-hidden="true" tabindex="-1"></a>    pcr_CO1_control_samples <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">sub</span>(<span class="st">"_[^_]*$"</span>, <span class="st">""</span>, sample_data<span class="sc">$</span>fullID[sample_data<span class="sc">$</span>sampleType <span class="sc">==</span> <span class="st">"lab-negative-control"</span> </span>
<span id="cb106-38"><a href="#cb106-38" aria-hidden="true" tabindex="-1"></a>                                                                            <span class="sc">&amp;</span> sample_data<span class="sc">$</span>batch_PCR_CO1 <span class="sc">==</span> batch]))</span>
<span id="cb106-39"><a href="#cb106-39" aria-hidden="true" tabindex="-1"></a>    pcr_CO1_control_samples <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(pcr_CO1_control_samples)</span>
<span id="cb106-40"><a href="#cb106-40" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb106-41"><a href="#cb106-41" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If there are no PCR CO1 control samples for the current batch, print a message and skip</span></span>
<span id="cb106-42"><a href="#cb106-42" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(pcr_CO1_control_samples) <span class="sc">&lt;</span> <span class="dv">1</span>) {</span>
<span id="cb106-43"><a href="#cb106-43" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No PCR (CO1) control samples found in batch:"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb106-44"><a href="#cb106-44" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb106-45"><a href="#cb106-45" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb106-46"><a href="#cb106-46" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb106-47"><a href="#cb106-47" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get indices of PCR (CO1) control and samples</span></span>
<span id="cb106-48"><a href="#cb106-48" aria-hidden="true" tabindex="-1"></a>    control_indices <span class="ot">&lt;-</span> <span class="fu">which</span>(sample_data<span class="sc">$</span>fieldID <span class="sc">%in%</span> pcr_CO1_control_samples)</span>
<span id="cb106-49"><a href="#cb106-49" aria-hidden="true" tabindex="-1"></a>    sample_indices <span class="ot">&lt;-</span> <span class="fu">which</span>(sample_data<span class="sc">$</span>batch_PCR_CO1 <span class="sc">==</span> batch <span class="sc">&amp;</span> <span class="sc">!</span>sample_data<span class="sc">$</span>fullID <span class="sc">%in%</span> pcr_CO1_control_samples)</span>
<span id="cb106-50"><a href="#cb106-50" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb106-51"><a href="#cb106-51" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If no control indices found</span></span>
<span id="cb106-52"><a href="#cb106-52" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(control_indices) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb106-53"><a href="#cb106-53" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No control indices found for batch:"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb106-54"><a href="#cb106-54" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb106-55"><a href="#cb106-55" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb106-56"><a href="#cb106-56" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb106-57"><a href="#cb106-57" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If no sample indices found</span></span>
<span id="cb106-58"><a href="#cb106-58" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(sample_indices) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb106-59"><a href="#cb106-59" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"No sample indices found for batch:"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb106-60"><a href="#cb106-60" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb106-61"><a href="#cb106-61" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb106-62"><a href="#cb106-62" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb106-63"><a href="#cb106-63" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If there are more than PCR CO1 control samples for the current locality, loop through the controls</span></span>
<span id="cb106-64"><a href="#cb106-64" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(pcr_CO1_control_samples) <span class="sc">&gt;</span> <span class="dv">2</span>) {</span>
<span id="cb106-65"><a href="#cb106-65" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"More than 1 PCR (CO1) control sample found in batch"</span>, batch, <span class="st">". Now printing individual controls:"</span>, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb106-66"><a href="#cb106-66" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb106-67"><a href="#cb106-67" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> (control <span class="cf">in</span> control_indices) {</span>
<span id="cb106-68"><a href="#cb106-68" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb106-69"><a href="#cb106-69" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Calculate mean abundance of each OTU in the control and samples</span></span>
<span id="cb106-70"><a href="#cb106-70" aria-hidden="true" tabindex="-1"></a>        control_abundance <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(otu_table_filtered[, control])</span>
<span id="cb106-71"><a href="#cb106-71" aria-hidden="true" tabindex="-1"></a>        sample_abundance <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(otu_table_filtered[, sample_indices])</span>
<span id="cb106-72"><a href="#cb106-72" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb106-73"><a href="#cb106-73" aria-hidden="true" tabindex="-1"></a>        <span class="co">#check there are OTUs with reads above minReads in the field control samples</span></span>
<span id="cb106-74"><a href="#cb106-74" aria-hidden="true" tabindex="-1"></a>        control_sum_above_minReads <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(otu_table_filtered[, control]) <span class="sc">&gt;</span> <span class="dv">100</span></span>
<span id="cb106-75"><a href="#cb106-75" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb106-76"><a href="#cb106-76" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="fu">any</span>(control_sum_above_minReads)) {</span>
<span id="cb106-77"><a href="#cb106-77" aria-hidden="true" tabindex="-1"></a>          <span class="cf">if</span> (<span class="fu">any</span>(control_abundance <span class="sc">&gt;</span> sample_abundance)) {</span>
<span id="cb106-78"><a href="#cb106-78" aria-hidden="true" tabindex="-1"></a>            contaminants <span class="ot">&lt;-</span> <span class="fu">rownames</span>(otu_table_filtered[control_abundance <span class="sc">&gt;</span> sample_abundance])</span>
<span id="cb106-79"><a href="#cb106-79" aria-hidden="true" tabindex="-1"></a>            <span class="fu">cat</span>(<span class="st">"Batch"</span>, batch, <span class="st">":"</span>, <span class="fu">sum</span>(control_sum_above_minReads <span class="sc">==</span> <span class="cn">TRUE</span>), <span class="st">"contaminant taxa found for control indices"</span>, control, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb106-80"><a href="#cb106-80" aria-hidden="true" tabindex="-1"></a>          } <span class="cf">else</span> {</span>
<span id="cb106-81"><a href="#cb106-81" aria-hidden="true" tabindex="-1"></a>            <span class="fu">cat</span>(<span class="st">"Batch"</span>, batch, <span class="st">": Control abundance is not greater than sample abundance. OTU table would be empty.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb106-82"><a href="#cb106-82" aria-hidden="true" tabindex="-1"></a>            <span class="cf">next</span></span>
<span id="cb106-83"><a href="#cb106-83" aria-hidden="true" tabindex="-1"></a>          }</span>
<span id="cb106-84"><a href="#cb106-84" aria-hidden="true" tabindex="-1"></a>        } <span class="cf">else</span> {</span>
<span id="cb106-85"><a href="#cb106-85" aria-hidden="true" tabindex="-1"></a>          <span class="fu">cat</span>(<span class="st">"Batch"</span>, batch, <span class="st">": No contaminants found for control indices"</span>, control, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb106-86"><a href="#cb106-86" aria-hidden="true" tabindex="-1"></a>          <span class="cf">next</span></span>
<span id="cb106-87"><a href="#cb106-87" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb106-88"><a href="#cb106-88" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb106-89"><a href="#cb106-89" aria-hidden="true" tabindex="-1"></a>        <span class="co">#Get original number of reads in samples for taxa with above minReads reads in controls</span></span>
<span id="cb106-90"><a href="#cb106-90" aria-hidden="true" tabindex="-1"></a>        <span class="co">#bind the previous output</span></span>
<span id="cb106-91"><a href="#cb106-91" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="fu">length</span>(otus_removed_reads) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb106-92"><a href="#cb106-92" aria-hidden="true" tabindex="-1"></a>          otus_removed_reads_total <span class="ot">&lt;-</span> <span class="fu">rbind</span>(otus_removed_reads, otus_removed_reads_total)</span>
<span id="cb106-93"><a href="#cb106-93" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb106-94"><a href="#cb106-94" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb106-95"><a href="#cb106-95" aria-hidden="true" tabindex="-1"></a>        <span class="co">#get reads</span></span>
<span id="cb106-96"><a href="#cb106-96" aria-hidden="true" tabindex="-1"></a>        otus_removed_reads <span class="ot">&lt;-</span> otu_table_filtered[contaminants,  sample_indices] <span class="sc">%&gt;%</span> reshape2<span class="sc">::</span><span class="fu">melt</span>()</span>
<span id="cb106-97"><a href="#cb106-97" aria-hidden="true" tabindex="-1"></a>        <span class="fu">colnames</span>(otus_removed_reads) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"taxa"</span>, <span class="st">"fullID"</span>, <span class="st">"reads_removed"</span>)</span>
<span id="cb106-98"><a href="#cb106-98" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb106-99"><a href="#cb106-99" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Set read count of OTUs in real samples to zero if reads in field control samples exceed 100</span></span>
<span id="cb106-100"><a href="#cb106-100" aria-hidden="true" tabindex="-1"></a>        otu_table_filtered[contaminants, sample_indices] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb106-101"><a href="#cb106-101" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb106-102"><a href="#cb106-102" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Remove the control samples from the phylo object</span></span>
<span id="cb106-103"><a href="#cb106-103" aria-hidden="true" tabindex="-1"></a>        otu_table_filtered <span class="ot">&lt;-</span> otu_table_filtered[<span class="sc">-</span>control]</span>
<span id="cb106-104"><a href="#cb106-104" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb106-105"><a href="#cb106-105" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb106-106"><a href="#cb106-106" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb106-107"><a href="#cb106-107" aria-hidden="true" tabindex="-1"></a>      <span class="co">#check there are OTUs with reads above minReads in the field control samples</span></span>
<span id="cb106-108"><a href="#cb106-108" aria-hidden="true" tabindex="-1"></a>      control_sum_above_minReads <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(otu_table_filtered[, control_indices]) <span class="sc">&gt;</span> minReads</span>
<span id="cb106-109"><a href="#cb106-109" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb106-110"><a href="#cb106-110" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="fu">any</span>(control_sum_above_minReads)) {</span>
<span id="cb106-111"><a href="#cb106-111" aria-hidden="true" tabindex="-1"></a>        contaminants <span class="ot">&lt;-</span> <span class="fu">rownames</span>(otu_table_filtered[<span class="fu">rowSums</span>(otu_table_filtered[, control_indices]) <span class="sc">&gt;</span> minReads])</span>
<span id="cb106-112"><a href="#cb106-112" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="fu">sum</span>(control_sum_above_minReads <span class="sc">==</span> T), <span class="st">"contaminant taxa found for batch"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb106-113"><a href="#cb106-113" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb106-114"><a href="#cb106-114" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"No contaminants found for batch"</span>, batch, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb106-115"><a href="#cb106-115" aria-hidden="true" tabindex="-1"></a>        <span class="cf">next</span></span>
<span id="cb106-116"><a href="#cb106-116" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb106-117"><a href="#cb106-117" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb106-118"><a href="#cb106-118" aria-hidden="true" tabindex="-1"></a>      <span class="co">#Get original number of reads in samples for taxa with above minReads reads in controls</span></span>
<span id="cb106-119"><a href="#cb106-119" aria-hidden="true" tabindex="-1"></a>      <span class="co">#bind the previous output</span></span>
<span id="cb106-120"><a href="#cb106-120" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="fu">length</span>(otus_removed_reads) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb106-121"><a href="#cb106-121" aria-hidden="true" tabindex="-1"></a>        otus_removed_reads_total <span class="ot">&lt;-</span> <span class="fu">rbind</span>(otus_removed_reads, otus_removed_reads_total)</span>
<span id="cb106-122"><a href="#cb106-122" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb106-123"><a href="#cb106-123" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb106-124"><a href="#cb106-124" aria-hidden="true" tabindex="-1"></a>      <span class="co">#get reads</span></span>
<span id="cb106-125"><a href="#cb106-125" aria-hidden="true" tabindex="-1"></a>      otus_removed_reads <span class="ot">&lt;-</span> otu_table_filtered[contaminants,  sample_indices] <span class="sc">%&gt;%</span> reshape2<span class="sc">::</span><span class="fu">melt</span>()</span>
<span id="cb106-126"><a href="#cb106-126" aria-hidden="true" tabindex="-1"></a>      <span class="fu">colnames</span>(otus_removed_reads) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"taxa"</span>, <span class="st">"fullID"</span>, <span class="st">"reads_removed"</span>)</span>
<span id="cb106-127"><a href="#cb106-127" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb106-128"><a href="#cb106-128" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Set read count of OTUs in real samples to zero if reads in field control samples exceed 100</span></span>
<span id="cb106-129"><a href="#cb106-129" aria-hidden="true" tabindex="-1"></a>      otu_table_filtered[contaminants, sample_indices] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb106-130"><a href="#cb106-130" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb106-131"><a href="#cb106-131" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Remove the control samples from the phylo object</span></span>
<span id="cb106-132"><a href="#cb106-132" aria-hidden="true" tabindex="-1"></a>      otu_table_filtered <span class="ot">&lt;-</span> otu_table_filtered[<span class="sc">-</span>control_indices]</span>
<span id="cb106-133"><a href="#cb106-133" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb106-134"><a href="#cb106-134" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb106-135"><a href="#cb106-135" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-136"><a href="#cb106-136" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print OTUs removed</span></span>
<span id="cb106-137"><a href="#cb106-137" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Reads removed:"</span>, <span class="fu">sum</span>(otus_removed_reads_total<span class="sc">$</span>reads_removed),<span class="st">"; Samples filtered:"</span>, <span class="fu">length</span>(<span class="fu">unique</span>(otus_removed_reads_total<span class="sc">$</span>fullID)), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb106-138"><a href="#cb106-138" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-139"><a href="#cb106-139" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get list of control samples</span></span>
<span id="cb106-140"><a href="#cb106-140" aria-hidden="true" tabindex="-1"></a>  control_samples <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">sub</span>(<span class="st">"_[^_]*$"</span>, <span class="st">""</span>, sample_data<span class="sc">$</span>fullID[sample_data<span class="sc">$</span>sampleType <span class="sc">==</span> <span class="st">"lab-negative-control"</span>])) <span class="co"># Build list of names of control samples</span></span>
<span id="cb106-141"><a href="#cb106-141" aria-hidden="true" tabindex="-1"></a>  control_samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">paste0</span>(control_samples, <span class="st">"_CO1"</span>)) <span class="co"># Create a list of control samples by appending "_CO1" to each ID</span></span>
<span id="cb106-142"><a href="#cb106-142" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-143"><a href="#cb106-143" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check control samples to filtered otu table (will be more in the list than reality because some aren't _CO1)</span></span>
<span id="cb106-144"><a href="#cb106-144" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(control_samples)</span>
<span id="cb106-145"><a href="#cb106-145" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">which</span>(control_samples <span class="sc">%in%</span> <span class="fu">colnames</span>(otu_table_filtered)))</span>
<span id="cb106-146"><a href="#cb106-146" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-147"><a href="#cb106-147" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove control samples from otu and sample table</span></span>
<span id="cb106-148"><a href="#cb106-148" aria-hidden="true" tabindex="-1"></a>  otu_table_filtered <span class="ot">&lt;-</span> otu_table_filtered[, <span class="sc">-</span><span class="fu">which</span>(<span class="fu">colnames</span>(otu_table_filtered) <span class="sc">%in%</span> control_samples)] <span class="co"># Remove control samples from otu_table</span></span>
<span id="cb106-149"><a href="#cb106-149" aria-hidden="true" tabindex="-1"></a>  sample_data_filtered <span class="ot">&lt;-</span> sample_data[<span class="sc">!</span>sample_data<span class="sc">$</span>fullID <span class="sc">%in%</span> control_samples, ] <span class="co"># Remove control samples from sample_data</span></span>
<span id="cb106-150"><a href="#cb106-150" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-151"><a href="#cb106-151" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print new dimensions</span></span>
<span id="cb106-152"><a href="#cb106-152" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"New dimensions of otu_table_filtered:"</span>)</span>
<span id="cb106-153"><a href="#cb106-153" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">dim</span>(otu_table_filtered))</span>
<span id="cb106-154"><a href="#cb106-154" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-155"><a href="#cb106-155" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Return the modified phyloseq object</span></span>
<span id="cb106-156"><a href="#cb106-156" aria-hidden="true" tabindex="-1"></a>  modified_phylo <span class="ot">&lt;-</span> <span class="fu">phyloseq</span>(otu_table_filtered, <span class="fu">tax_table</span>(phylo_object), sample_data_filtered)</span>
<span id="cb106-157"><a href="#cb106-157" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-158"><a href="#cb106-158" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove taxa with zero reads removed in otus_removed_reads_total</span></span>
<span id="cb106-159"><a href="#cb106-159" aria-hidden="true" tabindex="-1"></a>  otus_removed_reads_total <span class="ot">&lt;-</span> otus_removed_reads_total <span class="sc">%&gt;%</span> <span class="fu">subset</span>(otus_removed_reads_total<span class="sc">$</span>reads_removed <span class="sc">!=</span> <span class="dv">0</span>)</span>
<span id="cb106-160"><a href="#cb106-160" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb106-161"><a href="#cb106-161" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Return the modified phyloseq object and otus_removed data frame</span></span>
<span id="cb106-162"><a href="#cb106-162" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(modified_phylo, otus_removed_reads_total))</span>
<span id="cb106-163"><a href="#cb106-163" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Time to run the decontamination functions one after the other, using the output of the first as the input for the last.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>minReads <span class="ot">=</span> <span class="dv">100</span> <span class="co">#to identify a contaminant</span></span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Call remove_field_control_taxa_abund() function (site)</span></span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a>result_field_abun <span class="ot">&lt;-</span> <span class="fu">remove_field_control_taxa_abund</span>(phylo_eDNA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Original dimensions of otu_table_filtered:"
[1] 1725  655
1 contaminant taxa found for site: SCH01 
4 contaminant taxa found for site: SCH02 
1 contaminant taxa found for site: SCH03 
No contaminants found for site: SCH04 
2 contaminant taxa found for site: SCH05 
3 contaminant taxa found for site: SCH06 
1 contaminant taxa found for site: SCH07 
1 contaminant taxa found for site: SCH08 
2 contaminant taxa found for site: SW01 
2 contaminant taxa found for site: SW02 
1 contaminant taxa found for site: SW03 
7 contaminant taxa found for site: SW04 
1 contaminant taxa found for site: SW05 
No field control samples found for site: SW06 
2 contaminant taxa found for site: SW07 
1 contaminant taxa found for site: SW08 
73 contaminant taxa found for site: CN01 
6 contaminant taxa found for site: CN02 
45 contaminant taxa found for site: CN03 
17 contaminant taxa found for site: CN04 
8 contaminant taxa found for site: CN05 
5 contaminant taxa found for site: NE01 
2 contaminant taxa found for site: NE02 
3 contaminant taxa found for site: NE03 
5 contaminant taxa found for site: NE04 
8 contaminant taxa found for site: NE05 
No field control samples found for site: NW01 
2 contaminant taxa found for site: NW02 
6 contaminant taxa found for site: NW03 
3 contaminant taxa found for site: NW04 
3 contaminant taxa found for site: NW05 
1 contaminant taxa found for site: NW06 
Reads removed: 60165 ; Samples filtered: 474 
[1] "New dimensions of otu_table_filtered:"
[1] 1725  655</code></pre>
</div>
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_filtered_field_abun <span class="ot">&lt;-</span> result_field_abun[[<span class="dv">1</span>]]</span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>otus_removed_field_abun <span class="ot">&lt;-</span> result_field_abun[[<span class="dv">2</span>]]</span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb109-4"><a href="#cb109-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Call remove_extract_control_taxa_abund() function (extracts)</span></span>
<span id="cb109-5"><a href="#cb109-5" aria-hidden="true" tabindex="-1"></a>result_extract_abun <span class="ot">&lt;-</span> <span class="fu">remove_extract_control_taxa_abund</span>(phylo_eDNA_filtered_field_abun)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Original dimensions of otu_table_filtered:"
[1] 1725  594
1 contaminant taxa found for site: 1 
4 contaminant taxa found for site: 2 
2 contaminant taxa found for site: 3 
2 contaminant taxa found for site: 4 
3 contaminant taxa found for site: 5 
6 contaminant taxa found for site: 6 
69 contaminant taxa found for site: 9 
6 contaminant taxa found for site: 10 
10 contaminant taxa found for site: 8 
5 contaminant taxa found for site: 7 
Reads removed: 608314 ; Samples filtered: 497 
[1] "New dimensions of otu_table_filtered:"
[1] 1705  574</code></pre>
</div>
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_filtered_extract_abun <span class="ot">&lt;-</span> result_extract_abun[[<span class="dv">1</span>]]</span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>otus_removed_extract_abun <span class="ot">&lt;-</span> result_extract_abun[[<span class="dv">2</span>]]</span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb111-4"><a href="#cb111-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Call remove_PCR_CO1_control_taxa_abund() function (PCR)</span></span>
<span id="cb111-5"><a href="#cb111-5" aria-hidden="true" tabindex="-1"></a>result_PCR_CO1_abun <span class="ot">&lt;-</span> <span class="fu">remove_PCR_control_taxa_CO1_abund</span>(phylo_eDNA_filtered_extract_abun)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Original dimensions of otu_table_filtered:"
[1] 1705  574
More than 1 PCR (CO1) control sample found in batch 1 . Now printing individual controls: 
Batch 1 : No contaminants found for control indices 1 
Batch 1 : 1 contaminant taxa found for control indices 2 
Batch 1 : No contaminants found for control indices 3 
Batch 1 : No contaminants found for control indices 215 
1 contaminant taxa found for batch 2 
More than 1 PCR (CO1) control sample found in batch 10 . Now printing individual controls: 
Batch 10 : 4 contaminant taxa found for control indices 411 
Batch 10 : 8 contaminant taxa found for control indices 412 
Batch 10 : 20 contaminant taxa found for control indices 413 
Batch 10 : No contaminants found for control indices 414 
Batch 10 : 2 contaminant taxa found for control indices 415 
Batch 10 : No contaminants found for control indices 549 
No contaminants found for batch 9 
No contaminants found for batch 8 
Reads removed: 2461106 ; Samples filtered: 344 
 [1] "PHD1-NCbead_CO1"         "PHD1-NCrep_CO1"         
 [3] "PHD1-SCHNC-PCR_1703_CO1" "PHD1-SCHNC-PCR_2703_CO1"
 [5] "PHD1-SCHNC2803_CO1"      "PHD1-SWNC_PCR_2203_CO1" 
 [7] "PHD1-SWNC_PCR_2803_CO1"  "PHD2-Neg-PCR-0111_CO1"  
 [9] "PHD2-Neg-PCR-0311_CO1"   "PHD2-Neg-PCR-0611_CO1"  
[11] "PHD2-Neg-PCR-1411_CO1"   "PHD2-Neg-PCR-1511_CO1"  
[13] "PHD2-Neg-PCR-2011_CO1"   "PHD2-Neg-PCR-2111_CO1"  
[15] "PHD2-Neg-PCR-3010_CO1"   "PHD2-Neg-PCR-3110_CO1"  
[17] "PHD2-SCHNC2303_CO1"     
 [1]  1  2  3  5  6 10 11 12 13 14 17
[1] "New dimensions of otu_table_filtered:"
[1] 1698  563</code></pre>
</div>
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_filtered_PCR_CO1_abun <span class="ot">&lt;-</span> result_PCR_CO1_abun[[<span class="dv">1</span>]]</span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>otus_removed_PCR_CO1_abun <span class="ot">&lt;-</span> result_PCR_CO1_abun[[<span class="dv">2</span>]]</span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb113-4"><a href="#cb113-4" aria-hidden="true" tabindex="-1"></a><span class="co">#all remove_PCR_18S_control_taxa_abund() function (PCR)</span></span>
<span id="cb113-5"><a href="#cb113-5" aria-hidden="true" tabindex="-1"></a>result_PCR_18S_abun <span class="ot">&lt;-</span> <span class="fu">remove_PCR_control_taxa_18S_abund</span>(phylo_eDNA_filtered_PCR_CO1_abun)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Original dimensions of otu_table_filtered:"
[1] 1698  563
2 contaminant taxa found for batch 3 
No contaminants found for batch 4 
2 contaminant taxa found for batch 7 
5 contaminant taxa found for batch 6 
More than 1 PCR (18S) control sample found in batch 5 . Now printing individual controls: 
Batch 5 : 4 contaminant taxa found for control indices 403 
Batch 5 : No contaminants found for control indices 404 
Batch 5 : No contaminants found for control indices 406 
Reads removed: 1207399 ; Samples filtered: 297 
[1] "PHD1-NCrep_18S"          "PHD1-SCHNC-PCR_2703_18S"
[3] "PHD1-SWNC_PCR_2803_18S"  "PHD2-Neg-PCR-0111_18S"  
[5] "PHD2-Neg-PCR-0311_18S"   "PHD2-Neg-PCR-1511_18S"  
[7] "PHD2-Neg-PCR-2011_18S"   "PHD2-Neg-PCR-3010_18S"  
[9] "PHD2-Neg-PCR-3110_18S"  
[1] 1 2 3 4 5 6 7 8 9
[1] "New dimensions of otu_table_filtered:"
[1] 1692  554</code></pre>
</div>
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_filtered_PCR_18S_abun <span class="ot">&lt;-</span> result_PCR_18S_abun[[<span class="dv">1</span>]]</span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>otus_removed_PCR_18S_abun <span class="ot">&lt;-</span> result_PCR_18S_abun[[<span class="dv">2</span>]]</span>
<span id="cb115-3"><a href="#cb115-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb115-4"><a href="#cb115-4" aria-hidden="true" tabindex="-1"></a><span class="co"># rename file data</span></span>
<span id="cb115-5"><a href="#cb115-5" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam <span class="ot">&lt;-</span> phylo_eDNA_filtered_PCR_18S_abun</span>
<span id="cb115-6"><a href="#cb115-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb115-7"><a href="#cb115-7" aria-hidden="true" tabindex="-1"></a><span class="co"># remove taxa with no reads across all samples</span></span>
<span id="cb115-8"><a href="#cb115-8" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(phylo_eDNA_decontam, <span class="fu">taxa_sums</span>(phylo_eDNA_decontam) <span class="sc">&gt;</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s compared our decontaminated data with the original.</p>
<p>We can see that control samples have now been removed from the data and 51 taxa have also been removed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>phylo_eDNA</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 1725 taxa and 655 samples ]
sample_data() Sample Data:       [ 655 samples by 42 sample variables ]
tax_table()   Taxonomy Table:    [ 1725 taxa by 10 taxonomic ranks ]</code></pre>
</div>
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 1674 taxa and 554 samples ]
sample_data() Sample Data:       [ 554 samples by 42 sample variables ]
tax_table()   Taxonomy Table:    [ 1674 taxa by 10 taxonomic ranks ]</code></pre>
</div>
</div>
<p>We can also check to see how many samples are categorized under different sample types.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">meta</span>(phylo_eDNA_decontam)<span class="sc">$</span>sampleType, <span class="at">useNA =</span> <span class="st">"always"</span>) <span class="co">#sample type</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
       failed-repeat lab-positive-control      plate-replicate 
                   6                   30                   34 
       run-replicate               sample                 &lt;NA&gt; 
                  32                  452                    0 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">meta</span>(phylo_eDNA_decontam)<span class="sc">$</span>controlCheck, <span class="at">useNA =</span> <span class="st">"always"</span>) <span class="co">#control type</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
positive   sample     &lt;NA&gt; 
      30      524        0 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">meta</span>(phylo_eDNA_decontam)<span class="sc">$</span>shorePosition, <span class="at">useNA =</span> <span class="st">"always"</span>) <span class="co">#shore height</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    High      Low      Mid Subtidal    Tidal     &lt;NA&gt; 
     164      128       11        6      215       30 </code></pre>
</div>
</div>
<p>We need to tidy up the shore positions to high, low and open water only.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(phylo_eDNA_decontam)<span class="sc">$</span>shorePosition[<span class="fu">sample_data</span>(phylo_eDNA_decontam)<span class="sc">$</span>shorePosition <span class="sc">==</span> <span class="st">"Mid "</span>] <span class="ot">&lt;-</span> <span class="st">"High"</span></span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(phylo_eDNA_decontam)<span class="sc">$</span>shorePosition[<span class="fu">sample_data</span>(phylo_eDNA_decontam)<span class="sc">$</span>shorePosition <span class="sc">==</span> <span class="st">"Tidal"</span>] <span class="ot">&lt;-</span> <span class="st">"Open Water"</span></span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(phylo_eDNA_decontam)<span class="sc">$</span>shorePosition[<span class="fu">sample_data</span>(phylo_eDNA_decontam)<span class="sc">$</span>shorePosition <span class="sc">==</span> <span class="st">"Subtidal"</span>] <span class="ot">&lt;-</span> <span class="st">"Open Water"</span></span>
<span id="cb126-4"><a href="#cb126-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(phylo_eDNA_decontam)<span class="sc">$</span>shorePosition[<span class="fu">sample_data</span>(phylo_eDNA_decontam)<span class="sc">$</span>shorePosition <span class="sc">==</span> <span class="st">"Mid"</span>] <span class="ot">&lt;-</span> <span class="st">"High"</span></span>
<span id="cb126-5"><a href="#cb126-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb126-6"><a href="#cb126-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">meta</span>(phylo_eDNA_decontam)<span class="sc">$</span>shorePosition, <span class="at">useNA =</span> <span class="st">"always"</span>) <span class="co">#shore height</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
      High        Low Open Water       &lt;NA&gt; 
       175        128        221         30 </code></pre>
</div>
</div>
<p>We also don’t require our positive controls now.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam <span class="ot">&lt;-</span> <span class="fu">subset_samples</span>(phylo_eDNA_decontam, <span class="sc">!</span>(controlCheck <span class="sc">==</span> <span class="st">"positive"</span>)) </span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">meta</span>(phylo_eDNA_decontam)<span class="sc">$</span>sampleType, <span class="at">useNA =</span> <span class="st">"always"</span>) <span class="co">#sample type</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  failed-repeat plate-replicate   run-replicate          sample            &lt;NA&gt; 
              6              34              32             452               0 </code></pre>
</div>
</div>
</section>
</section>
<section id="filtering-steps" class="level2">
<h2 class="anchored" data-anchor-id="filtering-steps">Filtering steps</h2>
<section id="removal-of-low-read-samples" class="level3">
<h3 class="anchored" data-anchor-id="removal-of-low-read-samples">Removal of low read samples</h3>
<p>It’s good practice to remove any samples with the low read depth to prevent sampling bias driving ecological patterns. We are going to explore the distribution of read depths across samples to make a descision on a minimum depth threshold.</p>
<p>We first need to calculate read depth per sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of reads per sample (before filtering)</span></span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>sample_depths_decontam <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">readcount</span>(phylo_eDNA_decontam) <span class="sc">%&gt;%</span></span>
<span id="cb130-3"><a href="#cb130-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb130-4"><a href="#cb130-4" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"fullID"</span>)</span>
<span id="cb130-5"><a href="#cb130-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-6"><a href="#cb130-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sample_depths_decontam) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"fullID"</span>, <span class="st">"reads_prefilter"</span>)</span>
<span id="cb130-7"><a href="#cb130-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-8"><a href="#cb130-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the total ASVs by counting the number of rows </span></span>
<span id="cb130-9"><a href="#cb130-9" aria-hidden="true" tabindex="-1"></a>num_asvs_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(phylo_eDNA_decontam)))  </span>
<span id="cb130-10"><a href="#cb130-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(num_asvs_vec)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"no.taxa"</span> </span>
<span id="cb130-11"><a href="#cb130-11" aria-hidden="true" tabindex="-1"></a>num_asvs_vec</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>no.taxa 
   1674 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Total number asvs per sample </span></span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>asv_sum <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">colSums</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(phylo_eDNA_decontam)<span class="sc">!=</span><span class="dv">0</span>))  </span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Extract sample meta data as a separate R object </span></span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a>abundance_metadf <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(phylo_eDNA_decontam) </span>
<span id="cb132-6"><a href="#cb132-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-7"><a href="#cb132-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Check if vector of sample_depths &amp; asv_sum has the same order as our metadata rows</span></span>
<span id="cb132-8"><a href="#cb132-8" aria-hidden="true" tabindex="-1"></a><span class="fu">identical</span>(sample_depths_decontam<span class="sc">$</span>fullID,<span class="fu">row.names</span>(abundance_metadf)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="fu">identical</span>(<span class="fu">row.names</span>(asv_sum),<span class="fu">row.names</span>(abundance_metadf))  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Add sample depths and number of ASVs to metadata data frame </span></span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a>abundance_metadf[,<span class="st">"depth"</span>] <span class="ot">&lt;-</span> sample_depths_decontam<span class="sc">$</span>reads_prefilter </span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a>abundance_metadf[,<span class="st">"ASVs"</span>] <span class="ot">&lt;-</span> asv_sum</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check the distribution of sample reads.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>pcr_seq_depth_overall <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(abundance_metadf, <span class="fu">aes</span>(depth)) <span class="sc">+</span></span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">60</span>, <span class="at">color =</span> <span class="st">"grey36"</span>, <span class="at">fill =</span> <span class="st">"white"</span>) <span class="sc">+</span>   </span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">3000</span>, <span class="at">lty =</span> <span class="dv">2</span>,<span class="at">color =</span> <span class="st">"orange"</span>) <span class="sc">+</span> <span class="co"># threshold</span></span>
<span id="cb137-4"><a href="#cb137-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span>   </span>
<span id="cb137-5"><a href="#cb137-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"# Reads"</span>, <span class="at">y =</span> <span class="st">"# Samples"</span>) <span class="sc">+</span></span>
<span id="cb137-6"><a href="#cb137-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb137-7"><a href="#cb137-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span>
<span id="cb137-8"><a href="#cb137-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-9"><a href="#cb137-9" aria-hidden="true" tabindex="-1"></a>pcr_seq_depth_overall </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="intertidal-eDNA-formatting-and-quality-control_files/figure-html/unnamed-chunk-72-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now let’s flag samples with an acceptable (TRUE) or inacceptable (FALSE) sequencing depth. We should aim to keep at least 90% of samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>abundance_metadf<span class="sc">$</span>depth_ok <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(abundance_metadf<span class="sc">$</span>depth <span class="sc">&lt;</span> <span class="dv">3000</span>, F, T)</span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb138-3"><a href="#cb138-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(abundance_metadf<span class="sc">$</span>depth_ok[abundance_metadf<span class="sc">$</span>negCheckLogical<span class="sc">==</span><span class="st">"FALSE"</span>]) <span class="sc">/</span></span>
<span id="cb138-4"><a href="#cb138-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">nrow</span>(abundance_metadf[abundance_metadf<span class="sc">$</span>negCheckLogical<span class="sc">==</span><span class="st">"FALSE"</span>,])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     FALSE       TRUE 
0.03625954 0.96374046 </code></pre>
</div>
</div>
<p>A minimum threshold of 3000 reads looks like it could work well. Let’s now filter our data set to remove any samples below 3000.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a>sample_depths_object <span class="ot">&lt;-</span> <span class="fu">sample_sums</span>(phylo_eDNA_decontam) <span class="co"># Calculate sample sequencing depths</span></span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a>min_reads <span class="ot">&lt;-</span> <span class="dv">3000</span> <span class="co"># Define the minimum read threshold</span></span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-4"><a href="#cb140-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Prune samples with fewer than 3000 reads</span></span>
<span id="cb140-5"><a href="#cb140-5" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam_filtered <span class="ot">&lt;-</span> <span class="fu">prune_samples</span>(sample_depths_object <span class="sc">&gt;=</span> min_reads, phylo_eDNA_decontam)</span>
<span id="cb140-6"><a href="#cb140-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-7"><a href="#cb140-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Check the number of samples before and after filtering</span></span>
<span id="cb140-8"><a href="#cb140-8" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Number of samples before filtering:"</span>, <span class="fu">nsamples</span>(phylo_eDNA_decontam), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of samples before filtering: 524 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Number of samples after filtering:"</span>, <span class="fu">nsamples</span>(phylo_eDNA_decontam_filtered), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>) <span class="co">#removed 19 samples</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of samples after filtering: 505 </code></pre>
</div>
</div>
</section>
<section id="filtering-for-target-functional-groups-broad" class="level3">
<h3 class="anchored" data-anchor-id="filtering-for-target-functional-groups-broad">Filtering for target functional groups (broad)</h3>
<p>For this project, we were only interested in marine invertebrates and algae. So, we need to remove any other type of taxa. We’ll start with a broad removal, then focus on species phyla.</p>
<p>Let’s first examine what groups we have.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(<span class="fu">tax_table</span>(phylo_eDNA_decontam_filtered)[, <span class="st">"adult"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Taxonomy Table:     [22 taxa by 1 taxonomic ranks]:
                            adult          
Abra nitida                 "benthos"      
Acanthochiasmidae           "phytoplankton"
Acartia tonsa cryophylla    "mesoplankton" 
Acinetospora filamentosa    "macroalgae"   
Acoela                      "macrobenthos" 
Acremonium fuci             NA             
Alcyonium                   "megabenthos"  
Alexandrium andersonii      "mixoplankton" 
Alexandrium minutum         "microplankton"
Allas                       "zooplankton"  
Ameira scotti               "meiobenthos"  
Ammodytes marinus           "fish"         
Amphidiniopsis              "nanoplankton" 
Amphidiniopsis rotundata    "microbenthos" 
Archilopsis spinosa         "turbellaria"  
Aureococcus anophagefferens "algae"        
Bufo bufo                   "nekton"       
Calanus                     "macroplankton"
Crinoidea                   "epibenthos"   
Duboscquella                "plankton"     
Echinocardium cordatum      "endobenthos"  
Halodule uninervis          "phytobenthos" </code></pre>
</div>
</div>
<p>Define which groups not of interest.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a>fish_and_small_stuff <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"fish"</span>, </span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"microplankton"</span>, </span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"phytoplankton"</span>, </span>
<span id="cb146-4"><a href="#cb146-4" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"nanoplankton"</span>, </span>
<span id="cb146-5"><a href="#cb146-5" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"mixoplankton"</span>, </span>
<span id="cb146-6"><a href="#cb146-6" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"microbenthos"</span>)</span>
<span id="cb146-7"><a href="#cb146-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-8"><a href="#cb146-8" aria-hidden="true" tabindex="-1"></a>other_unwanted_class <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Aves"</span>, </span>
<span id="cb146-9"><a href="#cb146-9" aria-hidden="true" tabindex="-1"></a>                          <span class="st">"Mammalia"</span>, </span>
<span id="cb146-10"><a href="#cb146-10" aria-hidden="true" tabindex="-1"></a>                          <span class="st">"Amphibia"</span>)</span>
<span id="cb146-11"><a href="#cb146-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-12"><a href="#cb146-12" aria-hidden="true" tabindex="-1"></a>other_unwanted_order <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Squamata"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s remove them from the relevant groups.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam_filtered <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(</span>
<span id="cb147-2"><a href="#cb147-2" aria-hidden="true" tabindex="-1"></a>  phylo_eDNA_decontam_filtered,</span>
<span id="cb147-3"><a href="#cb147-3" aria-hidden="true" tabindex="-1"></a>  <span class="sc">!</span>(adult <span class="sc">%in%</span> fish_and_small_stuff) <span class="sc">&amp;</span> </span>
<span id="cb147-4"><a href="#cb147-4" aria-hidden="true" tabindex="-1"></a>  <span class="sc">!</span>(class <span class="sc">%in%</span> other_unwanted_class) <span class="sc">&amp;</span> </span>
<span id="cb147-5"><a href="#cb147-5" aria-hidden="true" tabindex="-1"></a>  <span class="sc">!</span>(order <span class="sc">%in%</span> other_unwanted_order) </span>
<span id="cb147-6"><a href="#cb147-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb147-7"><a href="#cb147-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb147-8"><a href="#cb147-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Rmove any taxa without occurances</span></span>
<span id="cb147-9"><a href="#cb147-9" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam_filtered <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(phylo_eDNA_decontam_filtered, <span class="fu">taxa_sums</span>(phylo_eDNA_decontam_filtered) <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb147-10"><a href="#cb147-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb147-11"><a href="#cb147-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb147-12"><a href="#cb147-12" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam_filtered</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 1398 taxa and 505 samples ]
sample_data() Sample Data:       [ 505 samples by 42 sample variables ]
tax_table()   Taxonomy Table:    [ 1398 taxa by 10 taxonomic ranks ]</code></pre>
</div>
</div>
</section>
<section id="filtering-for-target-taxa-more-specific" class="level3">
<h3 class="anchored" data-anchor-id="filtering-for-target-taxa-more-specific">Filtering for target taxa (more specific)</h3>
<p>Let’s define which phyla we are interested in.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a>target_phyla <span class="ot">&lt;-</span>   <span class="fu">c</span>(</span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Acoela"</span>,</span>
<span id="cb149-3"><a href="#cb149-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Amphioxus"</span>,</span>
<span id="cb149-4"><a href="#cb149-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Annelida"</span>,</span>
<span id="cb149-5"><a href="#cb149-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Arthropoda"</span>,</span>
<span id="cb149-6"><a href="#cb149-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Brachiopoda"</span>,</span>
<span id="cb149-7"><a href="#cb149-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Bryozoa"</span>,</span>
<span id="cb149-8"><a href="#cb149-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Cephalochordata"</span>,</span>
<span id="cb149-9"><a href="#cb149-9" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Chaetognatha"</span>,</span>
<span id="cb149-10"><a href="#cb149-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Chlorophyta"</span>,</span>
<span id="cb149-11"><a href="#cb149-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Chordata"</span>,</span>
<span id="cb149-12"><a href="#cb149-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Cnidaria"</span>,</span>
<span id="cb149-13"><a href="#cb149-13" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Crustacea"</span>,</span>
<span id="cb149-14"><a href="#cb149-14" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Ctenophora"</span>,</span>
<span id="cb149-15"><a href="#cb149-15" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Dinoflagellata"</span>,</span>
<span id="cb149-16"><a href="#cb149-16" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Echiura"</span>,</span>
<span id="cb149-17"><a href="#cb149-17" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Echinoderm"</span>,</span>
<span id="cb149-18"><a href="#cb149-18" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Echinodermata"</span>,</span>
<span id="cb149-19"><a href="#cb149-19" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Euglenophyta"</span>,</span>
<span id="cb149-20"><a href="#cb149-20" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Gastrotricha"</span>,</span>
<span id="cb149-21"><a href="#cb149-21" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Gnathostomulids"</span>,</span>
<span id="cb149-22"><a href="#cb149-22" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Gyrista"</span>,</span>
<span id="cb149-23"><a href="#cb149-23" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Hemichordata"</span>,</span>
<span id="cb149-24"><a href="#cb149-24" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Kamptozoa"</span>,</span>
<span id="cb149-25"><a href="#cb149-25" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Kinorhyncha"</span>,</span>
<span id="cb149-26"><a href="#cb149-26" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Loricifera"</span>,</span>
<span id="cb149-27"><a href="#cb149-27" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Mollusca"</span>,</span>
<span id="cb149-28"><a href="#cb149-28" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Monogenea"</span>,</span>
<span id="cb149-29"><a href="#cb149-29" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Myzostomida"</span>,</span>
<span id="cb149-30"><a href="#cb149-30" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Nematoda"</span>,</span>
<span id="cb149-31"><a href="#cb149-31" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Nemertinea"</span>,</span>
<span id="cb149-32"><a href="#cb149-32" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Ochrophyta"</span>,</span>
<span id="cb149-33"><a href="#cb149-33" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Orthonectida"</span>,</span>
<span id="cb149-34"><a href="#cb149-34" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Phaeophyta"</span>,</span>
<span id="cb149-35"><a href="#cb149-35" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Phoronida"</span>,</span>
<span id="cb149-36"><a href="#cb149-36" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Placozoa"</span>,</span>
<span id="cb149-37"><a href="#cb149-37" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Platyhelminthes"</span>,</span>
<span id="cb149-38"><a href="#cb149-38" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Porifera"</span>,</span>
<span id="cb149-39"><a href="#cb149-39" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Priapulida"</span>,</span>
<span id="cb149-40"><a href="#cb149-40" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Pycnogonida"</span>,</span>
<span id="cb149-41"><a href="#cb149-41" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Rhodophyta"</span>,</span>
<span id="cb149-42"><a href="#cb149-42" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Sipunculida"</span>,</span>
<span id="cb149-43"><a href="#cb149-43" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Thallophyta"</span>,</span>
<span id="cb149-44"><a href="#cb149-44" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Tunicata"</span>,</span>
<span id="cb149-45"><a href="#cb149-45" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Turbellaria"</span>,</span>
<span id="cb149-46"><a href="#cb149-46" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Xenoturbella"</span>,</span>
<span id="cb149-47"><a href="#cb149-47" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Xiphosura"</span></span>
<span id="cb149-48"><a href="#cb149-48" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s remove these from the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam_filtered <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(phylo_eDNA_decontam_filtered,                                              phylum <span class="sc">%in%</span> target_phyla)</span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Rmove any taxa without occurances</span></span>
<span id="cb150-4"><a href="#cb150-4" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam_filtered <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(phylo_eDNA_decontam_filtered, <span class="fu">taxa_sums</span>(phylo_eDNA_decontam_filtered) <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb150-5"><a href="#cb150-5" aria-hidden="true" tabindex="-1"></a>                                            </span>
<span id="cb150-6"><a href="#cb150-6" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam_filtered</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 1026 taxa and 505 samples ]
sample_data() Sample Data:       [ 505 samples by 42 sample variables ]
tax_table()   Taxonomy Table:    [ 1026 taxa by 10 taxonomic ranks ]</code></pre>
</div>
</div>
<p>Let’s check to see how many taxa we have for invertebrates and macrophytes.</p>
<p>First, let’s define our groups.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a>invert_phyla <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb152-2"><a href="#cb152-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Acoela"</span>,</span>
<span id="cb152-3"><a href="#cb152-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Amphioxus"</span>,</span>
<span id="cb152-4"><a href="#cb152-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Annelida"</span>,</span>
<span id="cb152-5"><a href="#cb152-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Arthropoda"</span>,</span>
<span id="cb152-6"><a href="#cb152-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Brachiopoda"</span>,</span>
<span id="cb152-7"><a href="#cb152-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Bryozoa"</span>,</span>
<span id="cb152-8"><a href="#cb152-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Cephalochordata"</span>,</span>
<span id="cb152-9"><a href="#cb152-9" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Chaetognatha"</span>,</span>
<span id="cb152-10"><a href="#cb152-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Chordata"</span>,</span>
<span id="cb152-11"><a href="#cb152-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Cnidaria"</span>,</span>
<span id="cb152-12"><a href="#cb152-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Crustacea"</span>,</span>
<span id="cb152-13"><a href="#cb152-13" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Echiura"</span>,</span>
<span id="cb152-14"><a href="#cb152-14" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Echinoderm"</span>,</span>
<span id="cb152-15"><a href="#cb152-15" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Echinodermata"</span>,</span>
<span id="cb152-16"><a href="#cb152-16" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Gastrotricha"</span>,</span>
<span id="cb152-17"><a href="#cb152-17" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Gnathostomulids"</span>,</span>
<span id="cb152-18"><a href="#cb152-18" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Hemichordata"</span>,</span>
<span id="cb152-19"><a href="#cb152-19" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Kamptozoa"</span>,</span>
<span id="cb152-20"><a href="#cb152-20" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Kinorhyncha"</span>,</span>
<span id="cb152-21"><a href="#cb152-21" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Loricifera"</span>,</span>
<span id="cb152-22"><a href="#cb152-22" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Mollusca"</span>,</span>
<span id="cb152-23"><a href="#cb152-23" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Monogenea"</span>,</span>
<span id="cb152-24"><a href="#cb152-24" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Myzostomida"</span>,</span>
<span id="cb152-25"><a href="#cb152-25" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Nematoda"</span>,</span>
<span id="cb152-26"><a href="#cb152-26" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Nemertinea"</span>,</span>
<span id="cb152-27"><a href="#cb152-27" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Orthonectida"</span>,</span>
<span id="cb152-28"><a href="#cb152-28" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Phaeophyta"</span>,</span>
<span id="cb152-29"><a href="#cb152-29" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Phoronida"</span>,</span>
<span id="cb152-30"><a href="#cb152-30" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Placozoa"</span>,</span>
<span id="cb152-31"><a href="#cb152-31" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Platyhelminthes"</span>,</span>
<span id="cb152-32"><a href="#cb152-32" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Porifera"</span>,</span>
<span id="cb152-33"><a href="#cb152-33" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Priapulida"</span>,</span>
<span id="cb152-34"><a href="#cb152-34" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Pycnogonida"</span>,</span>
<span id="cb152-35"><a href="#cb152-35" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Sipunculida"</span>,</span>
<span id="cb152-36"><a href="#cb152-36" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Thallophyta"</span>,</span>
<span id="cb152-37"><a href="#cb152-37" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Tunicata"</span>,</span>
<span id="cb152-38"><a href="#cb152-38" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Xenoturbella"</span>,</span>
<span id="cb152-39"><a href="#cb152-39" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Xiphosura"</span></span>
<span id="cb152-40"><a href="#cb152-40" aria-hidden="true" tabindex="-1"></a>)                    </span>
<span id="cb152-41"><a href="#cb152-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-42"><a href="#cb152-42" aria-hidden="true" tabindex="-1"></a>algae_phyla <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb152-43"><a href="#cb152-43" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Chlorophyta"</span>,</span>
<span id="cb152-44"><a href="#cb152-44" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Ctenophora"</span>,</span>
<span id="cb152-45"><a href="#cb152-45" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Dinoflagellata"</span>,</span>
<span id="cb152-46"><a href="#cb152-46" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Euglenophyta"</span>,</span>
<span id="cb152-47"><a href="#cb152-47" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Gyrista"</span>,</span>
<span id="cb152-48"><a href="#cb152-48" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Ochrophyta"</span>,</span>
<span id="cb152-49"><a href="#cb152-49" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Phaeophyta"</span>,</span>
<span id="cb152-50"><a href="#cb152-50" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Rhodophyta"</span>,</span>
<span id="cb152-51"><a href="#cb152-51" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Thallophyta"</span></span>
<span id="cb152-52"><a href="#cb152-52" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s look at invertebrates first.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam_filtered_invert <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(phylo_eDNA_decontam_filtered, phylum <span class="sc">%in%</span> invert_phyla)  </span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb153-3"><a href="#cb153-3" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam_filtered_invert   </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 737 taxa and 505 samples ]
sample_data() Sample Data:       [ 505 samples by 42 sample variables ]
tax_table()   Taxonomy Table:    [ 737 taxa by 10 taxonomic ranks ]</code></pre>
</div>
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Invertebrate species account for"</span>, (<span class="dv">737</span> <span class="sc">/</span> <span class="dv">1026</span>) <span class="sc">*</span> <span class="dv">100</span>, <span class="st">"% of detections."</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Invertebrate species account for 71.8323586744639 % of detections."</code></pre>
</div>
</div>
<p>Now let’s look at algae.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam_filtered_algae <span class="ot">&lt;-</span> <span class="fu">subset_taxa</span>(phylo_eDNA_decontam_filtered, phylum <span class="sc">%in%</span> algae_phyla)  </span>
<span id="cb157-2"><a href="#cb157-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-3"><a href="#cb157-3" aria-hidden="true" tabindex="-1"></a>phylo_eDNA_decontam_filtered_algae  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 289 taxa and 505 samples ]
sample_data() Sample Data:       [ 505 samples by 42 sample variables ]
tax_table()   Taxonomy Table:    [ 289 taxa by 10 taxonomic ranks ]</code></pre>
</div>
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Algae species account for"</span>, (<span class="dv">289</span> <span class="sc">/</span> <span class="dv">1026</span>) <span class="sc">*</span> <span class="dv">100</span>, <span class="st">"% of detections."</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Algae species account for 28.1676413255361 % of detections."</code></pre>
</div>
</div>
</section>
</section>
<section id="further-quality-control" class="level2">
<h2 class="anchored" data-anchor-id="further-quality-control">Further quality control</h2>
<p>Now we’ve got our decontaminated data, we can do a variety of quality control checks to make sure everything looks like it should.</p>
<section id="set-up-functions-and-variables" class="level3">
<h3 class="anchored" data-anchor-id="set-up-functions-and-variables">Set-up (functions and variables)</h3>
<p>Let’s set-up some functions we’ll use later on to extract random pieces of information from our data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb161"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb161-1"><a href="#cb161-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract fieldID when adding TRUE or FALSE for replicate status</span></span>
<span id="cb161-2"><a href="#cb161-2" aria-hidden="true" tabindex="-1"></a>extract_middle <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb161-3"><a href="#cb161-3" aria-hidden="true" tabindex="-1"></a>  middle <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"^[^-]+-([A-Za-z]{2}</span><span class="sc">\\</span><span class="st">d{2}-</span><span class="sc">\\</span><span class="st">d{2}).*_(18S|CO1)$"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, x)</span>
<span id="cb161-4"><a href="#cb161-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(middle)</span>
<span id="cb161-5"><a href="#cb161-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb161-6"><a href="#cb161-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb161-7"><a href="#cb161-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract site when adding TRUE or FALSE for replicate status</span></span>
<span id="cb161-8"><a href="#cb161-8" aria-hidden="true" tabindex="-1"></a>extract_site <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb161-9"><a href="#cb161-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extract the part after the first '-' and before the second '-' or '_' and remove anything after</span></span>
<span id="cb161-10"><a href="#cb161-10" aria-hidden="true" tabindex="-1"></a>  middle <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"^[^-]+-([^-]+)-.*_(18S|CO1)$"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, x)</span>
<span id="cb161-11"><a href="#cb161-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(middle)</span>
<span id="cb161-12"><a href="#cb161-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb161-13"><a href="#cb161-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb161-14"><a href="#cb161-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb161-15"><a href="#cb161-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the middle part (including fieldID and primer)</span></span>
<span id="cb161-16"><a href="#cb161-16" aria-hidden="true" tabindex="-1"></a>extract_middle_all <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb161-17"><a href="#cb161-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extract the middle part of the string</span></span>
<span id="cb161-18"><a href="#cb161-18" aria-hidden="true" tabindex="-1"></a>  middle <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"^[^-]+-(.+)$"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, x)</span>
<span id="cb161-19"><a href="#cb161-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb161-20"><a href="#cb161-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove the term 'rep' and any following number</span></span>
<span id="cb161-21"><a href="#cb161-21" aria-hidden="true" tabindex="-1"></a>  middle <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"rep</span><span class="sc">\\</span><span class="st">d*"</span>, <span class="st">""</span>, middle, <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb161-22"><a href="#cb161-22" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb161-23"><a href="#cb161-23" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Return the modified middle part</span></span>
<span id="cb161-24"><a href="#cb161-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(middle)</span>
<span id="cb161-25"><a href="#cb161-25" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb161-26"><a href="#cb161-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb161-27"><a href="#cb161-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Modify the extract_middle function to extract the fieldID part for CO1 only groups (run 3)</span></span>
<span id="cb161-28"><a href="#cb161-28" aria-hidden="true" tabindex="-1"></a>extract_middle_CO1 <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb161-29"><a href="#cb161-29" aria-hidden="true" tabindex="-1"></a>  middle <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"^[^-]+-([A-Za-z]{2}</span><span class="sc">\\</span><span class="st">d{2}-</span><span class="sc">\\</span><span class="st">d{2}).*_CO1$"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, x)</span>
<span id="cb161-30"><a href="#cb161-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(middle)</span>
<span id="cb161-31"><a href="#cb161-31" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We need to set some variables for plots later.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a>site_order <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Scourie"</span>, </span>
<span id="cb162-2"><a href="#cb162-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Rispond"</span>, </span>
<span id="cb162-3"><a href="#cb162-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Skerray"</span>, </span>
<span id="cb162-4"><a href="#cb162-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Murkle Bay"</span>, </span>
<span id="cb162-5"><a href="#cb162-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Portskerra"</span>, </span>
<span id="cb162-6"><a href="#cb162-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Borwick, Yesnaby"</span>, </span>
<span id="cb162-7"><a href="#cb162-7" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Sannick"</span>, </span>
<span id="cb162-8"><a href="#cb162-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Wick"</span>, <span class="co">#Scotland</span></span>
<span id="cb162-9"><a href="#cb162-9" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Aberystwyth"</span>, </span>
<span id="cb162-10"><a href="#cb162-10" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Neyland"</span>, </span>
<span id="cb162-11"><a href="#cb162-11" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Broad Haven"</span>, </span>
<span id="cb162-12"><a href="#cb162-12" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Skomer Island"</span>, </span>
<span id="cb162-13"><a href="#cb162-13" aria-hidden="true" tabindex="-1"></a>                <span class="st">"West Angle"</span>, </span>
<span id="cb162-14"><a href="#cb162-14" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Monkstone Point"</span>, </span>
<span id="cb162-15"><a href="#cb162-15" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Dale Jetty"</span>, </span>
<span id="cb162-16"><a href="#cb162-16" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Martin's Haven"</span>, <span class="co">#South Wales</span></span>
<span id="cb162-17"><a href="#cb162-17" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Castlehead Rocks"</span>, </span>
<span id="cb162-18"><a href="#cb162-18" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Filey Brigg"</span>, </span>
<span id="cb162-19"><a href="#cb162-19" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Newton Point"</span>, </span>
<span id="cb162-20"><a href="#cb162-20" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Rumbling Kern"</span>, </span>
<span id="cb162-21"><a href="#cb162-21" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Scalby Mills"</span>, <span class="co">#Northumbria</span></span>
<span id="cb162-22"><a href="#cb162-22" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Great Orme East"</span>, </span>
<span id="cb162-23"><a href="#cb162-23" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Little Orme"</span>, </span>
<span id="cb162-24"><a href="#cb162-24" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Menai Bridge"</span>, </span>
<span id="cb162-25"><a href="#cb162-25" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Porth Oer"</span>, </span>
<span id="cb162-26"><a href="#cb162-26" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Porth Swtan"</span>, </span>
<span id="cb162-27"><a href="#cb162-27" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Rhosneigr"</span>, <span class="co">#North Wales</span></span>
<span id="cb162-28"><a href="#cb162-28" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Lizard Point"</span>, </span>
<span id="cb162-29"><a href="#cb162-29" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Looe"</span>, </span>
<span id="cb162-30"><a href="#cb162-30" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Sennen Cove"</span>, </span>
<span id="cb162-31"><a href="#cb162-31" aria-hidden="true" tabindex="-1"></a>                <span class="st">"St Ives"</span>, </span>
<span id="cb162-32"><a href="#cb162-32" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Trevone"</span>) <span class="co">#Cornwall</span></span>
<span id="cb162-33"><a href="#cb162-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb162-34"><a href="#cb162-34" aria-hidden="true" tabindex="-1"></a>control_order <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"field-control"</span>, </span>
<span id="cb162-35"><a href="#cb162-35" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"lab-extraction-control"</span>, </span>
<span id="cb162-36"><a href="#cb162-36" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"lab-negative-control"</span>, </span>
<span id="cb162-37"><a href="#cb162-37" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"lab-positive-control"</span>, </span>
<span id="cb162-38"><a href="#cb162-38" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"sample"</span>, </span>
<span id="cb162-39"><a href="#cb162-39" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"run-replicate"</span>, </span>
<span id="cb162-40"><a href="#cb162-40" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"failed-repeat"</span>, </span>
<span id="cb162-41"><a href="#cb162-41" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"plate-replicate"</span>)</span>
<span id="cb162-42"><a href="#cb162-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb162-43"><a href="#cb162-43" aria-hidden="true" tabindex="-1"></a>control_order_tidy <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Field control"</span>, </span>
<span id="cb162-44"><a href="#cb162-44" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Lab extraction control"</span>, </span>
<span id="cb162-45"><a href="#cb162-45" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Lab negative control"</span>, </span>
<span id="cb162-46"><a href="#cb162-46" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Lab positive control"</span>, </span>
<span id="cb162-47"><a href="#cb162-47" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Sample"</span>, </span>
<span id="cb162-48"><a href="#cb162-48" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Run replicate"</span>, </span>
<span id="cb162-49"><a href="#cb162-49" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Failed repeat"</span>, </span>
<span id="cb162-50"><a href="#cb162-50" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Plate replicate"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="explore-read-depths" class="level3">
<h3 class="anchored" data-anchor-id="explore-read-depths">Explore read depths</h3>
<p>Let’s summarize the number of reads before and after decontamination and filtering. We calculated the pre-filtered data above, so now let’s calculate the post filtered data and compare them.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of reads per sample (after decontamination and filtering)</span></span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a>sample_depths_decontam_filter <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">readcount</span>(phylo_eDNA_decontam_filtered) <span class="sc">%&gt;%</span> </span>
<span id="cb163-3"><a href="#cb163-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb163-4"><a href="#cb163-4" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"fullID"</span>)</span>
<span id="cb163-5"><a href="#cb163-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb163-6"><a href="#cb163-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sample_depths_decontam_filter) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"fullID"</span>, <span class="st">"reads_postfilter"</span>)</span>
<span id="cb163-7"><a href="#cb163-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb163-8"><a href="#cb163-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Reads before and after decontamination together</span></span>
<span id="cb163-9"><a href="#cb163-9" aria-hidden="true" tabindex="-1"></a>all_depths <span class="ot">&lt;-</span> <span class="fu">left_join</span>(sample_depths_decontam, </span>
<span id="cb163-10"><a href="#cb163-10" aria-hidden="true" tabindex="-1"></a>                        sample_depths_decontam_filter, </span>
<span id="cb163-11"><a href="#cb163-11" aria-hidden="true" tabindex="-1"></a>                        <span class="at">by =</span> <span class="st">"fullID"</span>)</span>
<span id="cb163-12"><a href="#cb163-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb163-13"><a href="#cb163-13" aria-hidden="true" tabindex="-1"></a><span class="co">#add difference between start and end</span></span>
<span id="cb163-14"><a href="#cb163-14" aria-hidden="true" tabindex="-1"></a>all_depths<span class="sc">$</span>reads_removed <span class="ot">&lt;-</span> all_depths<span class="sc">$</span>reads_prefilter <span class="sc">-</span> all_depths<span class="sc">$</span>reads_postfilter </span>
<span id="cb163-15"><a href="#cb163-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb163-16"><a href="#cb163-16" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(all_depths)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  fullID reads_prefilter reads_postfilter reads_removed
519 PHD3-SCH04-09rep_CO1           66980             4956         62024
520 PHD3-SCH07-02rep_CO1           75403            54742         20661
521  PHD3-SW01-06rep_CO1           49183             9203         39980
522  PHD3-SW03-08rep_CO1           32311              498         31813
523  PHD3-SW05-01rep_CO1           20020              165         19855
524  PHD3-SW07-03rep_CO1           37441             1455         35986</code></pre>
</div>
</div>
<p>Let’s create some figures to visualize read depths before and after filtering. We are going to create some new data set for plotting.</p>
<p>We did the pre-filtered data above. Let’s do the filtered data now.</p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>